// @mtfe/sc-web/src/config/const.ts
/**
 * Created by dongchunhai on 6/9/20.
 */
interface DefaultSCAccessConfig {
  readonly [key: string]: {
    accessEnv: string,
    http: {
      apiHost: string,
      apiHost_out_mt?: string,
      apiHost_out_sk?: string,
    },
    id: string,
    loginUrl: string,
    loginUrl_out_mt?: string,
    loginUrl_out_sk?: string,
    loginOrigin: string,
    loginOrigin_out_mt?: string,
    loginOrigin_out_sk?: string,
  };
}

export const DEFAULT_SC_ACCESS_CONFIG: DefaultSCAccessConfig = {
  // accessEnv: 'development',
  // http: {
  //   // apiHost: 'http://118.24.147.98:3330',
  //   apiHost: 'https://proxy.fe.test.sankuai.com',
  //   // apiHost: 'http://localhost:7070',
  // },
  // id: 'development_master',
  // loginUrl: 'https://proxy.fe.test.sankuai.com',
  // // loginUrl: 'http://localhost:7070/login',
  // loginOrigin: 'https://proxy.fe.test.sankuai.com',
  // // loginOrigin: 'http://localhost:7070',
  test: {
    accessEnv: 'test',
    http: {
      apiHost: 'https://proxy.fe.test.sankuai.com',
    },
    id: 'test_master',
    loginUrl: 'https://proxy.fe.test.sankuai.com/login',
    loginOrigin: 'https://proxy.fe.test.sankuai.com',
  },
  staging: {
    accessEnv: 'staging',
    http: {
      apiHost: 'https://proxy.fe.st.sankuai.com',
    },
    id: 'staging_master',
    loginUrl: 'https://proxy.fe.st.sankuai.com/login',
    loginOrigin: 'https://proxy.fe.st.sankuai.com',
  },
  production: {
    accessEnv: 'production',
    http: {
      apiHost: 'https://proxy.sankuai.com',
      apiHost_out_mt: 'https://service-component.meituan.com',
      apiHost_out_sk: 'https://proxy-pub.sankuai.com',
    },
    id: 'production_master',
    loginUrl: 'https://proxy.sankuai.com/login',
    loginOrigin: 'https://proxy.sankuai.com',
    loginUrl_out_mt: 'https://service-component.meituan.com/login',
    loginOrigin_out_mt: 'https://service-component.meituan.com',
    loginUrl_out_sk: 'https://proxy-pub.sankuai.com/login',
    loginOrigin_out_sk: 'https://proxy-pub.sankuai.com',
  }
};

export const PROXY_PATH: string = 'sc-proxy';

export const LOGIN_COUNTER: number = 5;

export const IFRAME_ID: string = '__sc_login_iframe__';

export const LOGIN_TIMEOUT: number = 10000; // 一次login最长等待时间

export const LOGIN_THROTTLE_TIME: Number = 1000; // login iframe 限流时长，用于多个组件同时login的场景
