// @ss/mtd-vue/es/components/radio-button/radio-button.js
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _Symbol from 'babel-runtime/core-js/symbol';
import { CONFIG_PROVIDER, getPrefixCls, getSize } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('label', { class: _vm.wrapperCls, on: { "click": function click($event) {
        $event.preventDefault();return _vm.handleClick($event);
      } } }, [_c('input', { staticStyle: { "display": "none" }, attrs: { "type": "radio", "name": _vm.name }, domProps: { "checked": _vm.checked } }), _c('span', { class: _vm.prefix + "-inner" }, [_vm._t("default", [_vm._v(_vm._s(_vm.label))])], 2)]);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdDumbRadioButton',
  props: {
    value: {
      type: [String, Number, Boolean, Function, Object, Array, _Symbol]
    },
    disabled: {
      type: Boolean,
      default: false
    },
    name: {
      type: String
    },
    checked: {
      type: Boolean,
      default: false
    },
    size: String,
    label: [String, Number]
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getSize: getSize
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('radio-button');
    },
    _size: function _size() {
      return this.config.getSize(this);
    },
    wrapperCls: function wrapperCls() {
      var _ref;

      var prefix = this.prefix;

      return [prefix, (_ref = {}, _defineProperty(_ref, prefix + '-checked', this.checked), _defineProperty(_ref, prefix + '-disabled', this.disabled), _defineProperty(_ref, prefix + '-' + this._size, this._size), _ref)];
    }
  },
  methods: {
    handleClick: function handleClick($event) {
      if (!this.disabled && !this.checked) {
        this.$emit('input', true);
      }
      if (!this.disabled) {
        this.$emit('click', $event);
      }
    }
  }
};
