// @ss/mtd-vue/es/components/date-picker/panel/Date/date-panel-mixin.js

import { initTimeDate } from '@ss/mtd-vue/es/utils/date';

export default {
  props: {
    showTime: {
      type: Boolean,
      default: false
    },
    format: {
      type: String
    },
    selectionMode: {
      type: String,
      validator: function validator(value) {
        return ['year', 'month', 'date', 'time', 'quarter'].indexOf(value) > -1;
      },

      default: 'date'
    },
    shortcuts: {
      type: Array,
      default: function _default() {
        return [];
      }
    },
    disabledDate: {
      type: Function,
      default: function _default() {
        return false;
      }
    },
    value: {
      type: Array,
      default: function _default() {
        return [initTimeDate(), initTimeDate()];
      }
    },
    timePickerOptions: {
      default: function _default() {
        return {};
      },
      type: Object
    },
    showWeekNumbers: {
      type: Boolean,
      default: false
    },
    startDate: {
      type: Date
    },
    pickerType: {
      type: String,
      required: true
    },
    focusedDate: {
      type: Date,
      required: true
    },
    visible: Boolean,
    weekStart: {
      type: Number
    }
  },
  methods: {
    handleToggleTime: function handleToggleTime() {
      this.currentView = this.currentView === 'time' ? 'date' : 'time';
    }
  }
};
