// @ss/components-manage-cli/dist/lib/prompt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.componetnTechstackType = exports.templateType = exports.sourceType = exports.developNameType = exports.componentType = exports.nmpNameType = exports.descriptionType = exports.prefixType = exports.componentNameType = exports.techstackType = exports.platformType = exports.teamType = void 0;
exports.teamType = {
    type: 'list',
    message: 'Please pick an team.',
    name: 'teamName',
    choices: []
};
exports.platformType = {
    type: 'list',
    message: 'Please pick an platform.',
    name: 'platform',
    choices: []
};
exports.techstackType = {
    type: 'list',
    message: 'Please pick an techstack.',
    name: 'techstack',
    choices: []
};
exports.componentNameType = {
    type: 'input',
    message: 'Please input an componentName.',
    name: 'componentName',
    default: ''
};
exports.prefixType = [{
        type: 'input',
        message: 'Please input an classPrefix.',
        name: 'prefix',
        default: ''
    }];
exports.descriptionType = [{
        type: 'input',
        message: 'Please input description.',
        name: 'description',
        default: ''
    }];
exports.nmpNameType = [{
        type: 'input',
        message: 'Please input npmName.',
        name: 'npmName',
        default: ''
    }];
exports.componentType = [{
        type: 'list',
        message: 'Please pick an componentType.',
        name: 'type',
        choices: [
            "Component Lib",
            "Template Lib"
        ]
    }];
exports.developNameType = {
    type: 'list',
    message: 'Please pick an develop language.',
    name: 'developName',
    choices: []
};
exports.sourceType = [{
        type: 'input',
        message: 'Please input an source.',
        name: 'source',
        default: ''
    }];
exports.templateType = {
    type: 'list',
    message: 'Please pick an template.',
    name: 'templateName',
    choices: []
};
exports.componetnTechstackType = {
    type: 'list',
    message: 'Please pick an componetnTechstack.',
    name: 'componetnTechstack',
    choices: []
};
