// @ss/mtd-vue/components/select/index.js
import Select from './select.vue';
import { withFormItem } from '@components/form-item';

const MtdSelect = withFormItem('MtdSelect', Select, {
  withRef: true,
  methods: {
    setSearchValue (searchValue) {
      const { wrappedInstance } = this.$refs;
      wrappedInstance.setSearchValue(searchValue);
    },
    // 由于暂时不支持 visible 属性，先提供手动展开方法
    // https://ones.sankuai.com/ones/product/4348/workItem/requirement/detail/3278600
    toggle (visible) {
      const { wrappedInstance } = this.$refs;
      const nextVisible = visible !== undefined ? visible : !wrappedInstance.focused;
      nextVisible ? this.focus() : this.blur();
    },
    focus () {
      const { wrappedInstance } = this.$refs;
      wrappedInstance.focus();
    },
    blur () {
      const { wrappedInstance } = this.$refs;
      wrappedInstance.blur();
    },
  },
  key (props) {
    return props.multiple ? 'multiple' : '';
  },
});

/* istanbul ignore next */
MtdSelect.install = function (Vue) {
  Vue.component(MtdSelect.name, MtdSelect);
};

export default MtdSelect;
