// @ss/components-manage-cli/src/lib/init.ts
import { templateType, componentType, componentNameType, teamType, prefixType, descriptionType, nmpNameType, platformType, techstackType, developNameType, sourceType } from './prompt';
import { typeMap, developMap, techstackMap, platformMap, platformTextMap, techstackTextMap, developTextMap, materialTypeMap, envHost } from '../util/constant';
import { registerApi, checkLogin, fatalLog } from '@ss/components-manage-tools';

const path = require('path');
const fs = require('fs')
const inquirer = require('inquirer');
const chalk = require('chalk');
const { exec } = require('child_process');
const fetch = require('node-fetch');

const checkProjectName = (dir: string) => {
    const isExists = fs.existsSync(dir);
    if (isExists) {
        fatalLog('project is alreday exits.');
    }
}

export async function init() {
    console.log(chalk.green('\n start init project \n'));
    const registerParams: any = {};
    const token = checkLogin();
    console.log(token)
    const name = await getUserInfo(token);
    console.log(name)
    const teamList = await getTeamList(name);
    if (teamList.length === 0) {
        fatalLog('未获取到你的团队，请先去component平台创建团队。');
    }
    const teamNameList: any = [];
    teamList.map((item: any) => {
        if (item.name) {
            teamNameList.push(item.name)
        }
    });
    const { teamName } = await inquirer.prompt([{...teamType, choices: teamNameList}]);
    let teamId;
    teamList.map((item: any) => {
        if (item.name === teamName) {
            teamId = item.id;
        }
    });
    registerParams['teamId'] = teamId;
    // const { source } = await inquirer.prompt(sourceType);
    // if (!source) {
    //     fatalLog('请填写物料库来源。');
    // }
    // registerParams['source'] = source;
    const { type } = await inquirer.prompt(componentType);
    const { componentName } = await inquirer.prompt([{...componentNameType, message: typeMap[type] === 1 ? 'Please input an componentLibName.' : 'Please input an templateLibName.'}]);
    if (!componentName) {
        fatalLog('组件库/模板库名字不能为空。');
    }
    if (componentName.length <= 1) {
        fatalLog('组件库/模板库名字至少输入2个字符。');
    }
    if (!/^[a-zA-Z0-9_\\-]+[a-zA-Z0-9]$/.test(componentName)) {
        fatalLog('组件库/模板库名字只能输入英文、数字、中划线、下划线。');
    }
    registerParams['name'] = componentName;
    if (typeMap[type] === 1) {
        const { npmName } = await inquirer.prompt(nmpNameType);
        if (!npmName) {
            fatalLog('npmName不能为空。');
        }
        registerParams['npmName'] = npmName;
    }
    const pickResult = await pick(typeMap[type]);
    Object.keys(pickResult).map((item: any) => {
        registerParams[item] = pickResult[item];
    });
    if (typeMap[type] === 1) {
        const { prefix } = await inquirer.prompt(prefixType);
        if (!prefix) {
            fatalLog('样式前缀不能为空。');
        }
        if (prefix.length <= 1) {
            fatalLog('样式前缀长度至少为2个字符。');
        }
        if (!/^[a-zA-Z_\\-]+[a-zA-Z]$/.test(prefix)) {
            fatalLog('样式前缀只能为由英文、中划线、下划线。');
        }
        registerParams['classPrefix'] = prefix;
        registerParams['gitUrl'] = '';
        registerParams['importDoc'] = '';
        registerParams['installDoc'] = '';
    }
    const { description } = await inquirer.prompt(descriptionType);
    registerParams['description'] = description;
    registerParams['componentOrigin'] = 'create';
    registerParams['administrator'] = name;
    const { gitUrl, libId, libName } = await registerApi(registerParams, materialTypeMap[typeMap[type]]);
    if (!gitUrl) {
        fatalLog(`create project failed: url不存在.`);
    }
    createApi(gitUrl, libId, materialTypeMap[typeMap[type]], libName);
}

const pick = async (type: any) => {
    const pickData: any = {};
    const templateList = await getTemplateList(type);
    if (templateList.length === 0) {
        fatalLog('未获取到可用模板列表。');
    }
    const platformList: any = [];
    templateList.map((item: any) => {
        if (item.platform && platformList.indexOf(platformTextMap[item.platform]) === -1) {
            platformList.push(platformTextMap[item.platform]);
        }
    });
    const { platform } = await inquirer.prompt([{...platformType, choices: platformList}]);
    pickData['platform'] = platform;
    const techstackList: any = [];
    templateList.map((item: any) => {
        if (item.tech_stack && techstackList.indexOf(techstackTextMap[item.tech_stack]) === -1 && item.platform === platformMap[platform]) {
            techstackList.push(techstackTextMap[item.tech_stack]);
        }
    });
    const { techstack } = await inquirer.prompt([{...techstackType, choices: techstackList}]);
    pickData['framework'] = techstack;
    const developLangList: any = [];
    templateList.map((item: any) => {
        if (item.develop_lang && developLangList.indexOf(developTextMap[item.develop_lang]) === -1 && item.platform === platformMap[platform] && item.tech_stack === techstackMap[techstack]) {
            developLangList.push(developTextMap[item.develop_lang]);
        }
    });
    const { developName } = await inquirer.prompt([{...developNameType, choices: developLangList}]);
    pickData['developLanguage'] = developName;
    const templateNameList: any = [];
    templateList.map((item: any) => {
        if (item.tpl_name && item.platform === platformMap[platform] && item.tech_stack === techstackMap[techstack] && item.develop_lang === developMap[developName]) {
            templateNameList.push(item.tpl_name);
        }
    });
    const { templateName } = await inquirer.prompt([{...templateType, choices: templateNameList}]);
    templateList.map((item: any) => {
        if (item.tpl_name === templateName) {
            pickData['templateId'] = item.id;
        }
    });
    return pickData;
}

const createApi = async (url: any, libId: any, libType: any, projectName: any) => {
    await checkProjectName(path.join(process.cwd(), projectName));
    exec(`git clone ${url}`, (err: any, stdout: any, stderr: any) => {
        if (err) {
            fatalLog(`create project failed: ${err}.`);
        }
        console.log(chalk.green('\n init project success && create project success. \n'));
        console.log(chalk.gray(`\n cd ${projectName} && yarn install. \n`));
        process.exit(0);
    });
}

const getUserInfo = async (token: any) => {
    const value = await fetch(`${envHost}/openapi/getUserName?token=${token}`, {
        method: 'GET',
        headers: {
            "Content-Type": 'application/json',
        },
    });
    const data = await value.json();
    const name = data && data.data
    return name || '';
}

const getTeamList = async (name: any) => {
    const value = await fetch(`${envHost}/openapi/getTeamListByUser?name=${name}`, {
        method: 'GET',
        headers: {
            "Content-Type": 'application/json',
        },
    });
    const data = await value.json();
    const list = data && data.data;
    return list || [];
}

const getTemplateList = async (type: any) => {
    const value = await fetch(`${envHost}/openapi/getTemplateList?type=${type}`, {
        method: 'GET',
        headers: {
            "Content-Type": 'application/json',
        },
    });
    const data = await value.json();
    const list = data && data.data;
    return list || [];
}
