// @ss/components-manage-cli/src/lib/prompt.ts
export const teamType = {
    type: 'list',
    message: 'Please pick an team.',
    name: 'teamName',
    choices: []
};

export const platformType = {
    type: 'list',
    message: 'Please pick an platform.',
    name: 'platform',
    choices: []
};

export const techstackType = {
    type: 'list',
    message: 'Please pick an techstack.',
    name: 'techstack',
    choices: []
};

export const componentNameType = {
    type: 'input',
    message: 'Please input an componentName.',
    name: 'componentName',
    default: ''
};

export const prefixType = [{
    type: 'input',
    message: 'Please input an classPrefix.',
    name: 'prefix',
    default: ''
}];

export const descriptionType = [{
    type: 'input',
    message: 'Please input description.',
    name: 'description',
    default: ''
}];

export const nmpNameType = [{
    type: 'input',
    message: 'Please input npmName.',
    name: 'npmName',
    default: ''
}];

export const componentType = [{
    type: 'list',
    message: 'Please pick an componentType.',
    name: 'type',
    choices: [
        "Component Lib",
        "Template Lib"
    ]
}];

export const developNameType = {
    type: 'list',
    message: 'Please pick an develop language.',
    name: 'developName',
    choices: []
};

export const sourceType = [{
    type: 'input',
    message: 'Please input an source.',
    name: 'source',
    default: ''
}];

export const templateType = {
    type: 'list',
    message: 'Please pick an template.',
    name: 'templateName',
    choices: []
};

export const componetnTechstackType = {
    type: 'list',
    message: 'Please pick an componetnTechstack.',
    name: 'componetnTechstack',
    choices: []
};
