// @ss/mtdv-mpa/components/mpa-city-select/types/index.ts
export interface ICustomOptionItem {
    label: string;
    value: string;
    children?: ICustomOptionItem[];
    disabled?: string;
    active?: boolean;
    expand?: boolean;
    indeterminate?: boolean;
    highlight?: boolean;
    hidden?: boolean;
    cityCode?: string;
    cityId?: number;
    firstLetter?: string;
}

export interface ICustomDataProps {
    [propName: string]: string;
}

export interface ICustomObject {
    [propName: string]: any;
}

export interface IFullDataInfo {
    cityList: ICustomOptionItem[];
    provinceList: ICustomOptionItem[];
}

export interface ISideTitleInfo {
    sideBarTitle: string;
    gridTitle: string;
    gridSubTitle: string;
}

export interface ISideBarListItem {
    multiple: boolean;
    title: string;
    list: ICustomOptionItem[];
    showExpand: boolean;
    enableSelectAll: boolean;
}

export type DataSourceType = 'regions' | 'level' | 'sequence' | 'bussiness';

export type CategoryPositionType = 'inside' | 'outside';

export type CityValueType = string | number;

export type SelectedValueItem = CityValueType[];

export type SelectedValueList = SelectedValueItem[];
