// @ss/mtd-vue/es/components/sidebar/submenu.js
import MenuItem from './menu-item';
import MtdSubmenu from '@ss/mtd-vue/es/components/submenu';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('mtd-submenu', _vm._b({ attrs: { "name": _vm.item[_vm.itemKey], "disabled": _vm.item.disabled, "tooltip": _vm.item.tooltip, "tooltip-props": _vm.tooltipProps, "enabled-tooltip": _vm.item.enabledTooltip } }, 'mtd-submenu', _vm.$attrs, false), [_vm.item.icon ? _c('i', { class: _vm.item.icon, attrs: { "slot": "icon" }, slot: "icon" }) : _vm._e(), _c('span', { attrs: { "slot": "title" }, slot: "title" }, [_vm._t("title", [_vm._v(_vm._s(_vm.item.title))], { "item": _vm.item })], 2), _vm._l(_vm.item.children, function (child) {
    return _c(child.children ? 'submenu' : 'menu-item', { key: child[_vm.itemKey], tag: "component", attrs: { "item": child, "item-key": _vm.itemKey, "tooltip-props": _vm.tooltipProps }, scopedSlots: _vm._u([{ key: "title", fn: function fn(scope) {
          return _vm.$scopedSlots.title ? [_vm._t("title", null, null, scope)] : undefined;
        } }, { key: "item", fn: function fn(scope) {
          return _vm.$scopedSlots.item ? [_vm._t("item", null, null, scope)] : undefined;
        } }, { key: "default", fn: function fn(scope) {
          return _vm.$scopedSlots.item ? [_vm._t("item", null, null, scope)] : undefined;
        } }], null, true) });
  })], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'Submenu',
  components: {
    MenuItem: MenuItem,
    MtdSubmenu: MtdSubmenu
  },
  inheritAttrs: false,
  props: {
    item: {
      type: Object,
      required: true
    },
    itemKey: String,
    tooltipProps: {
      type: Object
    }
  }
};
