// @ss/components-manage-cli/src/util/constant.ts
export const typeMap = {
    'Component Lib': 1,
    'Template Lib': 2
};

export const developMap = {
    'es': 1,
    'ts': 2
};

export const developTextMap = {
    1: 'es',
    2: 'ts'
};

export const techstackTextMap = {
    1: 'vue',
    2: 'react',
    3: 'mrn',
    4: 'miniapp',
    5: 'webcomponent',
    6: 'ios',
    7: 'android',
};

export const techstackMap = {
    'vue': 1,
    'react': 2,
    'mrn': 3,
    'miniapp': 4,
    'webcomponent': 5,
    'ios': 6,
    'android': 7
};

export const platformMap = {
    'Pc': 1,
    'Mobile': 2
};

export const platformTextMap = {
    1: 'Pc',
    2: 'Mobile'
};

export const materialTypeMap = {
    1: 'component',
    2: 'template'
};

export const envHost = process.env.DEPLOY_ENV === 'test' ? 'http://material.bi.test.sankuai.com' : 'https://component.sankuai.com';

export const SCREEN_SHOTS_PREFIX = './screenshots';
// export const SKETCH_SCREEN_SHOTS_PREFIX = './sketch-screenshots';
// export const SKETCH_INVERSE_FILENAME = 'sketch-inverse.js';
export const execBuffer = 1 << 30;
export const VERSION_INCREMENTS = [
    'prerelease:alpha',
    'prerelease:beta',
    'patch',
    'minor',
    'major',
];
export const LOCK_FILE_NAME = 'template.lock';

export const ZIP_DIR = 'zips';
