// @ss/mtd-vue/es/components/breadcrumb/breadcrumb-item.js
import { CONFIG_PROVIDER, getPrefixCls, getIconCls } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('span', { class: _vm.prefix }, [_c(_vm.tag, _vm._b({ tag: "component", class: _vm.breadcrumbPrefix + "-inner" }, 'component', _vm.$attrs, false), [_vm.icon ? _c('i', { class: [_vm.prefix + "-icon", _vm.icon] }) : _vm._e(), _vm._t("default")], 2), _vm.separatorClass || !_vm.separator ? _c('i', { class: [_vm.breadcrumbPrefix + "-separator", _vm.separatorClass || "" + _vm.getIcon('right-thick')] }) : _c('span', { class: _vm.breadcrumbPrefix + "-separator" }, [_vm._v(_vm._s(_vm.separator))])], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdBreadcrumbItem',
  inheritAttrs: false,
  props: {
    tag: {
      type: [String, Object],
      default: 'span'
    },
    icon: String
  },
  inject: {
    breadcrumb: {
      from: 'breadcrumb'
    },
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getIconCls: getIconCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('breadcrumb-item');
    },
    breadcrumbPrefix: function breadcrumbPrefix() {
      return this.config.getPrefixCls('breadcrumb');
    },
    getIcon: function getIcon() {
      return this.config.getIconCls;
    },
    separator: function separator() {
      return this.breadcrumb.separator;
    },
    separatorClass: function separatorClass() {
      return this.breadcrumb.separatorClass;
    }
  }
};
