// @ss/mtd-vue/es/components/tabs/src/tab-pane.js
import _defineProperty from 'babel-runtime/helpers/defineProperty';

import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

export default {
  name: 'MtdTabPane',

  props: {
    label: String,
    value: {
      type: [String, Number, Object, Array],
      required: true
    },
    icon: String,
    closable: Boolean,
    disabled: Boolean,
    labelClass: String,
    lazy: Boolean,
    destroyOnHide: {
      type: Boolean,
      default: false
    }
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  data: function data() {
    return {
      id: Math.random(),
      index: 0,
      active: false,
      hasContent: false,
      $tab: undefined, // will set value in tab-one component.
      mounted: !this.lazy && !this.destroyOnHide
    };
  },


  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('tab-pane');
    },
    isClosable: function isClosable() {
      return this.closable || this.$parent.closable;
    },
    paneName: function paneName() {
      return this.value || this.index;
    },
    dropActive: function dropActive() {
      return this.value === this.$parent.currentValue;
    },
    act: function act() {
      return this.active || this.dropActive;
    }
  },

  watch: {
    label: function label() {
      this.$parent.$forceUpdate();
    },
    act: function act(v) {
      if (this.destroyOnHide) {
        this.mounted = v;
      } else if (v) {
        this.mounted = v;
      }
    }
  },

  mounted: function mounted() {
    this.$parent.addTabs && this.$parent.addTabs(this, this.$parent);
  },
  updated: function updated() {
    this.$parent.$emit('tab-nav-update');
  },
  render: function render(h) {
    var act = this.act,
        mounted = this.mounted;

    this.hasContent = this.$slots.default && this.$slots.default.length;
    return h(
      'div',
      {
        'class': _defineProperty({}, this.prefix, true),
        directives: [{
          name: 'show',
          value: act
        }],

        ref: 'pane'
      },
      [(act || mounted) && this.$slots.default]
    );
  }
};
