// @ss/mtd-vue/es/components/message/component.js
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import { getPrefix, getIconPrefix } from '@ss/mtd-vue/es/utils/config';

var typeMap = {
  success: 'success-circle',
  info: 'info-circle',
  warning: 'warning-circle',
  error: 'error-circle',
  loading: 'loading'
};

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('transition', { attrs: { "name": _vm.prefix + "-fade", "appear": true } }, [_c('div', _vm._b({ directives: [{ name: "show", rawName: "v-show", value: _vm.visible, expression: "visible" }], class: ["" + _vm.prefix, _vm.type ? _vm.prefix + "-" + _vm.type : '', _vm.showClose ? _vm.prefix + "-closable" : '', _vm.className], style: _vm.positionStyle, on: { "mouseenter": _vm.clearTimer, "mouseleave": _vm.startTimer } }, 'div', _vm.attrs, false), [_vm.typeIcon ? _c('i', { class: [_vm.prefix + "-icon", _vm.typeIcon] }) : _vm._e(), _vm._t("default", [!_vm.dangerouslyUseHTMLString ? _c('div', { class: _vm.prefix + "-content" }, [_vm._v(" " + _vm._s(_vm.message) + " ")]) : _c('div', { class: _vm.prefix + "-content", domProps: { "innerHTML": _vm._s(_vm.message) } })]), _vm.showClose ? _c('i', { class: _vm.prefix + "-close " + _vm.iconPrefixCls + " " + _vm.iconPrefixCls + "-close-thick", on: { "click": _vm.close } }) : _vm._e()], 2)]);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdMessage',

  data: function data() {
    return {
      visible: true,
      message: '',
      duration: 3000,
      type: '',
      icon: '',
      className: '',
      offset: 0,
      onClose: null,
      showClose: false,
      closed: false,
      verticalProperty: 'top',
      verticalOffset: 0,
      timer: null,
      dangerouslyUseHTMLString: false,
      prefixCls: getPrefix(),
      iconPrefixCls: getIconPrefix(),
      attrs: {}
    };
  },


  computed: {
    prefix: function prefix() {
      return this.prefixCls + '-message';
    },
    typeIcon: function typeIcon() {
      var iconPrefixCls = this.iconPrefixCls;

      return this.icon || (this.type && typeMap[this.type] ? iconPrefixCls + '-' + typeMap[this.type] : '');
    },
    positionStyle: function positionStyle() {
      return _defineProperty({}, this.verticalProperty, this.verticalOffset + 'px');
    }
  },

  watch: {
    closed: function closed(newVal) {
      if (newVal) {
        this.visible = false;
        this.$el.addEventListener('transitionend', this.destroyElement);
      }
    }
  },

  created: function created() {
    this.$on('esc', this.handleEsc);
  },
  mounted: function mounted() {
    this.startTimer();
  },


  methods: {
    destroyElement: function destroyElement() {
      this.$el.removeEventListener('transitionend', this.destroyElement);
      this.$destroy(true);
      this.$el.parentNode.removeChild(this.$el);
    },
    close: function close() {
      this.closed = true;
      if (typeof this.onClose === 'function') {
        this.onClose();
      }
    },
    clearTimer: function clearTimer() {
      clearTimeout(this.timer);
    },
    startTimer: function startTimer() {
      var _this = this;

      if (this.duration > 0) {
        this.timer = setTimeout(function () {
          if (!_this.closed) {
            _this.close();
          }
        }, this.duration);
      }
    },
    handleEsc: function handleEsc(e) {
      if (!this.closed) {
        this.close();
      }
    }
  }
};
