// @datafe/mdbi-common-tools/rollup.config.js
import vue from 'rollup-plugin-vue'
import typescript from 'rollup-plugin-typescript2'
import commonjs from '@rollup/plugin-commonjs'
import resolve from '@rollup/plugin-node-resolve'
import babel from '@rollup/plugin-babel'
import scss from 'rollup-plugin-scss'
// import postcss from 'rollup-plugin-postcss'
// import cssnano from 'cssnano'

const config = {
  input: './src/index.ts',
  output: [
    {
      file: 'dist/index.esm.js',
      format: 'es',
      exports: 'named',
      globals: {
        vue: 'Vue',
      },
    },
    {
      file: 'dist/index.cjs.js',
      format: 'cjs',
      exports: 'named',
      globals: {
        vue: 'Vue',
      },
    },
  ],
  plugins: [
    vue(),
    babel({ babelHelpers: 'bundled' }),
    resolve(),
    commonjs(),
    typescript(),
    scss(),
    // postcss({
    //   plugins: [cssnano()]
    // })
  ],
  external: [
    'vue',
    'vue-draggable-resizable',
    /^@ss\/mtd/,
    /^@mtfe\/sso-web/,
    /^@mt-material\/datafe-auth-apply-dynamic/,
    /^axios/,
    // /^vue-property-decorator/,
    // /^vue-class-component/,
  ],
  onwarn: function (warning) {
    if (warning.code === 'THIS_IS_UNDEFINED') {
      return
    }
    console.error(warning.message)
  },
}

export default config
