// @ss/components-manage-tools/src/lib/updateLibInfoApi.ts
import { fatalLog } from '../util/log';
import { checkLogin } from './login';
import { envHost, getSignature } from '../util/constant';

const fetch = require('node-fetch');
const chalk = require('chalk');

export const updateLibInfoApi = async (updateParams: any, libType: any) => {
    const token = checkLogin();
    const _octo = await getSignature();
    updateParams['token'] = token;
    updateParams['libType'] = libType;
    try {
        const value = await fetch(`${envHost}/openapi/v1/develop/updateLibInfo`, {
            method: 'POST',
            body: JSON.stringify(updateParams),
            headers: {
                "Content-Type": 'application/json',
                "oceanus-auth": _octo
            },
        });
        const res = await value.json();
        console.log('res====>', res)
        if (res.code !== 0) {
            fatalLog(res.message);
        } else {
            console.log(
                chalk.green('update lib info success.'),
            );
        }
    } catch (e: any) {
        fatalLog(`update lib info failed: ${e}.`)
    }
}
