// @ss/mtd-vue/es/utils/tests/hoc/dumb.js
var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('form', { staticClass: "test-form" }, [_c('div', { staticClass: "test-form-header" }, [_vm._t("header", [_c('p', { staticClass: "default-header" }, [_vm._v("this is an default header slot")])])], 2), _c('input', { domProps: { "value": _vm.innerValue }, on: { "input": _vm.handleInnerInput } }), _c('input', { domProps: { "value": _vm.syncValue }, on: { "input": _vm.handleSyncInput } }), _vm._t("default"), _c('div', { staticClass: "test-form-footer" }, [_vm._t("footer")], 2), _c('button', { staticClass: "btn-save", on: { "click": _vm.handleSave } }, [_vm._v("保存")]), _c('button', { staticClass: "btn-cancel", on: { "click": _vm.handleCancel } }, [_vm._v("取消")])], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'TestForm',
  props: {
    defaultInnerValue: {
      type: String,
      default: ''
    },
    syncValue: String
  },
  data: function data() {
    return {
      innerValue: this.defaultInnerValue
    };
  },

  methods: {
    handleInnerInput: function handleInnerInput(v) {
      this.innerValue = v;
    },
    handleSyncInput: function handleSyncInput(v) {
      this.$emit('update:syncValue', v);
    },
    handleSave: function handleSave() {
      this.$emit('save', {
        innerValue: this.innerValue,
        syncValue: this.syncValue
      });
    },
    handleCancel: function handleCancel(e) {
      this.$emit('cancel', e);
    }
  }
};
