// @ss/mtd-vue/components/picker/control.js
import createHoc from '@/utils/hoc';
import Picker from './picker.vue';

export default createHoc({
  name: 'MtdPicker',
  model: {
    prop: 'value',
  },
  props: {
    defaultVisible: {
      type: Boolean,
      default: false,
    },
    visible: Boolean,
  },
  data () {
    return {
      vs: this.defaultVisible,
    };
  },
  computed: {
    isControlled () {
      return 'visible' in this.$options.propsData;
    },
    _visible () {
      return this.isControlled ? this.visible : this.vs;
    },
  },
  methods: {
    handleVisibleChange (v) {
      this.$emit('update:visible', v);
      if (!this.isControlled) {
        this.vs = v;
      }
    },
    updatePopper () {
      const { wrappedInstance } = this.$refs;
      wrappedInstance && wrappedInstance.updatePopper();
    },
    getPopper () {
      const { wrappedInstance } = this.$refs;
      return wrappedInstance.$refs.popper;
    },
    focus () {
      const { wrappedInstance } = this.$refs;
      wrappedInstance.focus();
    },
    blur () {
      const { wrappedInstance } = this.$refs;
      wrappedInstance.blur();
    },
  },
}, {
  withRef: true,
  mapStateToProps (context) {
    return {
      visible: context._visible,
    };
  },
  mapMethodToListener (context) {
    return {
      'update:visible': context.handleVisibleChange,
    };
  },
})(Picker);
