// @ss/mtd-vue/es/components/radio-group/radio-group.js
import _Symbol from 'babel-runtime/core-js/symbol';
import { CONFIG_PROVIDER, getPrefixCls, getSize } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: _vm.prefix }, [_vm._t("default")], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'RadioGroup',
  props: {
    value: [String, Number, Boolean, Function, Object, Array, _Symbol],
    disabled: Boolean,
    size: String,
    name: String
  },
  provide: function provide() {
    return {
      radioGroup: this
    };
  },

  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getSize: getSize
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('radio-group');
    },
    _size: function _size() {
      return this.config.getSize(this);
    }
  },
  created: function created() {
    this.$on('radioChange', this.handleChange);
  },

  methods: {
    handleChange: function handleChange(radioValue) {
      if (this.value !== radioValue) {
        this.$emit('input', radioValue);
        this.$emit('change', radioValue);
      }
    }
  }
};
