// @ss/mtd-vue/es/components/tabs/tests/demo/v-if.js
import Tabs from '@ss/mtd-vue/es/components/tabs';
import TabPane from '@ss/mtd-vue/es/components/tab-pane';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('mtd-tabs', { ref: "tabs", model: { value: _vm.activeName1, callback: function callback($v) {
        _vm.activeName1 = $v;
      }, expression: "activeName1" } }, [_c('mtd-tab-pane', { attrs: { "label": "Tab1", "value": "Tab1" } }, [_vm.activeName1 === 'Tab1' ? _c('div', [_vm._v("示例内容: "), _c('br'), _vm._v(" 明月几时有？把酒问青天。不知天上宫阙，今夕是何年。 ")]) : _vm._e()]), _c('mtd-tab-pane', { attrs: { "label": "Tab2", "value": "Tab2" } }, [_vm.activeName1 === 'Tab2' ? _c('div', [_vm._v("示例内容: "), _c('br'), _vm._v(" 我欲乘风归去，又恐琼楼玉宇，高处不胜寒。起舞弄清影，何似在人间！ ")]) : _vm._e()]), _c('mtd-tab-pane', { attrs: { "label": "Tab3", "value": "Tab3" } }, [_vm.activeName1 === 'Tab3' ? _c('div', [_vm._v("示例内容: "), _c('br'), _vm._v(" 转朱阁，低绮户，照无眠。 不应有恨，何事长向别时圆？ ")]) : _vm._e()]), _c('mtd-tab-pane', { attrs: { "label": "Tab4", "value": "Tab4", "disabled": "" } }, [_vm.activeName1 === 'Tab4' ? _c('div', [_vm._v("示例内容: "), _c('br'), _vm._v(" 人有悲欢离合，月有阴晴圆缺，此事古难全。但愿人长久，千里共婵娟。 ")]) : _vm._e()])], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  components: {
    MtdTabs: Tabs,
    MtdTabPane: TabPane
  },
  data: function data() {
    return {
      activeName1: ''
    };
  },
  mounted: function mounted() {
    var _this = this;

    setTimeout(function () {
      _this.activeName1 = 'Tab1';
    }, 300);
  },

  methods: {
    assertion: function assertion() {
      expect(this.$refs.tabs.showPanel).to.be.true();
      var content = this.$el.querySelector('.mtd-tabs-content');
      expect(content.style.display).to.be.not.eql('none');
    },
    doTest: function doTest(done) {
      var _this2 = this;

      try {
        setTimeout(function () {
          _this2.assertion();
          done();
        }, 400);
      } catch (e) {
        done(e);
      }
    }
  }
};
