// @ss/mtd-vue/es/components/radio-group/tests/demo/slot.js
import RadioGroup, { withRadioGroup } from '../../index.js';

var TestRadio = withRadioGroup('Radio', {
  name: 'DumbRadio',
  props: {
    value: [String, Number],
    disabled: Boolean,
    size: String,
    checked: Boolean,
    name: String
  },
  methods: {
    handleClick: function handleClick(v) {
      this.$emit('input', true);
    }
  },
  render: function render(h) {
    return h('div', {
      on: {
        click: this.handleClick
      }
    });
  }
});

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('radio-group', _vm._b({}, 'radio-group', _vm.$attrs, false), [_c('radio', { attrs: { "value": "1" } }), _c('radio', { attrs: { "value": "2", "name": "other" } }), _c('radio', { attrs: { "value": "3", "disabled": "" } }), _c('radio', { attrs: { "value": "4", "checked": "" } }), _c('radio', { attrs: { "value": "5", "size": "large" } })], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'SlotTest',
  components: {
    radio: TestRadio,
    RadioGroup: RadioGroup
  }
};
