// @ss/mtd-vue/es/components/pagination/options.js
import MtdSelect from '@ss/mtd-vue/es/components/select';
import MtdOption from '@ss/mtd-vue/es/components/option';
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';
import Locale from '@ss/mtd-vue/es/mixins/locale';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('mtd-select', _vm._b({ class: _vm.prefix + "-options", attrs: { "value": _vm.pageSize, "size": _vm.size }, on: { "input": _vm.handleChange } }, 'mtd-select', _vm.selectProps, false), _vm._l(_vm.pageSizeOptions, function (item) {
    return _c('mtd-option', { key: item, attrs: { "label": item + _vm.pagesizeText, "value": item } });
  }), 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'PaginationOptions',

  components: {
    MtdSelect: MtdSelect,
    MtdOption: MtdOption
  },
  mixins: [Locale],
  props: {
    pageSizeOptions: Array,
    pageSize: Number,
    size: {
      type: String
    },
    selectProps: {
      type: Object,
      default: function _default() {
        return {};
      }
    }
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('pagination');
    },
    pagesizeText: function pagesizeText() {
      return this.t('el.pagination.pagesize');
    }
  },
  methods: {
    handleChange: function handleChange(val) {
      this.$emit('change', val);
    }
  }
};
