// @ss/mtd-vue/es/components/pagination/pagination.js
import _mergeJSXProps from 'babel-helper-vue-jsx-merge-props';

import MtdPage from './pager';
import MtdOptions from './options';

import { CONFIG_PROVIDER, getPrefixCls, getSize } from '@ss/mtd-vue/es/utils/config';
import Locale from '@ss/mtd-vue/es/mixins/locale';

export default {
  name: 'MtdPagination',
  components: {
    MtdPage: MtdPage,
    MtdOptions: MtdOptions
  },
  mixins: [Locale],
  props: {
    size: {
      type: String
    },
    unborder: Boolean,
    simple: {
      type: Boolean,
      default: false
    },
    currentPage: {
      type: Number,
      default: 1
    },
    pageSize: {
      type: Number,
      default: 10
    },
    pageSizeOptions: {
      type: Array,
      default: function _default() {
        return [10, 20, 50, 100];
      }
    },
    showSizeChanger: {
      type: Boolean,
      default: false
    },
    showQuickJumper: {
      type: Boolean,
      default: false
    },
    total: {
      type: Number,
      default: 0
    },
    showTotal: {
      type: Boolean,
      default: false
    },
    pagerCount: Number,
    selectClass: String,
    selectProps: Object,
    simpleReadonly: Boolean
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getSize: getSize
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('pagination');
    },
    _size: function _size() {
      return this.config.getSize(this);
    },
    defaultPageCount: function defaultPageCount() {
      var pageCount = Math.ceil(Number(this.total) / Number(this.pageSize));
      if (!isNaN(pageCount) && pageCount > 0) {
        return pageCount;
      }
      return 1;
    },
    defaultCurrentPage: function defaultCurrentPage() {
      if (isNaN(this.currentPage) || this.currentPage < 1) {
        this.handleCurrentChange(1);
        return 1;
      } else if (this.currentPage > this.defaultPageCount) {
        return this.defaultPageCount;
      }
      return this.currentPage;
    },
    defaultPageSizeOptions: function defaultPageSizeOptions() {
      var psizeOptions = this.pageSizeOptions;
      if (this.pageSizeOptions.indexOf(this.pageSize) === -1) {
        psizeOptions.unshift(this.pageSize);
      }
      return psizeOptions;
    }
  },
  methods: {
    handleChange: function handleChange(e) {
      e.target.value = ~~e.target.value;
      this.handleCurrentChange(Number(e.target.value));
    },
    handleKeyup: function handleKeyup(e) {
      e.keyCode === 13 && this.handleCurrentChange(Number(e.target.value));
    },
    handleCurrentChange: function handleCurrentChange(val) {
      var page = ~~(val > this.defaultPageCount ? this.defaultPageCount : val > 1 ? val : 1);
      if (page !== this.currentPage) {
        this.$emit('update:currentPage', page);
        this.$emit('change', page, this.pageSize);
      }
    },
    handlePageSizeChange: function handlePageSizeChange(val) {
      var pageCount = Math.max(1, Math.ceil(Number(this.total) / Number(val)));
      this.$emit('update:pageSize', val);
      var currentPage = pageCount < this.defaultCurrentPage ? pageCount : this.defaultCurrentPage;

      if (currentPage !== this.defaultCurrentPage) {
        this.$emit('update:currentPage', currentPage);
      }
      this.$emit('change', currentPage, val);

      // (pageCount >= this.defaultCurrentPage) &&
      //   this.$emit('change', this.currentPage, val);
    }
  },
  render: function render(h) {
    var _size = this._size,
        simple = this.simple,
        prefix = this.prefix;

    var simpleClass = simple ? prefix + '-simple' : '';
    var sizeClass = _size ? prefix + '-' + _size : '';
    var goto = this.t('el.pagination.goto');
    var totalText = this.t('el.pagination.total', { total: this.total });
    return h(
      'div',
      { 'class': prefix + ' ' + simpleClass + ' ' + sizeClass },
      [h(MtdPage, _mergeJSXProps([this.$attr, {
        attrs: {
          unborder: this.unborder,
          size: _size,
          simple: simple,
          currentPage: this.defaultCurrentPage,
          pageCount: this.defaultPageCount,
          pagerCount: this.pagerCount,
          simpleReadonly: this.simpleReadonly
        },
        on: {
          'change': this.handleCurrentChange
        }
      }])), this.showSizeChanger && h(MtdOptions, {
        attrs: { size: _size,
          selectProps: this.selectProps,
          pageSizeOptions: this.defaultPageSizeOptions,
          pageSize: this.pageSize
        },
        'class': this.selectClass || '',
        on: {
          'change': this.handlePageSizeChange
        }
      }), this.showQuickJumper && h(
        'span',
        { 'class': prefix + '-jumper' },
        [h('span', [goto]), h('input', {
          'class': prefix + '-jumper-input',
          attrs: { type: 'number'
          },
          domProps: {
            'value': this.defaultCurrentPage
          },
          on: {
            'keyup': this.handleKeyup,
            'blur': this.handleChange
          }
        })]
      ), this.showTotal && h(
        'span',
        { 'class': prefix + '-total' },
        [totalText]
      )]
    );
  }
};
