// @ss/components-manage-tools/dist/util/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPublishUrlMap = exports.publishUrlMap = exports.getSignature = exports.registerParamsKeyMap = exports.registerUrlMap = exports.envHost = void 0;
exports.envHost = process.env.DEPLOY_ENV === 'test' ? 'http://material.bi.test.sankuai.com' : 'https://component.sankuai.com';
exports.registerUrlMap = {
    'component': '/openapi/v1/develop/jointLib',
    'template': '/openapi/v1/develop/jointTemplate'
};
exports.registerParamsKeyMap = {
    'component': {
        "teamId": '',
        "libName": '',
        "framework": '',
        "platform": '',
        'componentOrigin': '',
        "developLanguage": '',
        "templateId": '',
        "description": '',
        "npmName": '',
        "classPrefix": '',
        "administrator": '',
        "source": '',
        "gitUrl": ''
    },
    'template': {
        "teamId": '',
        "templateName": '',
        "framework": '',
        "platform": '',
        "developLanguage": '',
        "templateId": '',
        "description": '',
        "administrator": '',
        "source": ''
    },
};
const getSignature = async () => {
    console.log('--- get signature start---');
    const { Auth } = require('@dp/patriot-sdk');
    const _auth = new Auth('com.sankuai.bfestatic.jointcomponent');
    const signature = await _auth.getTokenSignature('com.sankuai.bfestatic.jointcomponent');
    console.log('--- get signature end ---');
    return signature;
};
exports.getSignature = getSignature;
exports.publishUrlMap = {
    'component': '/openapi/third/publish',
    'template': '/openapi/third/publish/template'
};
exports.createPublishUrlMap = {
    'component': '/openapi/publish',
    'template': '/openapi/publish/template'
};
