// @ss/components-manage-tools/dist/util/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.infoLog = exports.warningLog = exports.successLog = exports.fatalLog = exports.welcome = exports.boxenLog = void 0;
const boxen = require('boxen');
const chalk = require('chalk');
function boxenLog(message, options = {}) {
    console.log(boxen(message, Object.assign({
        padding: { left: 4, right: 4, top: 1, bottom: 1 },
        margin: 1,
        borderStyle: 'round',
        borderColor: 'green',
    }, options)));
}
exports.boxenLog = boxenLog;
function welcome() {
    boxenLog(`${chalk.green('Welcome to use mtd manage commander')}`);
}
exports.welcome = welcome;
function fatalLog(content) {
    console.info(chalk.red(`[fatal]: ${content}`));
    process.exit(-1);
}
exports.fatalLog = fatalLog;
function successLog(content) {
    console.info(chalk.green(`[success]: ${content}`));
}
exports.successLog = successLog;
function warningLog(content) {
    console.info(chalk.yellow(`[warning]: ${content}`));
}
exports.warningLog = warningLog;
function infoLog(content) {
    console.info(chalk.blueBright(`[info]: ${content}`));
}
exports.infoLog = infoLog;
