// @ss/mtdv-mpa/components/mpa-date-picker/picker/date-picker.js
import Picker from '../mpa-date-picker.vue';
import DatePickerPanel from '../panel/Date/date.vue';
import RangeDatePickerPanel from '../panel/Date/date-range.vue';

export default {
  name: 'DatePicker',
  componentName: 'MpaDatePicker',
  mixins: [Picker],
  props: {
    type: {
      validator (value) {
        return [
          'year', 'yearrange', 'halfyear', 'halfyearrange',
          'month', 'monthrange',  'date', 'daterange', 'datetime', 'datetimerange',
          'singledatetimerange', 'week', 'workday', 'weekends', 'weekrange',
          'weekday', 'quarter', 'holiday', 'holidayrange', 'quarterrange'
        ].indexOf(value) > -1;
      },
      default: 'date',
    },
  },
  components: {
    DatePickerPanel,
    RangeDatePickerPanel,
  },
  computed: {
    panel () {
      const { type } = this;
      const isRange = type.indexOf('range') > -1;
      return isRange ? 'RangeDatePickerPanel' : 'DatePickerPanel';
    },
    ownPickerProps () {
      return this.options;
    },
  },
};
