// @ss/mtd-vue/es/components/config-provider/tests/demo/base.js
import _Array$from from 'babel-runtime/core-js/array/from';
import Icon from '@ss/mtd-vue/es/components/icon';
import Button from '@ss/mtd-vue/es/components/button';
import ConfigProvider from '@ss/mtd-vue/es/components/config-provider';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('ConfigProvider', { attrs: { "prefix-cls": "xx", "icon-prefix-cls": "xxicon" } }, [_c('Button', { staticClass: "test-config-btn" }, [_c('Icon', { staticClass: "icon", attrs: { "name": "search" } }), _vm._v(" xxx ")], 1)], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  components: {
    Icon: Icon,
    Button: Button,
    ConfigProvider: ConfigProvider
  },
  methods: {
    doTest: function doTest(done) {
      var el = this.$el.querySelector('.test-config-btn');
      expect(el, 'should mounted button element').to.be.not.eql(null);
      var classes = _Array$from(el.classList);
      expect(classes).to.be.include('xx-btn');

      var icon = this.$el.querySelector('.icon');
      expect(icon, 'should has icon').to.be.not.eql(null);
      var iconClasses = _Array$from(icon.classList);
      expect(iconClasses).to.be.include('xxicon-search');
      done();
    }
  }
};
