// @ss/mtd-vue/es/components/confirm/confirm.js
import _extends from 'babel-runtime/helpers/extends';
import _Promise from 'babel-runtime/core-js/promise';
import Vue from 'vue';
import confirmVue from './component';
import { isVNode } from '@ss/mtd-vue/es/utils/vnode';
import { isPromise } from '@ss/mtd-vue/es/utils/type';
import { getConfig } from '@ss/mtd-vue/es/utils/config';

var ConfirmConstructor = Vue.extend(confirmVue);
export var instances = [];

var seed = 0;

function remove(id) {
  var index = -1;
  for (var i = 0; i < instances.length; i++) {
    if (instances[i].id === id) {
      index = i;
      break;
    }
  }
  if (index > -1) {
    instances.splice(index, 1);
  }
}

var Confirm = function Confirm(options) {
  return new _Promise(function (resolve, reject) {
    var id = 'confirm_' + seed++;

    options = options || {};
    if (typeof options === 'string') {
      options = {
        message: options
      };
    }

    var data = _extends({}, options, {
      onOk: function onOk(params) {
        var result = options.onOk && options.onOk(params);
        var promise = isPromise(result) ? result : _Promise.resolve(params);
        return promise.then(function (r) {
          remove(id);
          return r;
        }).then(resolve);
      },
      onCancel: function onCancel(params) {
        var result = options.onCancel && options.onCancel(params);
        var promise = isPromise(result) ? result : _Promise.resolve(params);
        return promise.then(function (r) {
          remove(id);
          params.$$mtd = true;
          return params;
        }).then(reject);
      }
    });
    var instance = new ConfirmConstructor({
      data: data
    });

    instance.id = id;
    if (isVNode(data.message)) {
      instance.$slots.default = [options.message];
      instance.message = null;
    }
    // see https://ones.sankuai.com/ones/product/4348/workItem/requirement/detail/41067412
    options.onCreate && options.onCreate(instance, options);
    instance.$mount();
    var _options = options,
        getPopupContainer = _options.getPopupContainer;

    var parent = (getPopupContainer || getConfig().getPopupContainer)();
    parent.appendChild(instance.$el);
    instances.push(instance);
    Vue.nextTick(function () {
      instance.visible = true;
    });
  });
};

Confirm.closeAll = function () {
  for (var i = instances.length - 1; i >= 0; i--) {
    instances[i].handleCancel();
  }
  instances.splice(0, instances.length);
};

Confirm.getInstance = function () {
  return instances.length > 0 ? instances[instances.length - 1] : null;
};

Confirm.COMPONENT = confirmVue;

export default Confirm;
export { Confirm };
