// @ss/mtd-vue/es/components/checkbox-group/checkbox-group.js
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: _vm.prefix }, [_vm._t("default")], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'CheckboxGroup',
  props: {
    value: {
      type: Array,
      default: function _default() {
        return [];
      }
    },
    disabled: Boolean,
    name: String,
    min: Number,
    max: Number
  },
  provide: function provide() {
    return {
      checkboxGroup: this
    };
  },

  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('checkbox-group');
    }
  },
  created: function created() {
    this.$on('checkboxChange', this.handleChange);
  },

  methods: {
    handleChange: function handleChange(checked, value) {
      var index = this.value.indexOf(value);
      var nextValue = [].concat(_toConsumableArray(this.value));
      if (checked) {
        nextValue.push(value);
      } else if (index !== -1) {
        nextValue.splice(index, 1);
      }
      var length = nextValue.length;
      if (checked && this.max && length > this.max || !checked && this.min && length < this.min) {
        return;
      }
      this.$emit('input', nextValue);
      this.$emit('change', nextValue);
    }
  }
};
