// @ss/mtd-vue/es/components/checkbox-group/with.js
import _Symbol from 'babel-runtime/core-js/symbol';
import createHoc from '@ss/mtd-vue/es/utils/hoc';
import { hasProps } from '@ss/mtd-vue/es/utils/vnode';

export default function withCheckboxGroup(name, Component) {
  return createHoc({
    name: name,
    props: {
      value: [String, Number, Boolean, Function, Object, _Symbol],
      trueValue: {
        type: [String, Number, Boolean, Function, Object, _Symbol],
        default: true
      },
      falseValue: {
        type: [String, Number, Boolean, Function, Object, _Symbol],
        default: false
      },
      disabled: Boolean,
      name: String,
      checked: Boolean,
      isUnderControl: {
        type: Boolean,
        default: true
      }
    },
    inject: {
      checkboxGroup: {
        default: ''
      }
    },
    computed: {
      _isUnderControl: function _isUnderControl() {
        return this.isUnderControl && this.checkboxGroup;
      },
      _checked: function _checked() {
        return this._isUnderControl ? this.checkboxGroup.value.indexOf(this.value) > -1 : hasProps(this, 'checked') ? this.checked : this.value === this.trueValue;
      },
      _disabled: function _disabled() {
        return this._isUnderControl ? this.checkboxGroup.disabled || this.disabled : this.disabled;
      },
      inputName: function inputName() {
        return this._isUnderControl ? this.checkboxGroup.name : this.name;
      },
      _size: function _size() {
        return this._isUnderControl ? this.checkboxGroup.size : this.size;
      }
    },
    methods: {
      getChangeValue: function getChangeValue(checked) {
        return checked ? this.trueValue : this.falseValue;
      }
    }
  }, {
    mapStateToProps: function mapStateToProps(context) {
      return {
        checked: context._checked,
        disabled: context._disabled,
        name: context.inputName,
        value: context.value
      };
    },
    mapMethodToListener: function mapMethodToListener(context) {
      return {
        input: function input(checked) {
          if (context._isUnderControl) {
            context.checkboxGroup.$emit('checkboxChange', checked, context.value);
          } else {
            context.$emit('input', context.getChangeValue(checked));
            context.$emit('change', context.getChangeValue(checked));
          }
        }
      };
    }
  })(Component);
}
