// @ss/mtd-vue/helper/announcement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.announcement = {
    type: 'normal',
    prefix: 'm-announcement',
    name: 'mtd-announcement',
    props: [
        {
            key: 'title',
            default: '',
            detail: 'title',
            documentation: '标题，也可通过默认 slot 传入，**必选参数**。类型：string',
            options: [],
        },
        {
            key: 'description',
            default: '',
            detail: 'description',
            documentation: '辅助性文字。也可通过 slot=description 传入。类型：string',
            options: [],
        },
        {
            key: 'type',
            default: 'warning',
            detail: 'type',
            documentation: '主题。类型：string',
            options: [
                {
                    value: 'success',
                    detail: 'success',
                    documentation: '成功',
                },
                {
                    value: 'warning',
                    detail: 'warning',
                    documentation: '警告',
                },
                {
                    value: 'error',
                    detail: 'error',
                    documentation: '错误',
                },
                {
                    value: 'info',
                    detail: 'info',
                    documentation: '普通信息',
                },
            ],
        },
        {
            key: 'closable',
            isBoolean: true,
            detail: 'closable',
            documentation: '是否可关闭。类型：boolean。默认true',
            options: [],
        },
        {
            key: 'show-icon',
            isBoolean: true,
            detail: 'show-icon',
            documentation: '是否显示图标。类型：boolean。默认false。',
            options: [],
        },
    ],
    events: [
        {
            key: 'close',
            detail: 'close',
            documentation: '关闭 announcement 时触发的事件。',
        },
    ],
    detail: '公告',
    documentation: '用于展示系统的重要信息，在页面顶部贯穿静态显示，不会自动消失，用户可点击自行关闭。',
    snippets: {
        tag: 'mtd-announcement',
        attrs: {
            title: '公告消息',
            type: 'info'
        }
    },
    path: '/#/components/announcement',
};
exports.default = exports.announcement;
