// @ss/mtd-vue/helper/badge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const badge = {
    type: 'normal',
    prefix: 'm-badge',
    name: 'mtd-badge',
    props: [
        {
            key: 'value',
            default: '',
            options: [],
            detail: '显示值',
            documentation: '值类型：string/number'
        },
        {
            key: 'max',
            default: '',
            options: [],
            detail: '最大值',
            documentation: '最大值，number类型。超过最大值会显示 \'{max}+\'，要求 value 是 Number 类型'
        },
        {
            key: 'dot',
            isBoolean: true,
            options: [],
            detail: '小圆点',
            documentation: '不显示具体的值，只显示小圆点。默认false。'
        },
        {
            key: 'hidden',
            isBoolean: true,
            options: [],
            detail: '是否隐藏',
            documentation: '是否设置隐藏。默认false'
        },
    ],
    events: [],
    detail: '徽标组件',
    documentation: '出现在图标或文字右上角的红色标记（圆点、数字或者文字），表示有新内容或者待处理的信息。',
    snippets: {
        tag: 'mtd-badge',
        attrs: {
            ':value': 'value'
        }
    },
    path: '/#/components/badge'
};
exports.default = badge;
