// @ss/mtd-vue/helper/breadcrumb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.breadcrumb = {
    type: 'normal',
    prefix: 'm-breadcrumb',
    name: 'mtd-breadcrumb',
    props: [
        {
            key: 'separator',
            default: '',
            detail: 'separator',
            documentation: '分隔符。类型：string',
            options: [],
        },
        {
            key: 'separator-class',
            default: '',
            detail: 'separator-class',
            documentation: '图标分隔符 class。类型：string',
            options: [],
        },
        {
            key: 'size',
            default: 'small',
            detail: 'size',
            documentation: '尺寸。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                },
                {
                    value: 'normal',
                    detail: 'normal',
                    documentation: '正常',
                },
                {
                    value: 'large',
                    detail: 'large',
                    documentation: '大号',
                },
            ],
        },
    ],
    events: [],
    detail: '面包屑容器组件',
    documentation: `用于展示当前页面的路径或层级，并且能够快速返回之前的任意页面。`,
    snippets: {
        tag: 'mtd-breadcrumb',
        children: [
            {
                tag: 'mtd-breadcrumb-item',
                attrs: {
                    tag: 'router-link',
                    to: '/'
                }
            }
        ]
    },
    path: '/#/components/breadcrumb',
};
exports.breadcrumbItem = {
    type: 'normal',
    prefix: 'm-breadcrumb-item',
    name: 'mtd-breadcrumb-item',
    props: [
        {
            key: 'icon',
            default: '',
            detail: 'icon',
            documentation: '文字前方的图标名。类型：string',
            options: [],
        },
        {
            key: 'tag',
            default: '',
            detail: 'tag',
            documentation: '子元素的标签名。类型：string / object',
            options: [],
        },
    ],
    events: [],
    detail: '面包屑子组件',
    documentation: `用于展示当前页面的路径或层级，并且能够快速返回之前的任意页面。`,
    snippets: {
        tag: 'mtd-breadcrumb-item',
        attrs: {
            tag: 'router-link',
            to: '/'
        }
    },
    path: '/#/components/breadcrumb',
};
