// @ss/mtd-vue/helper/checkbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkboxGroup = {
    type: 'normal',
    prefix: 'm-checkbox-group',
    name: 'mtd-checkbox-group',
    props: [
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态'
        },
        {
            key: 'value',
            options: [],
            detail: '值',
            documentation: '一般通过v-model使用。'
        },
    ],
    events: [
        {
            key: 'input',
            detail: 'change事件',
            documentation: '在选项状态发生改变时触发，返回当前状态。'
        }
    ],
    detail: '多选框组组件',
    documentation: '一组多选框的容器。',
    snippets: {
        tag: 'mtd-checkbox-group',
        attrs: {
            'v-model': 'checkboxValue'
        },
        children: [
            {
                tag: 'mtd-checkbox',
                attrs: {
                    value: '1'
                }
            }
        ]
    },
    path: '/#/components/checkbox'
};
exports.checkbox = {
    type: 'normal',
    prefix: 'm-checkbox',
    name: 'mtd-checkbox',
    props: [
        {
            key: 'value',
            default: '',
            options: [],
            detail: '值',
            documentation: '值类型：Any'
        },
        {
            key: 'size',
            default: 'large',
            options: [
                {
                    value: 'large',
                    detail: '大号',
                    documentation: '大号'
                },
            ],
            detail: 'Checkbox的大小',
            documentation: 'Checkbox的大小，不设置则是普通大小。'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '禁用状态，继承CheckboxGroup组件的disabled值，默认false。'
        },
        {
            key: 'name',
            default: '',
            options: [],
            detail: 'name',
            documentation: 'Checkbox的name，继承CheckboxGroup组件的name值。类型：String'
        },
        {
            key: 'checked',
            isBoolean: true,
            options: [],
            detail: '是否选中',
            documentation: '是否选中, 支持v-model, 如果在CheckboxGroup中将根据value值判断得出，默认false。'
        },
        {
            key: 'indeterminate',
            isBoolean: true,
            options: [],
            detail: '设置indeterminate 状态',
            documentation: '设置indeterminate状态，只负责控制样式。默认false。'
        },
    ],
    events: [
        {
            key: 'input',
            detail: 'change事件',
            documentation: '在选项状态发生改变时触发，返回当前状态。'
        }
    ],
    detail: '多选框组件',
    documentation: '多选框组件。',
    snippets: {
        tag: 'mtd-checkbox',
        attrs: {
            value: '1'
        }
    },
    path: '/#/components/checkbox'
};
