// @ss/mtd-vue/helper/collapse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collapse = {
    type: 'normal',
    prefix: 'm-collapse',
    name: 'mtd-collapse',
    props: [
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: '当前展开的面板值，可以是数组也可以是单个值，当是数组时则表示可以同时展开。类型：any[]/any',
            options: [],
        },
        {
            key: 'type',
            default: '',
            detail: 'type',
            documentation: '面板类型。类型：string',
            options: [
                {
                    value: 'sample',
                    detail: 'sample',
                    documentation: '简洁型',
                },
                {
                    value: 'area',
                    detail: 'area',
                    documentation: '面型',
                },
            ],
        },
    ],
    events: [{
            key: 'input',
            detail: 'input',
            documentation: '当切换展开面板时触发。参数：(value: any[]/any)',
        }],
    detail: '折叠面板容器组件',
    documentation: '允许用户折叠和展开一块内容区域，利用折叠面板更加有效利用空间。',
    snippets: {
        tag: 'mtd-collapse',
        attrs: {
            'v-model': 'collapseValue'
        },
        children: [
            {
                tag: 'mtd-collapse-item',
                attrs: {
                    title: '标题1',
                    value: '1'
                }
            }
        ]
    },
    path: '/#/components/collapse',
};
exports.collapseItem = {
    type: 'normal',
    prefix: 'm-collapse-item',
    name: 'mtd-collapse-item',
    props: [
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: '当前面板值，必填。类型：any',
            options: [],
        },
        {
            key: 'title',
            default: '',
            detail: 'title',
            documentation: '面板标题。类型：string',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
    ],
    events: [],
    detail: '折叠面板子组件',
    documentation: '折叠面板子组件',
    snippets: {
        tag: 'mtd-collapse-item',
        attrs: {
            title: '标题2',
            value: '2'
        }
    },
    path: '/#/components/collapse',
};
