// @ss/mtd-vue/helper/datePicker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.datePicker = {
    type: 'normal',
    prefix: 'm-date-picker',
    name: 'mtd-date-picker',
    props: [
        {
            key: 'type',
            default: 'daterange',
            detail: 'type',
            documentation: '显示类型。类型：date / daterange / datetime / datetimerange / year / month',
            options: [
                {
                    value: 'date',
                    detail: 'date',
                    documentation: '日期选择'
                },
                {
                    value: 'daterange',
                    detail: 'daterange',
                    documentation: '日期段选择'
                },
                {
                    value: 'datetime',
                    detail: 'datetime',
                    documentation: '日期时间选择'
                },
                {
                    value: 'datetimerange',
                    detail: 'datetimerange',
                    documentation: '日期时间段选择'
                },
                {
                    value: 'year',
                    detail: 'year',
                    documentation: '年选择'
                },
                {
                    value: 'month',
                    detail: 'month',
                    documentation: '月选择'
                }
            ],
        },
        {
            key: 'format',
            default: '',
            detail: 'format',
            documentation: `展示的日期格式。类型：string。默认：
        date、daterange： yyyy-MM-dd；
        datetime、datetimerange：yyyy-MM-dd HH:mm:ss；
        year：yyyy；
        month：yyyy-MM`,
            options: [],
        },
        {
            key: 'editable',
            isBoolean: true,
            detail: 'editable',
            documentation: '文本框是否可输入。类型：boolean',
            options: [],
        },
        {
            key: 'readonly',
            isBoolean: true,
            detail: 'readonly',
            documentation: '是否只读。类型：boolean',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'clearable',
            isBoolean: true,
            detail: 'clearable',
            documentation: '是否可清空。类型：boolean',
            options: [],
        },
        {
            key: 'open',
            isBoolean: true,
            detail: 'open',
            documentation: '不必须，是否显示下方面板，提供 open 属性则为受控模式，不提供则为非受控模式，受控模式下则根据 open 值来决定显示。类型：boolean',
            options: [],
        },
        {
            key: 'multiple',
            isBoolean: true,
            detail: 'multiple',
            documentation: '是否可以多选，仅在 type = date 下生效。类型：boolean',
            options: [],
        },
        {
            key: 'placeholder',
            default: '',
            detail: 'placeholder',
            documentation: '空内容时的占位内容。类型：string',
            options: [],
        },
        {
            key: 'placement',
            default: 'bottom',
            detail: 'placement',
            documentation: '弹出面板默认的方向。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '顶部',
                },
                {
                    value: 'top-start',
                    detail: 'top-start',
                    documentation: '上左',
                },
                {
                    value: 'top-end',
                    detail: 'top-end',
                    documentation: '上右',
                },
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
                {
                    value: 'bottom-start',
                    detail: 'bottom-start',
                    documentation: '底左',
                },
                {
                    value: 'bottom-end',
                    detail: 'bottom-end',
                    documentation: '底右',
                },
                {
                    value: 'left',
                    detail: 'left',
                    documentation: '左侧',
                },
                {
                    value: 'left-start',
                    detail: 'left-start',
                    documentation: '左上',
                },
                {
                    value: 'left-end',
                    detail: 'left-end',
                    documentation: '左下',
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
                {
                    value: 'right-start',
                    detail: 'right-start',
                    documentation: '右上',
                },
                {
                    value: 'right-end',
                    detail: 'right-end',
                    documentation: '右下',
                },
            ],
        },
        {
            key: 'name',
            default: '',
            detail: 'name',
            documentation: '输入框的 name 属性。类型：string',
            options: [],
        },
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: '当前选中的值，当多选或者时间段时为数组。类型：Date / Array',
            options: [],
        },
        {
            key: 'genre',
            default: 'line',
            detail: 'genre',
            documentation: '类型 (同 Input 组件)。类型：string',
            options: [
                {
                    value: 'area',
                    detail: 'area',
                    documentation: '面型',
                },
                {
                    value: 'line',
                    detail: 'line',
                    documentation: '线型',
                },
            ],
        },
    ],
    events: [
        {
            key: 'update:open',
            detail: 'update:open',
            documentation: '当面板显示隐藏时触发。参数：(open: boolean)',
        },
        {
            key: 'input',
            detail: 'input',
            documentation: '当前 value 发生改变时触发，当为 range类型 或者 multiple 时为 date[]。参数：(value: date / date[])',
        },
        {
            key: 'clear',
            detail: 'clear',
            documentation: '当点击清除按钮时触发(此时会先触发一次 value 为空的 input 事件)。参数：()',
        },
        {
            key: 'confirm',
            detail: 'confirm',
            documentation: '当点击确定按钮时触发。参数：()',
        },
    ],
    detail: '时间选择框',
    documentation: `用于选择或输入日期，时间。`,
    snippets: {
        tag: 'mtd-date-picker',
        attrs: {
            'v-model': 'dateValue',
            placeholder: '选择日期'
        }
    },
    path: '/#/components/date-picker',
};
exports.timePicker = {
    type: 'normal',
    prefix: 'm-time-picker',
    name: 'mtd-time-picker',
    props: [
        {
            key: 'type',
            default: 'time',
            detail: 'type',
            documentation: '显示类型。类型：time',
            options: [],
        },
        {
            key: 'format',
            default: 'HH:mm:ss',
            detail: 'format',
            documentation: '展示的日期格式。类型：string',
            options: [],
        },
        {
            key: 'editable',
            isBoolean: true,
            detail: 'editable',
            documentation: '文本框是否可输入。类型：boolean',
            options: [],
        },
        {
            key: 'readonly',
            isBoolean: true,
            detail: 'readonly',
            documentation: '是否只读。类型：boolean',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'clearable',
            isBoolean: true,
            detail: 'clearable',
            documentation: '是否可清空。类型：boolean',
            options: [],
        },
        {
            key: 'open',
            isBoolean: true,
            detail: 'open',
            documentation: '非必须，同 DatePicker，是否显示下方面板，支持 `.sync`。类型：boolean',
            options: [],
        },
        {
            key: 'placeholder',
            default: '',
            detail: 'placeholder',
            documentation: '空内容时的占位内容。类型：string',
            options: [],
        },
        {
            key: 'placement',
            default: 'top',
            detail: 'placement',
            documentation: '弹出面板默认的方向。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '顶部',
                },
                {
                    value: 'top-start',
                    detail: 'top-start',
                    documentation: '上左',
                },
                {
                    value: 'top-end',
                    detail: 'top-end',
                    documentation: '上右',
                },
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
                {
                    value: 'bottom-start',
                    detail: 'bottom-start',
                    documentation: '底左',
                },
                {
                    value: 'bottom-end',
                    detail: 'bottom-end',
                    documentation: '底右',
                },
                {
                    value: 'left',
                    detail: 'left',
                    documentation: '左侧',
                },
                {
                    value: 'left-start',
                    detail: 'left-start',
                    documentation: '左上',
                },
                {
                    value: 'left-end',
                    detail: 'left-end',
                    documentation: '左下',
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
                {
                    value: 'right-start',
                    detail: 'right-start',
                    documentation: '右上',
                },
                {
                    value: 'right-end',
                    detail: 'right-end',
                    documentation: '右下',
                },
            ],
        },
        {
            key: 'name',
            default: '',
            detail: 'name',
            documentation: '输入框的 name 属性。类型：string',
            options: [],
        },
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: '当前选中的值，当多选或者时间段时为数组。类型：Date、Array',
            options: [],
        },
        {
            key: 'steps',
            default: '[]',
            detail: 'steps',
            documentation: '下拉列表的时间间隔，数组的三项分别对应小时、分钟、秒。例如设置为 [1, 15] 时，分钟会显示：00、15、30、45。类型：Array',
            options: [],
        },
        {
            key: 'genre',
            default: '不设置',
            detail: 'genre',
            documentation: '类型 (同 Input 组件)。类型：string',
            options: [
                {
                    value: 'area',
                    detail: 'area',
                    documentation: '面型',
                },
                {
                    value: 'line',
                    detail: 'line',
                    documentation: '线型',
                },
            ],
        },
    ],
    events: [
        {
            key: 'update:open',
            detail: 'update:open',
            documentation: '当面板显示隐藏时触发。参数：(opened: boolean)',
        },
        {
            key: 'input',
            detail: 'input',
            documentation: '当前 value 发生改变时触发。参数：(value: date)',
        },
        {
            key: 'clear',
            detail: 'clear',
            documentation: '当点击清除按钮时触发(之后会触发一次 value 为空的 input 事件)。参数：()',
        },
        {
            key: 'confirm',
            detail: 'confirm',
            documentation: '点击确认按钮时触发。参数：()',
        },
    ],
    detail: '时间选择框',
    documentation: `用于选择或输入日期，时间。`,
    snippets: {
        tag: 'mtd-time-picker',
        attrs: {
            'v-model': 'timeValue',
            placeholder: ''
        }
    },
    path: '/#/components/date-picker',
};
