// @ss/mtd-vue/helper/dropdown.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dropdown = {
    type: 'normal',
    prefix: 'm-dropdown',
    name: 'mtd-dropdown',
    props: [
        {
            key: 'visible',
            isBoolean: true,
            detail: 'visible',
            documentation: '手动控制是否显示 DropdownMenu, 支持 v-model。类型：boolean',
            options: [],
        },
        {
            key: 'placement',
            default: 'bottom-start',
            detail: 'placement',
            documentation: '显示的位置。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '顶部',
                },
                {
                    value: 'top-start',
                    detail: 'top-start',
                    documentation: '上左',
                },
                {
                    value: 'top-end',
                    detail: 'top-end',
                    documentation: '上右',
                },
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
                {
                    value: 'bottom-start',
                    detail: 'bottom-start',
                    documentation: '底左',
                },
                {
                    value: 'bottom-end',
                    detail: 'bottom-end',
                    documentation: '底右',
                },
                {
                    value: 'left',
                    detail: 'left',
                    documentation: '左侧',
                },
                {
                    value: 'left-start',
                    detail: 'left-start',
                    documentation: '左上',
                },
                {
                    value: 'left-end',
                    detail: 'left-end',
                    documentation: '左下',
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
                {
                    value: 'right-start',
                    detail: 'right-start',
                    documentation: '右上',
                },
                {
                    value: 'right-end',
                    detail: 'right-end',
                    documentation: '右下',
                },
            ],
        },
        {
            key: 'popper-class',
            default: '',
            detail: 'popper-class',
            documentation: '为 Dropdown 的 popper 添加类名。类型：string',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'trigger',
            default: 'hover',
            detail: 'trigger',
            documentation: '触发行为。类型：string',
            options: [
                {
                    value: 'hover',
                    detail: 'hover',
                    documentation: '鼠标悬浮触发',
                },
                {
                    value: 'click',
                    detail: 'click',
                    documentation: '点击触发',
                },
                {
                    value: 'focus',
                    detail: 'focus',
                    documentation: '聚焦触发',
                }
            ],
        },
    ],
    events: [],
    detail: '下拉面板触发器容器',
    documentation: `收纳更多操作元素，由按钮、操作图标、指示器或其他控件触发。`,
    snippets: {
        tag: 'mtd-dropdown',
        children: [
            {
                tag: 'mtd-dropdown-menu',
                attrs: {
                    slot: 'dropdown'
                },
                children: [
                    {
                        tag: 'mtd-dropdown-menu-item',
                        children: [
                            {
                                text: '选项1'
                            }
                        ]
                    }
                ]
            }
        ]
    },
    path: '/#/components/dropdown-menu',
};
exports.dropdownMenu = {
    type: 'normal',
    prefix: 'm-dropdown-menu',
    name: 'mtd-dropdown-menu',
    props: [
        {
            key: 'size',
            default: 'small',
            detail: 'size',
            documentation: '大小。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                }
            ],
        },
    ],
    events: [],
    detail: '下拉面板容器',
    documentation: `收纳更多操作元素，由按钮、操作图标、指示器或其他控件触发。`,
    snippets: {
        tag: 'mtd-dropdown-menu',
        attrs: {
            slot: 'dropdown'
        },
        children: [
            {
                tag: 'mtd-dropdown-menu-item',
                children: [
                    {
                        text: '选项1'
                    }
                ]
            }
        ]
    },
    path: '/#/components/dropdown-menu',
};
exports.dropdownMenuItem = {
    type: 'normal',
    prefix: 'm-dropdown-menu-item',
    name: 'mtd-dropdown-menu-item',
    props: [
        {
            key: 'size',
            default: 'small',
            detail: 'size',
            documentation: '大小。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                }
            ],
        },
    ],
    events: [],
    detail: '下拉面板选项',
    documentation: `收纳更多操作元素，由按钮、操作图标、指示器或其他控件触发。`,
    snippets: {
        tag: 'mtd-dropdown-menu-item',
        children: [
            {
                text: '选项2'
            }
        ]
    },
    path: '/#/components/dropdown-menu',
};
