// @ss/mtd-vue/helper/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.form = {
    type: 'normal',
    prefix: 'm-form',
    name: 'mtd-form',
    props: [
        {
            key: 'model',
            default: '',
            detail: 'model',
            documentation: '表单数据对象。类型：object',
            options: [],
        },
        {
            key: 'rules',
            default: 'formRules',
            detail: 'rules',
            documentation: '表单验证规则，具体下方校验规则。',
            options: [],
        },
        {
            key: 'inline',
            isBoolean: true,
            detail: 'inline',
            documentation: '是否使用行内模式。类型：boolean',
            options: [],
        },
        {
            key: 'label-position',
            default: 'top',
            detail: 'label-position',
            documentation: '表单标签的位置。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '标签放置在上方'
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '标签放置在右侧'
                }
            ],
        },
        {
            key: 'label-width',
            default: '80',
            detail: 'label-width',
            documentation: '表单标签宽度。类型：number',
            options: [],
        },
        {
            key: 'show-message',
            isBoolean: true,
            detail: 'show-message',
            documentation: '是否显示校验错误信息。类型：boolean',
            options: [],
        },
    ],
    events: [],
    detail: '表单',
    documentation: `由输入框、选择器、单选框、多选框等控件组成，用以收集、校验、提交数据。`,
    snippets: {
        tag: 'mtd-form',
        attrs: {
            ':model': 'formData',
            ref: 'form',
            ':rules': 'formRules'
        },
        children: [
            {
                tag: 'mtd-form-item',
                attrs: {
                    label: '名称',
                    prop: ''
                }
            }
        ]
    },
    path: '/#/components/form',
};
exports.formItem = {
    type: 'normal',
    prefix: 'm-form-item',
    name: 'mtd-form-item',
    props: [
        {
            key: 'prop',
            default: '',
            detail: 'prop',
            documentation: '对应表单域 model 里的字段。类型：string',
            options: [],
        },
        {
            key: 'label',
            default: '',
            detail: 'label',
            documentation: '标签文本。类型：string',
            options: [],
        },
        {
            key: 'label-width',
            default: '',
            detail: 'label-width',
            documentation: '标签的宽度，不设置时继承 Form 中的配置。类型：number',
            options: [],
        },
        {
            key: 'label-for',
            default: '',
            detail: 'label-for',
            documentation: '指定原生的 label 标签的 for 属性。类型：string',
            options: [],
        },
        {
            key: 'rules',
            default: '',
            detail: 'rules',
            documentation: '表单验证规则。类型：object',
            options: [],
        },
        {
            key: 'required',
            isBoolean: true,
            detail: 'required',
            documentation: '是否必填。类型：boolean',
            options: [],
        },
        {
            key: 'error',
            default: '',
            detail: 'error',
            documentation: '表单域验证错误信息。类型：string',
            options: [],
        },
        {
            key: 'show-message',
            isBoolean: true,
            detail: 'show-message',
            documentation: '是否显示校验错误信息。类型：boolean',
            options: [],
        },
        {
            key: 'helperPlacement',
            default: 'bottom',
            detail: 'helperPlacement',
            documentation: 'helper的位置。类型：string',
            options: [
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
            ],
        },
        {
            key: 'validatePlacement',
            default: 'right',
            detail: 'validatePlacement',
            documentation: '校验信息的位置。类型：string',
            options: [
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
            ],
        },
    ],
    events: [],
    detail: '表单项',
    documentation: `由输入框、选择器、单选框、多选框等控件组成，用以收集、校验、提交数据。`,
    snippets: {
        tag: 'mtd-form-item',
        attrs: {
            label: '名称',
            prop: ''
        }
    },
    path: '/#/components/form',
};
