// @ss/mtd-vue/helper/input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.input = {
    type: 'normal',
    prefix: 'm-input',
    name: 'mtd-input',
    props: [
        {
            key: 'genre',
            default: '',
            options: [
                {
                    value: 'area',
                    detail: '面型',
                    documentation: '面型'
                },
                {
                    value: 'line',
                    detail: '线型',
                    documentation: '线型。样式类似下划线。'
                }
            ],
            detail: '类型',
            documentation: '不设置时为框型。'
        },
        {
            key: 'size',
            default: '',
            options: [
                {
                    value: 'large',
                    detail: '大号',
                    documentation: '大号'
                },
                {
                    value: 'normal',
                    detail: '正常',
                    documentation: '正常'
                },
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                }
            ],
            detail: '控制输入框大小',
            documentation: '控制输入框大小，默认normal。'
        },
        {
            key: 'clearable',
            isBoolean: true,
            options: [],
            detail: '是否有清空按钮',
            documentation: '是否有清空按钮，默认false。'
        },
        {
            key: 'invalid',
            isBoolean: true,
            options: [],
            detail: '是否是错误状态',
            documentation: '是否是错误状态，默认false'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态，默认false。'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态，默认false。'
        },
        {
            key: 'readonly',
            options: [],
            detail: '是否只读',
            documentation: '是否只读，默认false。'
        },
        {
            key: 'value',
            options: [],
            detail: '输入框的内容',
            documentation: '输入框的内容，一般通过v-model使用。'
        },
        {
            key: 'prefix-icon',
            options: [],
            detail: '输入框头部图标',
            documentation: '输入框头部图标'
        },
        {
            key: 'suffix-icon',
            options: [],
            detail: '输入框尾部图标',
            documentation: '输入框尾部图标。'
        },
    ],
    events: [
        {
            key: 'input',
            detail: '输入事件',
            documentation: '当输入值改变时触发，回调参数：(value: string)。'
        },
        {
            key: 'clear',
            detail: '清空事件',
            documentation: '当点击清空按钮时触发。'
        }
    ],
    detail: '输入框组件',
    documentation: '指允许用户在区域内输入文本、数值等内容的组件。',
    snippets: {
        tag: 'mtd-input',
        attrs: {
            'v-model': 'inputValue',
            placeholder: '占位文案'
        }
    },
    path: '/#/components/input'
};
exports.textarea = {
    type: 'normal',
    prefix: 'm-textarea',
    name: 'mtd-textarea',
    props: [
        {
            key: 'invalid',
            isBoolean: true,
            options: [],
            detail: '是否是错误状态',
            documentation: '是否是错误状态，默认false'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态，默认false。'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态，默认false。'
        },
        {
            key: 'readonly',
            options: [],
            detail: '是否只读',
            documentation: '是否只读，默认false。'
        },
        {
            key: 'value',
            options: [],
            detail: '输入框的内容',
            documentation: '输入框的内容，一般通过v-model使用。'
        },
    ],
    events: [],
    detail: '多行输入框组件',
    documentation: '指允许用户在区域内输入文本、数值等内容的组件。',
    snippets: {
        tag: 'mtd-textarea',
        attrs: {
            'v-model': 'inputValue',
            placeholder: '占位文案',
            rows: '3',
            maxlength: '50'
        }
    },
    path: '/#/components/input'
};
