// @ss/mtd-vue/helper/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list = {
    type: 'normal',
    prefix: 'm-list',
    name: 'mtd-list',
    props: [
        {
            key: 'size',
            default: 'normal',
            detail: 'size',
            documentation: '列表大小。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                },
                {
                    value: 'normal',
                    detail: 'normal',
                    documentation: '正常',
                },
                {
                    value: 'large',
                    detail: 'large',
                    documentation: '大号',
                },
            ],
        },
        {
            key: 'loading',
            isBoolean: true,
            detail: 'loading',
            documentation: '加载中。类型：boolean',
            options: [],
        },
    ],
    events: [],
    detail: '列表',
    documentation: `多组同类、同级复杂数据信息的纵向排列，满足信息展示和操作指引等交互场景`,
    snippets: {
        tag: 'mtd-list',
        children: [
            {
                tag: 'mtd-list-item',
                attrs: {
                    'v-for': '(item, index) in listData',
                    ':key': 'index'
                }
            }
        ]
    },
    path: '/#/components/list',
};
exports.listItemGroup = {
    type: 'normal',
    prefix: 'm-list-item-group',
    name: 'mtd-list-item-group',
    props: [
        {
            key: 'header',
            default: '',
            detail: 'header',
            documentation: '分类列表名。类型：string',
            options: [],
        },
    ],
    events: [],
    detail: '列表分组容器',
    documentation: `多组同类、同级复杂数据信息的纵向排列，满足信息展示和操作指引等交互场景`,
    snippets: {
        tag: 'mtd-list-item-group',
        attrs: {
            header: '组名'
        },
        children: [
            {
                tag: 'mtd-list-item',
                attrs: {
                    'v-for': '(item, index) in listData',
                    ':key': 'index'
                }
            }
        ]
    },
    path: '/#/components/list',
};
exports.listItem = {
    type: 'normal',
    prefix: 'm-list-item',
    name: 'mtd-list-item',
    props: [],
    events: [],
    detail: '列表项',
    documentation: `多组同类、同级复杂数据信息的纵向排列，满足信息展示和操作指引等交互场景`,
    snippets: {
        tag: 'mtd-list-item'
    },
    path: '/#/components/list',
};
