// @ss/mtd-vue/helper/loading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const loading = {
    type: 'normal',
    prefix: 'm-loading',
    name: 'mtd-loading',
    props: [
        {
            key: 'size',
            default: 'small',
            detail: 'size',
            documentation: '组件大小。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号'
                },
                {
                    value: 'large',
                    detail: 'large',
                    documentation: '大号'
                }
            ],
        },
        {
            key: 'message',
            default: '',
            detail: 'message',
            documentation: '描述文案。类型：string',
            options: [],
        },
        {
            key: 'loading',
            isBoolean: true,
            detail: 'loading',
            documentation: '是否为加载中状态。类型：boolean',
            options: [],
        },
        {
            key: 'delay',
            default: '0',
            detail: 'delay',
            documentation: '延迟显示加载效果（防止闪烁）。类型：number(毫秒)',
            options: [],
        },
    ],
    events: [],
    detail: '加载/Loading',
    documentation: `向用户显示系统正在积极响应。`,
    snippets: {
        tag: 'mtd-loading'
    },
    path: '/#/components/loading',
};
exports.default = loading;
