// @ss/mtd-vue/helper/modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const modal = {
    type: 'normal',
    prefix: 'm-modal',
    name: 'mtd-modal',
    props: [
        {
            key: 'mask',
            isBoolean: true,
            detail: 'mask',
            documentation: '是否需要遮罩层。类型：boolean',
            options: [],
        },
        {
            key: 'closable',
            isBoolean: true,
            detail: 'closable',
            documentation: '是否显示关闭按钮，同时是否支持 esc 关闭。类型：boolean',
            options: [],
        },
        {
            key: 'mask-closable',
            isBoolean: true,
            detail: 'mask-closable',
            documentation: '点击蒙层是否允许关闭。类型：boolean',
            options: [],
        },
        {
            key: 'fullscreen',
            isBoolean: true,
            detail: 'fullscreen',
            documentation: '是否全屏。类型：boolean',
            options: [],
        },
        {
            key: 'visible',
            isBoolean: true,
            detail: 'visible',
            documentation: '是否显示，支持 v-model。类型：boolean',
            options: [],
        },
        {
            key: 'destroy-on-close',
            isBoolean: true,
            detail: 'destroy-on-close',
            documentation: '是否在关闭时销毁。类型：boolean',
            options: [],
        },
        {
            key: 'lock-scroll',
            isBoolean: true,
            detail: 'lock-scroll',
            documentation: '是否在 Dialog 出现时将 body 滚动锁定。类型：boolean',
            options: [],
        },
        {
            key: 'title',
            default: '',
            detail: 'title',
            documentation: 'Dialog 的标题，也可通过具名 slot （见下表）传入。类型：string',
            options: [],
        },
        {
            key: 'placement',
            default: 'top',
            detail: 'placement',
            documentation: '位置。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '顶部'
                },
                {
                    value: 'center',
                    detail: 'center',
                    documentation: '正中'
                }
            ],
        },
        {
            key: 'width',
            default: '',
            detail: 'width',
            documentation: '宽度 (当为 string 时需要带上单位)。类型：string、number',
            options: [],
        },
    ],
    events: [
        {
            key: 'input',
            detail: 'input事件',
            documentation: 'visible 发生改变时触发。参数：function (visible: boolean)'
        }
    ],
    detail: '对话框',
    documentation: `在当前页面基础上，通过浮层承载相关信息和操作，暂停当前流程，但保留特定的上下文信息。`,
    snippets: {
        tag: 'mtd-modal',
        attrs: {
            title: '标题',
            'v-model': 'modalVisible'
        }
    },
    path: '/#/components/modal',
};
exports.default = modal;
