// @ss/mtd-vue/helper/notification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.message = {
    type: 'function',
    prefix: 'mtdmessage',
    name: 'message',
    detail: '$mtd.message',
    documentation: '打开message弹窗。',
    path: '/#/components/notification',
    snippets: {
        funcName: 'message',
        params: [
            {
                type: 'string',
                key: 'message',
                value: ''
            },
            {
                type: 'string',
                key: 'type',
                value: 'info'
            },
        ]
    }
};
exports.notify = {
    type: 'function',
    prefix: 'mtdnotify',
    name: 'notify',
    detail: '$mtd.notify',
    documentation: '打开notify弹窗。',
    path: '/#/components/notification',
    snippets: {
        funcName: 'notify',
        params: [
            {
                type: 'string',
                key: 'message',
                value: ''
            },
            {
                type: 'string',
                key: 'title',
                value: '通知标题'
            },
        ]
    }
};
