// @ss/mtd-vue/helper/pagination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pagination = {
    type: 'normal',
    prefix: 'm-pagination',
    name: 'mtd-pagination',
    props: [
        {
            key: 'size',
            default: 'small',
            options: [
                {
                    value: 'small',
                    detail: '小尺寸',
                    documentation: '小尺寸'
                }
            ],
            detail: '分页控件大小',
            documentation: '分页控件大小'
        },
        {
            key: 'unborder',
            isBoolean: true,
            detail: '简洁模式',
            documentation: '简洁模式只有数字不带边框，默认false'
        },
        {
            key: 'simple',
            isBoolean: true,
            detail: '简单分页',
            documentation: '简单分页'
        },
        {
            key: 'current-page',
            options: [],
            detail: '当前页数',
            documentation: '当前页数，支持 .sync 修饰符，默认1'
        },
        {
            key: 'page-size',
            options: [],
            detail: '每页显示条数',
            documentation: '每页显示条数，支持 .sync 修饰符，默认10'
        },
        {
            key: 'page-size-options',
            options: [],
            detail: '每页显示个数选择器的选项设置',
            documentation: '每页显示个数选择器的选项设置，默认[10, 20, 50, 100]'
        },
        {
            key: 'show-size-changer',
            isBoolean: true,
            detail: '是否显示可以改变 page-size 的下拉',
            documentation: '是否显示可以改变 page-size 的下拉，默认false'
        },
        {
            key: 'total',
            options: [],
            detail: '总条目数',
            documentation: '总条目数，默认0'
        },
        {
            key: 'show-total',
            isBoolean: true,
            detail: '是否显示总条目数',
            documentation: '是否显示总条目数，默认false'
        },
        {
            key: 'pager-count',
            options: [],
            detail: '页码按钮的数量',
            documentation: '页码按钮的数量，当总页数超过该值时会折叠。值必须满足大于等于 5 且小于等于 21 的奇数，默认7'
        },
        {
            key: 'show-quick-jumper',
            isBoolean: true,
            detail: '是否显示快速跳转至某页功能',
            documentation: '是否显示快速跳转至某页功能，默认false'
        },
        {
            key: 'select-class',
            detail: '每页显示个数选择器的下拉框类名',
            documentation: '每页显示个数选择器的下拉框类名'
        }
    ],
    events: [
        {
            key: 'change',
            detail: '当前页码或者每页条数改变时会触发',
            documentation: '当前页码或者每页条数改变时会触发，回调参数(current, size)'
        }
    ],
    detail: '分页组件',
    documentation: '用于当数据量大时，分开页面展示和浏览数据。',
    snippets: {
        tag: 'mtd-pagination',
        attrs: {
            ':total': 'total',
            ':current-page.sync': 'currentPage'
        },
    },
    path: '/#/components/pagination'
};
exports.default = pagination;
