// @ss/mtd-vue/helper/popover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const popover = {
    type: 'normal',
    prefix: 'm-popover',
    name: 'mtd-popover',
    props: [
        {
            key: 'show-arrow',
            isBoolean: true,
            detail: 'show-arrow',
            documentation: '是否显示箭头。类型：boolean',
            options: [],
        },
        {
            key: 'size',
            default: 'samll',
            detail: 'size',
            documentation: '大小。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                },
                {
                    value: 'normal',
                    detail: 'normal',
                    documentation: '正常',
                },
            ],
        },
        {
            key: 'default-visible',
            isBoolean: true,
            detail: 'default-visible',
            documentation: '默认是否显示。类型：boolean',
            options: [],
        },
        {
            key: 'visible',
            isBoolean: true,
            detail: 'visible',
            documentation: '手动控制是否显示 popover, 支持 v-model。类型：boolean',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'transition',
            default: '',
            detail: 'transition',
            documentation: '定义渐变动画。类型：string',
            options: [],
        },
        {
            key: 'placement',
            default: 'top',
            detail: 'placement',
            documentation: '显示的位置。类型：string',
            options: [
                {
                    value: 'top',
                    detail: 'top',
                    documentation: '顶部',
                },
                {
                    value: 'top-start',
                    detail: 'top-start',
                    documentation: '上左',
                },
                {
                    value: 'top-end',
                    detail: 'top-end',
                    documentation: '上右',
                },
                {
                    value: 'bottom',
                    detail: 'bottom',
                    documentation: '底部',
                },
                {
                    value: 'bottom-start',
                    detail: 'bottom-start',
                    documentation: '底左',
                },
                {
                    value: 'bottom-end',
                    detail: 'bottom-end',
                    documentation: '底右',
                },
                {
                    value: 'left',
                    detail: 'left',
                    documentation: '左侧',
                },
                {
                    value: 'left-start',
                    detail: 'left-start',
                    documentation: '左上',
                },
                {
                    value: 'left-end',
                    detail: 'left-end',
                    documentation: '左下',
                },
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '右侧',
                },
                {
                    value: 'right-start',
                    detail: 'right-start',
                    documentation: '右上',
                },
                {
                    value: 'right-end',
                    detail: 'right-end',
                    documentation: '右下',
                },
            ],
        },
        {
            key: 'trigger',
            default: 'hover',
            detail: 'trigger',
            documentation: '触发行为 (注：不支持动态更改)。类型：string',
            options: [
                {
                    value: 'hover',
                    detail: 'hover',
                    documentation: '鼠标悬浮触发',
                },
                {
                    value: 'click',
                    detail: 'click',
                    documentation: '点击触发',
                },
                {
                    value: 'focus',
                    detail: 'focus',
                    documentation: '聚焦触发',
                }
            ],
        },
        {
            key: 'popper-class',
            default: '',
            detail: 'popper-class',
            documentation: '为 Popover 的 popper 添加类名。类型：string',
            options: [],
        },
        {
            key: 'content',
            default: '',
            detail: 'content',
            documentation: '显示的内容。类型：string',
            options: [],
        },
        {
            key: 'open-delay',
            default: '0',
            detail: 'open-delay',
            documentation: 'hover 触发时出现的延迟，单位毫秒, 0 表示不延迟。类型：number',
            options: [],
        },
        {
            key: 'close-delay',
            default: '200',
            detail: 'close-delay',
            documentation: 'hover 触发时消失的延迟，单位毫秒, 0 表示不延迟。类型：number',
            options: [],
        },
    ],
    events: [],
    detail: '弹出框',
    documentation: `用于鼠标点击或 hover 时，弹出的气泡式非模态浮层，在内容较少、操作较轻时使用。`,
    snippets: {
        tag: 'mtd-popover',
        attrs: {
            trigger: 'hover'
        }
    },
    path: '/#/components/popover',
};
exports.default = popover;
