// @ss/mtd-vue/helper/progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const progress = {
    type: 'normal',
    prefix: 'm-progress',
    name: 'mtd-progress',
    props: [
        {
            key: 'percentage',
            default: '0',
            detail: 'percentage',
            documentation: '百分比。类型：number。可选值：0-100。',
        },
        {
            key: 'status',
            default: 'success',
            detail: 'status',
            documentation: '状态。类型：string',
            options: [
                {
                    value: 'success',
                    detail: 'success',
                    documentation: '成功'
                },
                {
                    value: 'error',
                    detail: 'error',
                    documentation: '失败'
                }
            ],
        },
        {
            key: 'type',
            default: 'circle',
            detail: 'type',
            documentation: '类型。类型：string',
            options: [{
                    value: 'line',
                    detail: 'line',
                    documentation: '线型'
                },
                {
                    value: 'circle',
                    detail: 'circle',
                    documentation: '圆型'
                }],
        },
        {
            key: 'color',
            default: '',
            detail: 'color',
            documentation: '进度条背景色（会覆盖 status 状态颜色）。类型：string',
            options: [],
        },
        {
            key: 'stroke-width',
            default: '',
            detail: 'stroke-width',
            documentation: '进度条大小。类型：string',
            options: [],
        },
        {
            key: 'width',
            default: '28',
            detail: 'width',
            documentation: '环形进度条画布宽度（只在 type=circle 时可用）。类型：number',
            options: [],
        },
    ],
    events: [],
    detail: '进度条',
    documentation: '向用户传达特定进程的进度，告知用户当前状态和预期。',
    snippets: {
        tag: 'mtd-progress',
        attrs: {
            ':percentage': 'percentageValue'
        }
    },
    path: '/#/components/process',
};
exports.default = progress;
