// @ss/mtd-vue/helper/radio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.radioGroup = {
    type: 'normal',
    prefix: 'm-radio-group',
    name: 'mtd-radio-group',
    props: [
        {
            key: 'value',
            default: '',
            options: [],
            detail: '当前选中的值',
            documentation: '值类型：Any'
        },
        {
            key: 'size',
            default: '',
            options: [
                {
                    value: 'large',
                    detail: '大号',
                    documentation: '大号'
                },
                {
                    value: 'normal',
                    detail: '正常',
                    documentation: '正常'
                },
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                },
            ],
            detail: '选项卡大小',
            documentation: '选项卡的按钮大小,仅限按钮样式'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '默认false。'
        },
    ],
    events: [{
            key: 'input',
            detail: 'change事件',
            documentation: '在选项状态发生改变时触发，返回当前状态。'
        }],
    detail: '单选按钮组组件',
    documentation: '一组单选组件的容器',
    snippets: {
        tag: 'mtd-radio-group',
        attrs: {
            'v-model': 'radioValue'
        },
        children: [
            {
                tag: 'mtd-radio',
                attrs: {
                    value: ''
                }
            }
        ]
    },
    path: '/#/components/radio'
};
exports.radio = {
    type: 'normal',
    prefix: 'm-radio',
    name: 'mtd-radio',
    props: [
        {
            key: 'value',
            default: '',
            options: [],
            detail: '值',
            documentation: '值类型：Any'
        },
        {
            key: 'size',
            default: 'small',
            options: [
                {
                    value: 'normal',
                    detail: '正常',
                    documentation: '正常'
                },
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                },
            ],
            detail: '选项卡大小',
            documentation: '选项卡的按钮大小,仅限按钮样式'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '默认false。'
        },
        {
            key: 'name',
            default: '',
            options: [],
            detail: 'name',
            documentation: 'radio 的 name，继承 RadioGroup 组件的 name 值。类型：String'
        },
        {
            key: 'checked',
            isBoolean: true,
            options: [],
            detail: '是否选中',
            documentation: '是否选中, 根据 group 组件的 value 计算得出，默认false。'
        },
    ],
    events: [{
            key: 'input',
            detail: 'change事件',
            documentation: '在选项状态发生改变时触发，返回当前状态。回调参数：true/false。'
        }],
    detail: '单选按钮',
    documentation: '一个单选项，一般成组出现。',
    snippets: {
        tag: 'mtd-radio',
        attrs: {
            value: ''
        }
    },
    path: '/#/components/radio'
};
exports.radioBtn = {
    type: 'normal',
    prefix: 'm-radio-button',
    name: 'mtd-radio-button',
    props: [
        {
            key: 'value',
            default: '',
            options: [],
            detail: '值',
            documentation: '值类型：Any'
        },
        {
            key: 'size',
            default: 'small',
            options: [
                {
                    value: 'normal',
                    detail: '正常',
                    documentation: '正常'
                },
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                },
            ],
            detail: '选项卡大小',
            documentation: '选项卡的按钮大小,仅限按钮样式'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '默认false。'
        },
        {
            key: 'name',
            default: '',
            options: [],
            detail: 'name',
            documentation: 'radio 的 name，继承 RadioGroup 组件的 name 值。类型：String'
        },
        {
            key: 'checked',
            isBoolean: true,
            options: [],
            detail: '是否选中',
            documentation: '是否选中, 根据 group 组件的 value 计算得出，默认false。'
        },
    ],
    events: [{
            key: 'input',
            detail: 'change事件',
            documentation: '在选项状态发生改变时触发，返回当前状态。回调参数：true/false。'
        }],
    detail: '单选按钮',
    documentation: '一个按钮样式的单选项，一般成组出现。',
    snippets: {
        tag: 'mtd-radio-button',
        attrs: {
            value: ''
        }
    },
    path: '/#/components/radio'
};
