// @ss/mtd-vue/helper/select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.select = {
    type: 'normal',
    prefix: 'm-select',
    name: 'mtd-select',
    props: [
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: 'select 当前选中值, 多选时为数组。类型：string / object / array',
            options: [],
        },
        {
            key: 'value-key',
            default: '',
            detail: 'value-key',
            documentation: '仅当 value 值为 object、object[] 时生效，作为 value 唯一标识的键名。类型：string',
            options: [],
        },
        {
            key: 'size',
            default: 'normal',
            detail: 'size',
            documentation: '选择框尺寸。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号'
                },
                {
                    value: 'normal',
                    detail: 'normal',
                    documentation: '正常'
                },
                {
                    value: 'large',
                    detail: 'large',
                    documentation: '大号'
                }
            ],
        },
        {
            key: 'icon',
            default: '',
            detail: 'icon',
            documentation: '选择框右侧的 icon。类型：string',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'clearable',
            isBoolean: true,
            detail: 'clearable',
            documentation: '是否有清空按钮。类型：boolean',
            options: [],
        },
        {
            key: 'filterable',
            isBoolean: true,
            detail: 'filterable',
            documentation: '是否可搜索。类型：boolean',
            options: [],
        },
        {
            key: 'autoClearQuery',
            isBoolean: true,
            detail: 'autoClearQuery',
            documentation: '是否在获取焦点时清空搜索条件，仅在远程搜索情况下生效。类型：boolean',
            options: [],
        },
        {
            key: 'debounce',
            default: '0',
            detail: 'debounce',
            documentation: '搜索防抖延迟。类型：number',
            options: [],
        },
        {
            key: 'multiple',
            isBoolean: true,
            detail: 'multiple',
            documentation: '是否多选。类型：boolean',
            options: [],
        },
        {
            key: 'append-to-container',
            isBoolean: true,
            detail: 'append-to-container',
            documentation: '是否将弹出框插入至 body 元素。在弹出框的定位出现问题时，可将该属性设置为 false。类型：boolean',
            options: [],
        },
        {
            key: 'name',
            default: '',
            detail: 'name',
            documentation: 'select input 的 name 属性。类型：string',
            options: [],
        },
        {
            key: 'placeholder',
            default: '请选择',
            detail: 'placeholder',
            documentation: '占位符。类型：string',
            options: [],
        },
        {
            key: 'no-match-text',
            default: '',
            detail: 'no-match-text',
            documentation: '搜索条件无匹配时显示的文字。类型：string',
            options: [],
        },
        {
            key: 'no-data-text',
            default: '',
            detail: 'no-data-text',
            documentation: '选项为空时显示的文字。类型：string',
            options: [],
        },
        {
            key: 'filter-method',
            default: '',
            detail: 'filter-method',
            documentation: '自定义搜索方法， query 为当前查询条件，value 为当前 option 的值。类型：function (query, value)',
            options: [],
        },
        {
            key: 'remote',
            isBoolean: true,
            detail: 'remote',
            documentation: '是否远程搜索。类型：boolean',
            options: [],
        },
        {
            key: 'remote-method',
            default: '',
            detail: 'remote-method',
            documentation: '远程搜索的方法。类型：function (query)',
            options: [],
        },
        {
            key: 'loading',
            isBoolean: true,
            detail: 'loading',
            documentation: '是否正在从远程获取数据。类型：boolean',
            options: [],
        },
        {
            key: 'loading-text',
            default: '搜索中',
            detail: 'loading-text',
            documentation: '搜索时的文案。类型：string',
            options: [],
        },
        {
            key: 'collapse-tags',
            isBoolean: true,
            detail: 'collapse-tags',
            documentation: '多选时是否将选中值按文字的形式展示。类型：boolean',
            options: [],
        },
        {
            key: 'popper-class',
            default: '',
            detail: 'popper-class',
            documentation: 'Select 下拉框的类名。类型：string',
            options: [],
        },
    ],
    events: [
        {
            key: 'change',
            detail: 'change',
            documentation: '选中值发生变化时触发。参数：目前的选中值,上次选中值',
        },
        {
            key: 'update:visible',
            detail: 'update:visible',
            documentation: '下拉框显示／隐藏状态变化时触发。参数：出现则为 true，隐藏则为 false',
        },
        {
            key: 'remove',
            detail: 'remove',
            documentation: '多选模式下移除 tag 时触发。参数：移除tab的key',
        },
    ],
    detail: '选择框',
    documentation: `用于使用户能够从预定义的列表中选择一个或多个选项`,
    snippets: {
        tag: 'mtd-select',
        attrs: {
            'v-model': 'selectValue'
        },
        children: [
            {
                tag: 'mtd-option',
                attrs: {
                    'v-for': 'item in selectOptions',
                    ':key': "item.value",
                    ':label': 'item.label',
                    ':value': 'item.value'
                }
            }
        ]
    },
    path: '/#/components/select',
};
exports.optionGroup = {
    type: 'normal',
    prefix: 'm-option-group',
    name: 'mtd-option-group',
    props: [
        {
            key: 'label',
            default: '',
            detail: 'label',
            documentation: '分组的组名。类型：string',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否将该分组下所有选项置为禁用。类型：boolean',
            options: [],
        },
    ],
    events: [],
    detail: '选择框选项组容器',
    documentation: `选择框选项组容器`,
    snippets: {
        tag: 'mtd-option-group',
        attrs: {
            label: '分组名'
        },
        children: [
            {
                tag: 'mtd-option',
                attrs: {
                    'v-for': 'item in selectOptions',
                    ':key': "item.value",
                    ':label': 'item.label',
                    ':value': 'item.value'
                }
            }
        ]
    },
    path: '/#/components/select',
};
exports.option = {
    type: 'normal',
    prefix: 'm-option',
    name: 'mtd-option',
    props: [
        {
            key: 'value',
            default: '',
            detail: 'value',
            documentation: '选项的值。类型：string / number / object',
            options: [],
        },
        {
            key: 'label',
            default: '',
            detail: 'label',
            documentation: '选项的展示内容，若不设置则默认使用 value.toString()。类型：string',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用该选项。类型：boolean',
            options: [],
        },
    ],
    events: [],
    detail: '选择框选项',
    documentation: `用于使用户能够从预定义的列表中选择一个或多个选项`,
    snippets: {
        tag: 'mtd-option',
        attrs: {
            ':label': 'label',
            ':value': 'value'
        }
    },
    path: '/#/components/select',
};
