// @ss/mtd-vue/helper/sideBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sidebar = {
    type: 'normal',
    prefix: 'm-sidebar',
    name: 'mtd-sidebar',
    props: [
        {
            key: 'title',
            default: '',
            detail: 'title',
            documentation: '导航名称。类型：String',
            options: [],
        },
        {
            key: 'collapse',
            isBoolean: true,
            detail: 'collapse',
            documentation: '支持.sync，是否水平折叠收起导航。类型：boolean',
            options: [],
        },
        {
            key: 'active-key',
            default: '',
            detail: 'active-key',
            documentation: '(受控)，支持.sync，激活菜单的 key。类型：string / number',
            options: [],
        },
        {
            key: 'default-active-key',
            default: '',
            detail: 'default-active-key',
            documentation: '默认激活菜单的 key。类型：string / number',
            options: [],
        },
        {
            key: 'expand-keys',
            default: '',
            detail: 'expand-keys',
            documentation: '(受控)，支持.sync， 当前打开菜单的 key 的数组。类型：Array',
            options: [],
        },
        {
            key: 'default-expand-keys',
            default: '',
            detail: 'default-expand-keys',
            documentation: '当前打开菜单的 key 的数组。类型：Array',
            options: [],
        },
        {
            key: 'accordion',
            isBoolean: true,
            detail: 'accordion',
            documentation: '是否开启手风琴模式,开启后每次至多展开一个。类型：boolean',
            options: [],
        },
        {
            key: 'data',
            default: '',
            detail: 'data',
            documentation: '导航数据(包含一级导航、二级导航)。类型：Array',
            options: [],
        },
        {
            key: 'item-key',
            default: '',
            detail: 'item-key',
            documentation: '唯一标识的属性名。类型：string',
            options: [],
        },
        {
            key: 'tooltip-props',
            default: '',
            detail: 'tooltip-props',
            documentation: 'tooltip 组件属性，详细参考 [tooltip 组件](#/components/tooltip)。类型：object',
            options: [],
        },
    ],
    events: [
        {
            key: 'collapse-change',
            detail: 'collapse-change',
            documentation: '侧栏展开收起状态发生变化时触发。参数：(collaspe: boolean)',
        },
        {
            key: 'menu-expand-change',
            detail: 'menu-expand-change',
            documentation: '菜单项展开状态发生改变时触发。参数：({ expandKeys: string[], key: string, item: data, expanded: boolean })',
        },
        {
            key: 'menu-item-click',
            detail: 'menu-item-click',
            documentation: '点击菜单项时触发。参数：({ key: string, item: data })',
        },
    ],
    detail: 'MtdSidebar 侧导航',
    documentation: `aplha 版本, 后续会更改样式及交互，基础 API 功能不会有太大改动，目前只支持二级`,
    snippets: {
        tag: 'mtd-sidebar',
        attrs: {
            ':data': 'sidebarData',
            title: '侧边栏标题'
        }
    },
    path: '/#/components/sidebar',
};
exports.default = sidebar;
