// @ss/mtd-vue/helper/slider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const slider = {
    type: 'normal',
    prefix: 'm-slider',
    name: 'mtd-slider',
    props: [
        {
            key: 'min',
            default: '0',
            detail: 'min',
            documentation: '最小值。类型：number',
            options: [],
        },
        {
            key: 'max',
            default: '100',
            detail: 'max',
            documentation: '最大值。类型：number',
            options: [],
        },
        {
            key: 'value',
            default: '0',
            detail: 'value',
            documentation: '当前值。类型：number/number[]',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'step',
            default: '',
            detail: 'step',
            documentation: '步长，如果设置步长则将根据步长计算出 steps。类型：number',
            options: [],
        },
        {
            key: 'steps',
            default: '',
            detail: 'steps',
            documentation: '离散点。类型：number[]',
            options: [],
        },
        {
            key: 'thresholds',
            default: '',
            detail: 'thresholds',
            documentation: '阈值。类型：{ values: number[], colors: color[] }',
            options: [],
        },
        {
            key: 'fixed-value',
            default: '',
            detail: 'fixed-value',
            documentation: '固定起始值。类型：number',
            options: [],
        },
        {
            key: 'format-tooltip',
            default: '',
            detail: 'format-tooltip',
            documentation: '格式化tooltip message。类型：string function(value)',
            options: [],
        },
        {
            key: 'vertical',
            isBoolean: true,
            detail: 'vertical',
            documentation: '垂直。类型：  boolean',
            options: [],
        },
        {
            key: 'size',
            default: '',
            detail: 'size',
            documentation: '大小。类型: string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                },
            ],
        },
        {
            key: 'range',
            isBoolean: true,
            detail: 'range',
            documentation: '是否为范围选择。类型: boolean',
            options: [],
        },
        {
            key: 'marks',
            default: '',
            detail: 'marks',
            documentation: '刻度列表, key 为刻度的值，value 为刻度显示的内容。类型：  { [number]: string }',
            options: [],
        },
    ],
    events: [
        {
            key: 'change',
            detail: 'change',
            documentation: '值改变时触发（使用鼠标拖曳时，只在松开鼠标后触发）。参数：改变后的值',
        },
    ],
    detail: '滑块',
    documentation: `用于在一个数值区间内进行数值选择，并展示当前值。`,
    snippets: {
        tag: 'mtd-slider',
        attrs: {
            'v-model': 'sliderValue'
        }
    },
    path: '/#/components/slider',
};
exports.default = slider;
