// @ss/mtd-vue/helper/step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.steps = {
    type: 'normal',
    prefix: 'm-steps',
    name: 'mtd-steps',
    props: [
        {
            key: 'size',
            default: 'small',
            options: [
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                }
            ],
            detail: '尺寸',
            documentation: '不填则是normal大小。'
        },
        {
            key: 'dot',
            isBoolean: true,
            options: [],
            detail: '点状步骤条',
            documentation: '点状步骤条，默认false。'
        },
        {
            key: 'space',
            default: '',
            detail: '宽度/高度',
            documentation: '每个 step 的宽度/高度，不填写则自适应。支持百分比。类型：number / string'
        },
        {
            key: 'direction',
            default: 'vertical',
            options: [
                {
                    value: 'vertical',
                    detail: '垂直',
                    documentation: '按垂直方向显示'
                },
                {
                    value: 'horizontal',
                    detail: '水平',
                    documentation: '按水平方向显示'
                }
            ],
            detail: '显示方向',
            documentation: '显示方向。默认水平方向。'
        },
        {
            key: 'active',
            options: [],
            detail: '设置当前激活步骤',
            documentation: '设置当前激活步骤。从0开始计数，默认第0步。'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态'
        },
        {
            key: 'status',
            default: '',
            options: [
                {
                    value: 'wait',
                    detail: '等待中',
                    documentation: '等待中'
                },
                {
                    value: 'process',
                    detail: '处理中',
                    documentation: '处理中'
                },
                {
                    value: 'finish',
                    detail: '已完成',
                    documentation: '已完成'
                },
                {
                    value: 'error',
                    detail: '出错',
                    documentation: '出错'
                },
                {
                    value: 'cancel',
                    detail: '取消',
                    documentation: '取消'
                }
            ],
            detail: '当前步骤的状态',
            documentation: '指定当前步骤的状态。默认process。'
        },
        {
            key: 'is-center',
            isBoolean: true,
            options: [],
            detail: '是否居中对齐',
            documentation: '是否居中对齐，默认true。'
        },
    ],
    detail: '步骤条容器组件',
    documentation: '是引导用户按照流程完成任务的分步导航条，可根据实际应用场景设定步骤，步骤不得少于2步。',
    snippets: {
        tag: 'mtd-steps',
        attrs: {
            ':active': 'activeStep'
        },
        children: [
            {
                tag: 'mtd-step',
                attrs: {
                    title: '步骤1'
                }
            }
        ]
    },
    path: '/#/components/steps'
};
exports.step = {
    type: 'normal',
    prefix: 'm-step',
    name: 'mtd-step',
    props: [
        {
            key: 'title',
            default: '',
            detail: '标题',
            documentation: '标题。类型：String'
        },
        {
            key: 'description',
            options: [],
            detail: '描述性文字',
            documentation: '描述性文字。'
        },
        {
            key: 'icon',
            options: [],
            detail: '图标',
            documentation: '传入icon的class全名来自定义icon。'
        },
        {
            key: 'status',
            default: '',
            options: [
                {
                    value: 'wait',
                    detail: '等待中',
                    documentation: '等待中'
                },
                {
                    value: 'process',
                    detail: '处理中',
                    documentation: '处理中'
                },
                {
                    value: 'finish',
                    detail: '已完成',
                    documentation: '已完成'
                },
                {
                    value: 'error',
                    detail: '出错',
                    documentation: '出错'
                },
                {
                    value: 'cancel',
                    detail: '取消',
                    documentation: '取消'
                }
            ],
            detail: '当前步骤的状态',
            documentation: '设置当前步骤的状态，不设置则根据steps来确定状态。'
        },
    ],
    events: [],
    detail: '步骤组件',
    documentation: '步骤条中单个步骤组件。',
    snippets: {
        tag: 'mtd-step',
        attrs: {
            title: '步骤2'
        }
    },
    path: '/#/components/steps'
};
