// @ss/mtd-vue/helper/switch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const $switch = {
    type: 'normal',
    prefix: 'm-switch',
    name: 'mtd-switch',
    props: [
        {
            key: 'size',
            default: 'small',
            options: [
                {
                    value: 'normal',
                    detail: '正常',
                    documentation: '正常'
                },
                {
                    value: 'small',
                    detail: '小号',
                    documentation: '小号'
                }
            ],
            detail: '尺寸',
            documentation: '尺寸。可选项：small/normal，默认normal。'
        },
        {
            key: 'disabled',
            isBoolean: true,
            options: [],
            detail: '禁用状态',
            documentation: '是否设置禁用状态'
        },
        {
            key: 'loading',
            isBoolean: true,
            options: [],
            detail: '加载状态',
            documentation: '是否处于加载中状态，默认false'
        },
        {
            key: 'actived',
            isBoolean: true,
            options: [],
            detail: '是否开启',
            documentation: '指定当前是否开启,可以使用 v-model 双向绑定数据，默认true'
        },
        {
            key: 'name',
            options: [],
            detail: 'name',
            documentation: 'switch组件的name属性'
        },
    ],
    events: [
        {
            key: 'change',
            detail: 'change事件',
            documentation: '开关变化时触发，返回当前的状态。回调参数：Function(value: Boolean)'
        }
    ],
    detail: '开关组件',
    documentation: '表达同一行为两种互斥状态之间的切换。',
    snippets: {
        tag: 'mtd-switch',
        attrs: {
            'v-model': 'switchValue'
        }
    },
    path: '/#/components/switch'
};
exports.default = $switch;
