// @ss/mtd-vue/helper/tabPane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tabPane = {
    type: 'normal',
    prefix: 'm-tab-pane',
    name: 'mtd-tab-pane',
    props: [
        {
            key: 'label',
            options: [],
            detail: '选项卡标签',
            documentation: '选项卡标签'
        },
        {
            key: 'value',
            options: [],
            detail: '选项卡值',
            documentation: '如果选中该标签，父组件v-model的值则变为此值。默认false'
        },
        {
            key: 'closable',
            isBoolean: true,
            detail: '标签是否可关闭',
            documentation: '标签是否可关闭，优先级高于父组件的同名属性。'
        },
    ],
    events: [],
    detail: '标签页组件',
    documentation: '用于展示同一信息对象的不同类别或视角，可以互相之间快速切换。',
    snippets: {
        tag: 'mtd-tab-pane',
        attrs: {
            label: '标签1',
            value: 'value1'
        },
        children: [
            {
                text: '\n'
            }
        ]
    },
    path: '/#/components/tabs'
};
exports.default = tabPane;
