// @ss/mtd-vue/helper/tabs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tabs = {
    type: 'normal',
    prefix: 'm-tabs',
    name: 'mtd-tabs',
    props: [
        {
            key: 'size',
            default: 'normal',
            options: [
                {
                    value: 'large',
                    detail: '大号标签页',
                    documentation: '大号标签页'
                },
                {
                    value: 'small',
                    detail: '小号标签页',
                    documentation: '小号标签页'
                },
                {
                    value: 'normal',
                    detail: '常规标签页',
                    documentation: '常规标签页'
                }
            ],
            detail: '标签页大小',
            documentation: '控制标签页大小'
        },
        {
            key: 'line-size',
            default: '',
            options: [],
            detail: '分割线高度',
            documentation: '设置分隔线高度'
        },
        {
            key: 'type',
            options: [
                {
                    value: 'card',
                    detail: '卡片类型',
                    documentation: '普通卡片类型'
                },
                {
                    value: 'border-card',
                    detail: '带边框卡片类型',
                    documentation: '带边框卡片类型'
                }
            ],
            detail: '风格类型',
            documentation: '风格类型'
        },
        {
            key: 'tab-position',
            default: 'top',
            options: [
                {
                    value: 'top',
                    detail: '顶部选显卡',
                    documentation: '顶部选显卡'
                },
                {
                    value: 'left',
                    detail: '左侧选显卡',
                    documentation: '左侧选显卡'
                }
            ],
            detail: '选显卡位置',
            documentation: '选显卡位置，top/left'
        },
        {
            key: 'value',
            options: [],
            detail: '当前激活值',
            documentation: '当前激活值，一般使用v-model配合使用'
        },
        {
            key: 'closable',
            isBoolean: true,
            detail: '标签是否可关闭',
            documentation: '标签是否可关闭'
        },
        {
            key: 'addable',
            isBoolean: true,
            detail: '标签是否可添加',
            documentation: '标签是否可添加'
        },
        {
            key: 'scroll-unit',
            options: [],
            detail: '滚动单位',
            documentation: '滚动单位，默认1'
        },
        {
            key: 'vertical-height-number',
            options: [],
            detail: '垂直标签页tab展示数',
            documentation: '垂直标签页tab展示数，默认1000'
        }
    ],
    events: [
        {
            key: 'tab-click',
            detail: '标签点击事件',
            documentation: '标签被选中时触发，回调参数(event, el)'
        },
        {
            key: 'tab-remove',
            detail: '标签移除事件',
            documentation: '标签被移除时触发，回调参数(event, targetValue)'
        },
        {
            key: 'tab-add',
            detail: '标签添加事件',
            documentation: '标签被添加时触发，回调参数(event)'
        },
        {
            key: 'prev-click',
            detail: '向前翻页事件',
            documentation: '向前翻页时触发，回调参数(event)'
        },
        {
            key: 'next-click',
            detail: '向后翻页事件',
            documentation: '向后翻页时触发，回调参数(event)'
        }
    ],
    detail: '标签页容器组件',
    documentation: '用于展示同一信息对象的不同类别或视角，可以互相之间快速切换。',
    snippets: {
        tag: 'mtd-tabs',
        attrs: {
            'v-model': 'activeTab'
        },
        children: [
            {
                tag: 'mtd-tab-pane',
                attrs: {
                    label: '标签1',
                    value: 'value1'
                },
                children: [
                    {
                        text: '\n'
                    }
                ]
            }
        ]
    },
    path: '/#/components/tabs'
};
exports.default = tabs;
