// @ss/mtd-vue/helper/tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tag = {
    type: 'normal',
    prefix: 'm-tag',
    name: 'mtd-tag',
    props: [
        {
            key: 'size',
            default: 'large',
            detail: 'size',
            documentation: '尺寸。类型：string',
            options: [
                {
                    value: 'small',
                    detail: 'small',
                    documentation: '小号',
                },
                {
                    value: 'normal',
                    detail: 'normal',
                    documentation: '正常',
                },
                {
                    value: 'large',
                    detail: 'large',
                    documentation: '大号',
                },
            ],
        },
        {
            key: 'theme',
            default: 'blue',
            detail: 'theme',
            documentation: '主题颜色。类型：string',
            options: [
                {
                    value: 'gray',
                    detail: 'gray',
                    documentation: '灰',
                },
                {
                    value: 'blue',
                    detail: 'blue',
                    documentation: '蓝',
                },
                {
                    value: 'green',
                    detail: 'green',
                    documentation: '绿',
                },
                {
                    value: 'orange',
                    detail: 'orange',
                    documentation: '橙',
                },
                {
                    value: 'red',
                    detail: 'red',
                    documentation: '红',
                },
                {
                    value: 'purple',
                    detail: 'purple',
                    documentation: '紫',
                },
            ],
        },
        {
            key: 'type',
            default: 'unbordered',
            detail: 'type',
            documentation: '类型。类型：string。默认不设置。',
            options: [
                {
                    value: 'unbordered',
                    detail: 'unbordered',
                    documentation: '无边框',
                },
                {
                    value: 'pure',
                    detail: 'pure',
                    documentation: '纯色背景',
                },
                {
                    value: 'text',
                    detail: 'text',
                    documentation: '文本型',
                },
            ],
        },
        {
            key: 'rounded',
            isBoolean: true,
            detail: 'rounded',
            documentation: '是否是圆形。类型：boolean',
            options: [],
        },
        {
            key: 'closable',
            isBoolean: true,
            detail: 'closable',
            documentation: '是否可以关闭。类型：boolean',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
    ],
    events: [
        {
            key: 'click',
            detail: 'click',
            documentation: '点击事件的回调。参数：Function(e: Event)',
        },
        {
            key: 'close',
            detail: 'close',
            documentation: '关闭事件的回调。参数：Function(e: Event)',
        },
    ],
    detail: '标签',
    documentation: '是小块的信息项，用来通过视觉快速区分选择的内容。',
    snippets: {
        tag: 'mtd-tag',
        attrs: {}
    },
    path: '/#/components/tag',
};
exports.default = tag;
