// @ss/mtd-vue/helper/timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timeline = {
    type: 'normal',
    prefix: 'm-timeline',
    name: 'mtd-timeline',
    props: [
        {
            key: 'mode',
            default: 'alternate',
            detail: 'mode',
            documentation: '时间轴的位置。类型：String',
            options: [
                {
                    value: 'right',
                    detail: 'right',
                    documentation: '靠右',
                },
                {
                    value: 'alternate',
                    detail: 'alternate',
                    documentation: '交替',
                },
            ],
        },
    ],
    events: [],
    detail: '时间轴',
    documentation: `时间轴`,
    snippets: {
        tag: 'mtd-timeline',
        children: [
            {
                tag: 'mtd-timeline-item',
                attrs: {
                    title: ''
                }
            }
        ]
    },
    path: '/#/components/timeline',
};
exports.timelineItem = {
    type: 'normal',
    prefix: 'm-timeline-item',
    name: 'mtd-timeline-item',
    props: [
        {
            key: 'title',
            default: '',
            detail: 'title',
            documentation: '侧标题。类型：String',
            options: [],
        },
    ],
    events: [],
    detail: '时间轴子组件',
    documentation: `时间轴子组件`,
    snippets: {
        tag: 'mtd-timeline-item',
        attrs: {
            title: ''
        }
    },
    path: '/#/components/timeline',
};
