// @ss/mtd-vue/helper/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const transfer = {
    type: 'normal',
    prefix: 'm-transfer',
    name: 'mtd-transfer',
    props: [
        {
            key: 'data',
            default: 'initData',
            detail: 'data',
            documentation: '数据源。类型：Array',
            options: [],
        },
        {
            key: 'filterable',
            isBoolean: true,
            detail: 'filterable',
            documentation: '是否显示搜索框。类型：boolean',
            options: [],
        },
        {
            key: 'filter-placeholder',
            default: '请输入字段名称',
            detail: 'filter-placeholder',
            documentation: '搜索框placeholder。类型：string',
            options: [],
        },
        {
            key: 'filter-method',
            default: '',
            detail: 'filter-method',
            documentation: '自定义搜索函数，入参为query和item，item为每一项，query为当前输入的搜索词。类型：function',
            options: [],
        },
        {
            key: 'titles',
            default: '',
            detail: 'titles',
            documentation: '标题集合顺序从左到右。类型：Array',
            options: [],
        },
        {
            key: 'operations',
            default: '',
            detail: 'operations',
            documentation: '操作文案集合，顺序从上到下。类型：Array',
            options: [],
        },
        {
            key: 'props',
            default: '',
            detail: 'props',
            documentation: '数据源字段的别名。类型：object',
            options: [],
        },
        {
            key: 'v-model',
            default: '[]',
            detail: 'v-model',
            documentation: '显示在右侧的数据的key集合。类型：Array',
            options: [],
        },
        {
            key: 'source-default-checked',
            default: '',
            detail: 'source-default-checked',
            documentation: '初始状态下左侧列表的已勾选项的 key 数组。类型：Array',
            options: [],
        },
        {
            key: 'target-default-checked',
            default: '',
            detail: 'target-default-checked',
            documentation: '初始状态下右侧列表的已勾选项的 key 数组。类型：Array',
            options: [],
        },
    ],
    events: [
        {
            key: 'change',
            detail: 'change',
            documentation: '选项在两栏之间转移时的回调函数。参数：(targetKeys, direction, moveKeys): void',
        },
        {
            key: 'selectChange',
            detail: 'selectChange',
            documentation: '左右两侧选中项发生变化时触发。参数：(selectedKeys, direction): void',
        },
    ],
    detail: '穿梭框',
    documentation: `穿梭框`,
    snippets: {
        tag: 'mtd-transfer',
        attrs: {
            ':data': 'initData',
            'v-model': 'checkedData'
        }
    },
    path: '/#/components/transfer',
};
exports.default = transfer;
