// @ss/mtd-vue/helper/upload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const upload = {
    type: 'normal',
    prefix: 'm-upload',
    name: 'mtd-upload',
    props: [
        {
            key: 'action',
            default: '',
            detail: 'action',
            documentation: '必选参数，上传的地址。类型：string',
            options: [],
        },
        {
            key: 'headers',
            default: '',
            detail: 'headers',
            documentation: '设置上传的请求头部。类型：object',
            options: [],
        },
        {
            key: 'multiple',
            isBoolean: true,
            detail: 'multiple',
            documentation: '是否支持多选文件。类型：boolean',
            options: [],
        },
        {
            key: 'data',
            default: '',
            detail: 'data',
            documentation: '上传时附带的额外参数。类型：object',
            options: [],
        },
        {
            key: 'name',
            default: 'file',
            detail: 'name',
            documentation: '上传的文件字段名。类型：string',
            options: [],
        },
        {
            key: 'with-credentials',
            isBoolean: true,
            detail: 'with-credentials',
            documentation: '支持发送 cookie 凭证信息。类型：boolean',
            options: [],
        },
        {
            key: 'show-file-list',
            isBoolean: true,
            detail: 'show-file-list',
            documentation: '是否显示已上传文件列表。类型：boolean',
            options: [],
        },
        {
            key: 'drag',
            isBoolean: true,
            detail: 'drag',
            documentation: '是否启用拖拽上传。类型：boolean',
            options: [],
        },
        {
            key: 'accept',
            default: '',
            detail: 'accept',
            documentation: '接受上传的[文件类型](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-accept)（thumbnail-mode 模式下此参数无效）。类型：string',
            options: [],
        },
        {
            key: 'on-preview',
            default: '',
            detail: 'on-preview',
            documentation: '点击文件列表中已上传的文件时的钩子。类型：function(file)',
            options: [],
        },
        {
            key: 'on-remove',
            default: '',
            detail: 'on-remove',
            documentation: '文件列表移除文件时的钩子。类型：function(file, fileList)',
            options: [],
        },
        {
            key: 'on-success',
            default: '',
            detail: 'on-success',
            documentation: '文件上传成功时的钩子。类型：function(response, file, fileList)',
            options: [],
        },
        {
            key: 'on-error',
            default: '',
            detail: 'on-error',
            documentation: '文件上传失败时的钩子。类型：function(err, file, fileList)',
            options: [],
        },
        {
            key: 'on-progress',
            default: '',
            detail: 'on-progress',
            documentation: '文件上传时的钩子。类型：function(event, file, fileList)',
            options: [],
        },
        {
            key: 'on-change',
            default: '',
            detail: 'on-change',
            documentation: '文件状态改变时的钩子，添加文件、上传成功和上传失败时都会被调用。类型：function(file, fileList)',
            options: [],
        },
        {
            key: 'before-upload',
            default: '',
            detail: 'before-upload',
            documentation: '上传文件之前的钩子，参数为上传的文件，若返回 false 或者返回 Promise 且被 reject，则停止上传。。类型：function(file)',
            options: [],
        },
        {
            key: 'before-remove',
            default: '',
            detail: 'before-remove',
            documentation: '删除文件之前的钩子，参数为上传的文件和文件列表，若返回 false 或者返回 Promise 且被 reject，则停止上传。。类型：function(file, fileList)',
            options: [],
        },
        {
            key: 'list-type',
            default: 'text',
            detail: 'list-type',
            documentation: '文件列表的类型。类型：string',
            options: [
                {
                    value: 'text',
                    detail: 'text',
                    documentation: '文本型',
                },
                {
                    value: 'picture-card',
                    detail: 'picture-card',
                    documentation: '图片卡型',
                },
            ],
        },
        {
            key: 'auto-upload',
            isBoolean: true,
            detail: 'auto-upload',
            documentation: '是否在选取文件后立即进行上传。类型：boolean',
            options: [],
        },
        {
            key: 'file-list',
            default: 'uploadFileList',
            detail: 'file-list',
            documentation: '上传的文件列表, 例如: [{ name: food.jpg, url: https://xxx.cdn.com/xxx.jpg }]。类型：array',
            options: [],
        },
        {
            key: 'http-request',
            default: '',
            detail: 'http-request',
            documentation: '覆盖默认的上传行为，可以自定义上传的实现。类型：function',
            options: [],
        },
        {
            key: 'disabled',
            isBoolean: true,
            detail: 'disabled',
            documentation: '是否禁用。类型：boolean',
            options: [],
        },
        {
            key: 'limit',
            default: '',
            detail: 'limit',
            documentation: '最大允许上传个数。类型：number',
            options: [],
        },
        {
            key: 'on-exceed',
            default: '',
            detail: 'on-exceed',
            documentation: '文件超出个数限制时的钩子。类型：function(files, fileList)',
            options: [],
        },
    ],
    events: [],
    detail: 'Upload',
    documentation: `通过选择和拖拽方式,将文件上传至服务端。`,
    snippets: {
        tag: 'mtd-upload',
        attrs: {
            'action': '',
        }
    },
    path: '/#/components/upload',
};
exports.default = upload;
