// @datafe/eslint-config-base/eslintrc.vue.js
// const rootPath = process.cwd()

module.exports = {
  root: true,
  env: {
    browser: true,
    node: true,
    es2021: true,
  },
  extends: [
    'eslint:recommended',
    'plugin:vue/recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  parser: 'vue-eslint-parser',
  parserOptions: {
    ecmaVersion: 2021,
    parser: '@typescript-eslint/parser',
    sourceType: 'module',
    // tsconfigRootDir: rootPath,
    // 会导致 jsx option 不工作，因而先注释掉。参见：https://github.com/typescript-eslint/typescript-eslint/issues/566
    // project: ["./tsconfig.json"],
    extraFileExtensions: ['.vue'],
    ecmaFeatures: {
      jsx: true,
    },
  },
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      node: {
        extensions: ['.mjs', '.js', '.json', '.ts'],
      },
      typescript: {},
    },
    'import/extensions': ['.js', '.mjs', '.jsx', '.ts', '.tsx'],
    'import/external-module-folders': ['node_modules', 'node_modules/@types'],
  },
  plugins: ['import', 'vue', '@typescript-eslint'],
  rules: {
    // 允许类方法中可以不使用 this
    'class-methods-use-this': 'off',

    // 支持导入 ts, tsx 文件时忽略后缀名
    'import/extensions': [
      'error',
      'ignorePackages',
      {
        js: 'never',
        mjs: 'never',
        jsx: 'never',
        ts: 'never',
        tsx: 'never',
      },
    ],

    // 允许 require
    '@typescript-eslint/no-var-requires': 'off',
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx', '*.vue'],
      rules: {
        'no-undef': 'off',
      },
    },
  ],
}
