// @ss/mtd-vue/es/components/tooltip/tooltip.js
import { Popper, Drop, Reference } from '@ss/mtd-vue/es/components/popper';
import { CONFIG_PROVIDER, getPrefixCls, getSize } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _obj;
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('popper', _vm._b({ ref: "popper", class: _vm.prefix + "-rel", attrs: { "tag": _vm.tag, "popper-disabled": _vm.disabled, "trigger": _vm.trigger, "visible": !_vm.disabled && _vm.visible, "show-arrow": _vm.showArrow, "open-delay": _vm.openDelay }, on: { "update:visible": _vm.handleVisibleChange, "clickoutside": _vm.clickoutside } }, 'popper', _vm.$attrs, false), [_c('reference', [_vm._t("default")], 2), _c('drop', _vm._b({ class: [_vm.prefix, (_obj = {}, _obj[_vm.prefix + "-" + _vm._size] = _vm._size, _obj[_vm.prefix + "-" + _vm.theme] = _vm.theme, _obj), _vm.popperClass] }, 'drop', _vm.popperAttrs, false), [_vm._t("content", [_vm._v(_vm._s(_vm.content))])], 2)], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'Tooltip',
  components: {
    Popper: Popper,
    Drop: Drop,
    Reference: Reference
  },
  inheritAttrs: false,
  model: {
    prop: 'visible'
  },
  props: {
    content: [String, Number],
    visible: Boolean,
    size: String,
    showArrow: {
      type: Boolean,
      default: true
    },
    trigger: {
      type: String,
      default: 'hover'
    },
    popperClass: [String, Object, Array],
    popperAttrs: Object,
    disabled: Boolean,
    theme: {
      type: String
    },
    openDelay: {
      type: Number,
      default: 300
    },
    tag: {
      type: [String, Object],
      default: 'span'
    }
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getSize: getSize
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('tooltip');
    },
    _size: function _size() {
      return this.config.getSize(this);
    }
  },
  methods: {
    clickoutside: function clickoutside(e) {
      this.$emit('clickoutside', e);
    },
    handleVisibleChange: function handleVisibleChange(v) {
      this.$emit('input', v);
    },
    updatePopper: function updatePopper() {
      var popper = this.$refs.popper;

      if (this.visible && popper) {
        popper.updatePopper();
      }
    }
  }
};
