// @mtfe/sso-web/example/talos-react/src/App.jsx
import React from 'react'
import logo from './assets/logo.svg'
import talos from './api/talos'
import ssoWeb from './utils/sso';
import axios from 'axios';
import './App.css'

class App extends React.Component {
  componentDidMount() {
    // Talos Demo Public API by @hfe/common-cli-plugin-axios
    talos.alive()
  }
  constructor(props) {
    super(props);
    this.handleLogout = this.handleLogout.bind(this);
    this.handleMock401Rep = this.handleMock401Rep.bind(this)
  }
  async handleLogout() {
    await ssoWeb.logout();
    window.location.href = ssoWeb.getLoginUrl();
  }
  async handleMock401Rep() {
    // 请求这个地址会返回
    /*
    {
      status: 401,
      data: {
        message: 'auth failed'
      }
    }
    */
    await axios.get('http://yapi.sankuai.com/mock/1232/mock401')
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Welcome to React</h1>
        </header>
        <div>
          <button onClick={this.handleLogout}>登出</button>
          <button onClick={this.handleMock401Rep}>模拟ssoid失效</button>
        </div>
      </div>
    )
  }
}

export default App
