// @ss/components-manage-tools/dist/lib/updateLibInfoApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateLibInfoApi = void 0;
const log_1 = require("../util/log");
const login_1 = require("./login");
const constant_1 = require("../util/constant");
const fetch = require('node-fetch');
const chalk = require('chalk');
const updateLibInfoApi = async (updateParams, libType) => {
    const token = (0, login_1.checkLogin)();
    const _octo = await (0, constant_1.getSignature)();
    updateParams['token'] = token;
    updateParams['libType'] = libType;
    try {
        const value = await fetch(`${constant_1.envHost}/openapi/v1/develop/updateLibInfo`, {
            method: 'POST',
            body: JSON.stringify(updateParams),
            headers: {
                "Content-Type": 'application/json',
                "oceanus-auth": _octo
            },
        });
        const res = await value.json();
        console.log('res====>', res);
        if (res.code !== 0) {
            (0, log_1.fatalLog)(res.message);
        }
        else {
            console.log(chalk.green('update lib info success.'));
        }
    }
    catch (e) {
        (0, log_1.fatalLog)(`update lib info failed: ${e}.`);
    }
};
exports.updateLibInfoApi = updateLibInfoApi;
