// @mtfe/sso-web/src/store/token.ts
/**
 * Created by leon on 6/14/18.
 */
import * as util from '../util';

interface TokenObj {
    getToken(): string;
    setToken(ssoid: string,sameSite: boolean): void;
    clearToken(): void;
    setCookieStr(name: string,value: string, expiredTime: string): void;
}

export interface TokenArgs {
    clientId: string;
}

const DEFAULT_TOKEN_CFG = {
    name: 'ssoid',
    path: '/',
    domain: '',
};

export class TokenStorage implements TokenObj {
    clientId: string;

    constructor(tokenArgs: TokenArgs) {
        this.clientId = tokenArgs.clientId;
    }

    setToken(ssoid, sameSite) {
        if (!ssoid) return void 0;

        const options={
            value: ssoid,
            name: `${this.clientId}_${DEFAULT_TOKEN_CFG.name}`,
            path: DEFAULT_TOKEN_CFG.path,
            domain: DEFAULT_TOKEN_CFG.domain,
            expires: new Date(Date.now() + 1000 * 60 * 60 * 24 * 30).toString(), // 30 days
        }
        if(!sameSite && util.getBrowserVersion('chrome')>=80){
            Object.assign(options,{
                SameSite: 'None',
                Secure: true
            })
        }
        document.cookie = util.parseCookie(options);
    }

    getToken() {
        // 添加分号防止取的key值在cookie最后
        const cookie = `${document.cookie};`;
        const reg = new RegExp(`${this.clientId}_${DEFAULT_TOKEN_CFG.name}=(.+?);`)
        const parts = cookie.match(reg)
        if (parts) return parts[1];
        return '';
    }

    clearToken() {
        document.cookie = util.parseCookie({
            value: '',
            name: `${this.clientId}_${DEFAULT_TOKEN_CFG.name}`,
            path: DEFAULT_TOKEN_CFG.path,
            domain: DEFAULT_TOKEN_CFG.domain,
            expires: 'expires=Thu, 01 Jan 1970 00:00:01 GMT',
        });
    }

    setCookieStr(name, value, expiredTime?) {
        const options={
            value,
            name,
            path: DEFAULT_TOKEN_CFG.path,
            domain: DEFAULT_TOKEN_CFG.domain,
            expires: expiredTime || "Fri, 31 Dec 9999 23:59:59 GMT", // 默认永久
        }
        document.cookie = util.parseCookie(options);
    }
}
