// @ss/components-manage-cli/dist/util/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZIP_DIR = exports.LOCK_FILE_NAME = exports.VERSION_INCREMENTS = exports.execBuffer = exports.SCREEN_SHOTS_PREFIX = exports.envHost = exports.materialTypeMap = exports.platformTextMap = exports.platformMap = exports.techstackMap = exports.techstackTextMap = exports.developTextMap = exports.developMap = exports.typeMap = void 0;
exports.typeMap = {
    'Component Lib': 1,
    'Template Lib': 2
};
exports.developMap = {
    'es': 1,
    'ts': 2
};
exports.developTextMap = {
    1: 'es',
    2: 'ts'
};
exports.techstackTextMap = {
    1: 'vue',
    2: 'react',
    3: 'mrn',
    4: 'miniapp',
    5: 'webcomponent',
    6: 'ios',
    7: 'android',
};
exports.techstackMap = {
    'vue': 1,
    'react': 2,
    'mrn': 3,
    'miniapp': 4,
    'webcomponent': 5,
    'ios': 6,
    'android': 7
};
exports.platformMap = {
    'Pc': 1,
    'Mobile': 2
};
exports.platformTextMap = {
    1: 'Pc',
    2: 'Mobile'
};
exports.materialTypeMap = {
    1: 'component',
    2: 'template'
};
exports.envHost = process.env.DEPLOY_ENV === 'test' ? 'http://material.bi.test.sankuai.com' : 'https://component.sankuai.com';
exports.SCREEN_SHOTS_PREFIX = './screenshots';
exports.execBuffer = 1 << 30;
exports.VERSION_INCREMENTS = [
    'prerelease:alpha',
    'prerelease:beta',
    'patch',
    'minor',
    'major',
];
exports.LOCK_FILE_NAME = 'template.lock';
exports.ZIP_DIR = 'zips';
