// @datafe/mdbi-common-tools/src/utils/storage.ts
const client = 'mdbi-common-tools'

export interface StorageData {
  [key: string]: StorageData | string | number | Array<string>
}

// 拼接完整的Key: localStorage
const getStorageKey = (key: string): string => {
  return [client, key].join('__')
}

/*
 * localStorage取值方法
 */
export const getLocalStorage = (key: string): StorageData => {
  const storageKey = getStorageKey(key)
  const data = window.localStorage.getItem(storageKey)
  return data ? JSON.parse(data) : {}
}

/*
 * localStorage存储方法
 */
export const setLocalStorage = (key: string, data: StorageData) => {
  const storageKey = getStorageKey(key)
  window.localStorage.setItem(storageKey, JSON.stringify(data))
}

export const removeLocalStorage = (key: string) => {
  const storageKey = getStorageKey(key)
  window.localStorage.removeItem(storageKey)
}

/*
 * sessionStorage取值方法
 */
export const getSessionStorage = (key: string): any => {
  const storageKey = getStorageKey(key)
  const data = window.sessionStorage.getItem(storageKey)
  return data ? JSON.parse(data) : {}
}

/*
 * sessionStorage存储方法
 */
export const setSessionStorage = (key: string, data: StorageData) => {
  const storageKey = getStorageKey(key)
  window.sessionStorage.setItem(storageKey, JSON.stringify(data))
}

export const removeSessionStorage = (key: string) => {
  const storageKey = getStorageKey(key)
  window.sessionStorage.removeItem(storageKey)
}
