// @datafe/mdbi-common-tools/src/utils/systemSetting.ts
declare interface SettingCollect {
  default: {
    clientId: string
  }
  sso: {
    [key: string]: {
      rewriteLocation: string
    }
  }
}
// 魔数各项目配置汇总:
const settingCollect: SettingCollect = {
  default: {
    clientId: 'com.sankuai.fetc.mdbi.home',
  },
  sso: {
    dashboard: {
      rewriteLocation: '/dashboard',
    },
    sql: {
      rewriteLocation: '/sql',
    },
    mail: {
      rewriteLocation: '/mail',
    },
    dataset: {
      rewriteLocation: '/dataset',
    },
    excel: {
      rewriteLocation: '/excel',
    },
    portal: {
      rewriteLocation: '/portal',
    },
    analysis: {
      rewriteLocation: '/analysis',
    },
    newgetdata: {
      rewriteLocation: '/newgetdata',
    },
    getdata: {
      rewriteLocation: '/getdata',
    },
    monitor: {
      rewriteLocation: '/monitor',
    },
    dacs: {
      rewriteLocation: '/dacs',
    },
    query: {
      rewriteLocation: '/query',
    },
    cube: {
      rewriteLocation: '/cube',
    },
    space: {
      rewriteLocation: '/space',
    },
    'source-manager': {
      rewriteLocation: '/source-manager',
    },
  },
}

export default class SystemSetting {
  private systemName: string
  private module: string

  constructor(module: string = '') {
    this.systemName = this.getSystemName()
    this.module = module
  }

  private getSystemName = () => {
    const pathList = window.location.pathname.split('/')
    return pathList[1] || ''
  }

  getSystemSetting() {
    const defaultSetting = settingCollect.default
    const moduleSetting = settingCollect[this.module]
    if (!moduleSetting) {
      return defaultSetting
    }
    return Object.assign({}, defaultSetting, moduleSetting[this.systemName])
  }
}
