// @datafe/mdbi-common-tools/src/permission/auth/constants.ts
export const actionMap = {
  DASHBOARD: {
    APP: {
      DELETE: 'DASHBOARD-APP-DELETE',
      CREATE: 'DASHBOARD-APP-CREATE',
      HANDOVER: 'DASHBOARD-APP-HANDOVER',
      GO_OFFLINE: 'DASHBOARD-APP-GO_OFFLINE',
      EDIT: 'DASHBOARD-APP-EDIT',
      VIEW: 'DASHBOARD-APP-VIEW',
      ROLLBACK: 'DASHBOARD-APP-ROLLBACK',
      GO_ONLINE: 'DASHBOARD-APP-GO_ONLINE',
      AUTH_APPLY: 'DASHBOARD-APP-AUTH_APPLY',
      AUTH_MANAGE: 'DASHBOARD-APP-AUTH_MANAGE',
      OUTBASE: 'DASHBOARD-APP-OUTBASE',
      REVOKE_OUTBASE: 'DASHBOARD-APP-REVOKE_OUTBASE',
      SYSTEM_SAFEMETA: 'DASHBOARD-APP-SYSTEM_SAFEMETA',
      EXPORT: 'DASHBOARD-APP-EXPORT',
    },
    DATASOURCE: { EXECUTION: 'DASHBOARD-DATASOURCE-EXECUTION' },
    DATAPUSH: {
      DELETE: 'DASHBOARD-DATAPUSH-DELETE',
      GO_OFFLINE: 'DASHBOARD-DATAPUSH-GO_OFFLINE',
      EDIT: 'DASHBOARD-DATAPUSH-EDIT',
      GO_ONLINE: 'DASHBOARD-DATAPUSH-GO_ONLINE',
      CREATE: 'DASHBOARD-DATAPUSH-CREATE',
      EXECUTION: 'DASHBOARD-DATAPUSH-EXECUTION',
      RECEIVE: 'DASHBOARD-DATAPUSH-RECEIVE',
    },
    COMPONENT: { VIEW: 'DASHBOARD-COMPONENT-VIEW' },
    FILTER: { VIEW: 'DASHBOARD-FILTER-VIEW' },
    TAB: { VIEW: 'DASHBOARD-TAB-VIEW' },
    SCHEDULE: {
      DELETE: 'DASHBOARD-SCHEDULE-DELETE',
      GO_OFFLINE: 'DASHBOARD-SCHEDULE-GO_OFFLINE',
      EXECUTION: 'DASHBOARD-SCHEDULE-EXECUTION',
      EDIT: 'DASHBOARD-SCHEDULE-EDIT',
      GO_ONLINE: 'DASHBOARD-SCHEDULE-GO_ONLINE',
      VIEW: 'DASHBOARD-SCHEDULE-VIEW',
    },
  },
  SQL: {
    APP: {
      CREATE: 'SQL-APP-CREATE',
      EDIT: 'SQL-APP-EDIT',
      EXECUTION: 'SQL-APP-EXECUTION',
      VIEW: 'SQL-APP-VIEW',
      AUTH_APPLY: 'SQL-APP-AUTH_APPLY',
      AUTH_MANAGE: 'SQL-APP-AUTH_MANAGE',
      DELETE: 'SQL-APP-DELETE',
      SAVE: 'SQL-APP-SAVE',
    },
    DATASOURCE: { EXECUTION: 'SQL-DATASOURCE-EXECUTION', SYSTEM_SAFEMETA: 'SQL-DATASOURCE-SYSTEM_SAFEMETA' },
    TABLE: { EXECUTION: 'SQL-TABLE-EXECUTION', SYSTEM_SAFEMETA: 'SQL-TABLE-SYSTEM_SAFEMETA' },
    RESULT: { VIEW: 'SQL-RESULT-VIEW', EXPORT: 'SQL-RESULT-EXPORT', KILL: 'SQL-RESULT-KILL' },
  },
  WORMHOLE: {
    APP: {
      CREATE: 'WORMHOLE-APP-CREATE',
      EDIT: 'WORMHOLE-APP-EDIT',
      VIEW: 'WORMHOLE-APP-VIEW',
      EXECUTION: 'WORMHOLE-APP-EXECUTION',
      DELETE: 'WORMHOLE-APP-DELETE',
      GO_ONLINE: 'WORMHOLE-APP-GO_ONLINE',
      GO_OFFLINE: 'WORMHOLE-APP-GO_OFFLINE',
      AUTH_APPLY: 'WORMHOLE-APP-AUTH_APPLY',
      AUTH_MANAGE: 'WORMHOLE-APP-AUTH_MANAGE',
    },
    DATASOURCE: { EXECUTION: 'WORMHOLE-DATASOURCE-EXECUTION' },
    RESULT: { EXPORT: 'WORMHOLE-RESULT-EXPORT', KILL: 'WORMHOLE-RESULT-KILL' },
  },
  MONITOR: {
    APP: {
      CREATE: 'MONITOR-APP-CREATE',
      EDIT: 'MONITOR-APP-EDIT',
      VIEW: 'MONITOR-APP-VIEW',
      EXECUTION: 'MONITOR-APP-EXECUTION',
      DELETE: 'MONITOR-APP-DELETE',
      GO_ONLINE: 'MONITOR-APP-GO_ONLINE',
      GO_OFFLINE: 'MONITOR-APP-GO_OFFLINE',
      AUTH_APPLY: 'MONITOR-APP-AUTH_APPLY',
      AUTH_MANAGE: 'MONITOR-APP-AUTH_MANAGE',
      RECEIVE: 'MONITOR-APP-RECEIVE',
      KILL: 'MONITOR-APP-KILL',
    },
    DATASOURCE: { EXECUTION: 'MONITOR-DATASOURCE-EXECUTION' },
  },
  PORTAL: {
    APP: {
      CREATE: 'PORTAL-APP-CREATE',
      EDIT: 'PORTAL-APP-EDIT',
      VIEW: 'PORTAL-APP-VIEW',
      DELETE: 'PORTAL-APP-DELETE',
      GO_ONLINE: 'PORTAL-APP-GO_ONLINE',
      GO_OFFLINE: 'PORTAL-APP-GO_OFFLINE',
      AUTH_APPLY: 'PORTAL-APP-AUTH_APPLY',
      AUTH_MANAGE: 'PORTAL-APP-AUTH_MANAGE',
    },
    TREE: {
      VIEW: 'PORTAL-TREE-VIEW',
    },
  },
  DATASET: {
    APP: {
      EDIT: 'DATASET-APP-EDIT',
      VIEW: 'DATASET-APP-VIEW',
      EXECUTION: 'DATASET-APP-EXECUTION',
      DELETE: 'DATASET-APP-DELETE',
      VIEW_SQL: 'DATASET-APP-VIEW_SQL',
      REFRESH: 'DATASET-APP-REFRESH',
      AUTH_APPLY: 'DATASET-APP-AUTH_APPLY',
      AUTH_MANAGE: 'DATASET-APP-AUTH_MANAGE',
    },
    DATASOURCE: {
      EXECUTION: 'DATASET-DATASOURCE-EXECUTION',
    },
  },
  LID: {
    SAFEMETA: {
      CHECK: 'LID-SAFEMETA-CHECK',
    },
  },
  NOT: { FOUND: { NOT_FOUND: 'NOT-FOUND-NOT_FOUND' } },
} as const

export type APP_TYPE = 'DASHBOARD' | 'SQL' | 'WORMHOLE' | 'MONITOR' | 'PORTAL' | 'LID' | 'DATAPUSH' | 'DATASET'

/** 权限申请状态 */
export enum ApplyStatus {
  SUCCESS = 'success',
  FAIL = 'fail',
  NO_OPERATE = 'no-operate',
}
