// @datafe/mdbi-common-tools/src/permission/auth/jiangjunling/jiangjunlingManager.ts
// @ts-ignore
import AuthApplyDynamic from '@mt-material/datafe-auth-apply-dynamic'
import JiangjunlingApplyModal from './JiangjunlingApplyModal.vue'
import { ExtendedVue } from 'vue/types/vue'
import Vue from 'vue'
import { ApplyStatus, APP_TYPE } from '../constants'

class JiangjunlingManager {
  private AuthApplyDynamicContructor: ExtendedVue<Vue, unknown, unknown, unknown, Record<never, any>> | null = null

  private getAuthApplyDynamicContructor() {
    if (!this.AuthApplyDynamicContructor) {
      Vue.use(AuthApplyDynamic)
      this.AuthApplyDynamicContructor = Vue.extend(JiangjunlingApplyModal)
    }
    return this.AuthApplyDynamicContructor
  }

  applyAuth(options: { id: string; mis: string; appType: APP_TYPE }) {
    let applyStatus = ApplyStatus.NO_OPERATE
    return new Promise<ApplyStatus>((resolve, reject) => {
      const params = {
        system: 'MOSHU', // 系统 code
        resourceCode: [`MOSHU::${options.appType}::${options.id}`], // 资源根节点 code
        misId: options.mis, // 申请用户
      }
      const AuthApplyDynamicContructor = this.getAuthApplyDynamicContructor()
      const instance = new AuthApplyDynamicContructor({
        data: {
          params,
          env: 'production',
          value: true,
        },
      })
      instance.$on('input', (v: boolean) => {
        if (!v) {
          resolve(applyStatus)
        }
      })
      instance.$on('status', (v: string) => {
        applyStatus = v as ApplyStatus
      })
      instance.$on('initFail', () => {
        reject('加载将军令申请组件失败')
      })
      instance.$mount()
      window.document.body.appendChild(instance.$el)
    })
  }
}

export default new JiangjunlingManager()
