// @ss/components-manage-cli/dist/contant/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FRAMEWORKS = exports.execBuffer = exports.serverHost = exports.SKETCH_INVERSE_FILENAME = exports.SKETCH_SCREEN_SHOTS_PREFIX = exports.SCREEN_SHOTS_PREFIX = exports.PACKAGE_PREFIX = exports.DISABLE_COMPONENT_NAME_INSPECTION = exports.DISABLE_CM_VERSION_INSPECTION = exports.DISABLE_UNIT_TEST = exports.DEPLOY_ENV = void 0;
exports.DEPLOY_ENV = process.env.deployEnv;
exports.DISABLE_UNIT_TEST = process.env.DISABLE_UNIT_TEST;
exports.DISABLE_CM_VERSION_INSPECTION = process.env.DISABLE_CM_VERSION_INSPECTION;
exports.DISABLE_COMPONENT_NAME_INSPECTION = true;
exports.PACKAGE_PREFIX = './components';
exports.SCREEN_SHOTS_PREFIX = './screenshots';
exports.SKETCH_SCREEN_SHOTS_PREFIX = './sketch-screenshots';
exports.SKETCH_INVERSE_FILENAME = 'sketch-inverse.js';
const serverHost = () => exports.DEPLOY_ENV === 'test' ? 'http://material.bi.test.sankuai.com' : 'https://component.sankuai.com';
exports.serverHost = serverHost;
exports.execBuffer = 1 << 30;
exports.FRAMEWORKS = ['vue', 'react', 'mrn', 'mmp'];
