// @ss/mtd-vue/es/components/input-group/input-group.js
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _obj;
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('span', _vm._g({ class: (_obj = {}, _obj[_vm.prefix] = true, _obj[_vm.prefix + "-compact"] = _vm.compact, _obj) }, _vm.$listeners), [_vm.$slots.prepend ? _c('div', { class: _vm.prefix + "-prepend" }, [_vm._t("prepend")], 2) : _vm._e(), _vm._t("default"), _vm.$slots.append ? _c('div', { class: _vm.prefix + "-append" }, [_vm._t("append")], 2) : _vm._e()], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdInputGroup',
  props: {
    compact: Boolean
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('input-group');
    }
  }
};
