// @ss/mtdv-mpa/components/mpa-condition-expression/dataSource/index.ts
export const conditionData = {
    name: "",
    conditionType: "and",
    childConditionList: [
        {
            name: '123123123',
            conditionType: "and",
            operatorList: ['add', 'delete'],
            childConditionList: [
                {
                    conditionType: "and",
                    data: {
                        input: "",
                        selector: 1,
                    },
                    comp: "ua-selector",
                    childConditionList: [
                        {
                            data: {
                                input: "xxxx",
                                selector: 1,
                            },
                            comp: "ua-selector",
                        },
                    ],
                },
                {
                    name: "",
                    comp: "ua-selector",
                    data: {
                        input: "xxxx",
                        selector: 1,
                    },
                },
            ],
        },
        {
            name: "xxxxx",
            conditionType: "and",
            operatorList: ['add', 'delete'],
            childConditionList: [
                {
                    data: {
                        input: "xxxx",
                        selector: 1,
                    },
                    comp: "ua-selector",
                },
                {
                    data: {
                        input: "xxxx",
                        selector: 1,
                    },
                    comp: "ua-selector",
                },
            ],
        },
        {
            name: "-----123123",
            comp: "ua-selector",
            data: {
                input: "xxxx",
                selector: 1,
            },
        },
    ],
};
