// @ss/components-manage-cli/dist/lowcode/lowcode.config.js
const path = require('path');
const mtdConfig = require(path.resolve(process.cwd(), './mtd.config.js'));
module.exports = {
    entry: {
        lowcode: path.resolve(process.cwd(), './lowcode/index.ts'),
    },
    resolve: {
        extensions: ['.ts', '.js'],
    },
    output: {
        library: mtdConfig.lowcode.outputName,
        libraryTarget: 'umd',
        filename: 'lowcode.umd.js',
        path: path.resolve(process.cwd(), './lib'),
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            presets: ['@babel/preset-env', '@babel/preset-typescript'],
                        },
                    },
                ],
            },
            {
                test: /\.js$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            presets: ['@babel/preset-env'],
                        },
                    },
                ],
            },
        ],
    },
    devServer: {
        hot: true,
        port: 8009,
        host: '127.0.0.1',
        historyApiFallback: true,
        headers: {
            'Access-Control-Allow-Origin': '*',
        },
    },
};
