// @mtfe/sso-web/example/react/src/App.js
import React, { Component } from 'react';
import './App.css';
import ssoWeb from './utils/sso';
import axios from 'axios';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLogout = this.handleLogout.bind(this);
    this.handleMock401Rep = this.handleMock401Rep.bind(this)
  }
  async handleLogout() {
    await ssoWeb.logout();
    window.location.href = ssoWeb.getLoginUrl();
  }
  async handleMock401Rep() {
    // 请求这个地址会返回
    /*
    {
      status: 401,
      data: {
        message: 'auth failed'
      }
    }
    */
    await axios.get('http://yapi.sankuai.com/mock/1232/mock401')
  }
  render() {
    return (
      <div className="App">
        <span style={{
          width: '50vw',
          wordBreak: 'break-word',
        }}>
          ssoid:<br />{this.props.ssoid}
        </span>
        <div>
          <button onClick={this.handleLogout}>登出</button>
          <button onClick={this.handleMock401Rep}>模拟ssoid失效</button>
        </div>
      </div>
    );
  }
}

export default App;
