// @ss/mtd-vue/es/components/select/select-tag.js
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';
import MtdTag from '@ss/mtd-vue/es/components/tag';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('li', { class: _vm.prefix + "-choice" }, [_vm._t("tag", [_c('mtd-tag', { class: { 'focus': _vm.option.hitState }, attrs: { "size": _vm.size, "theme": _vm.theme, "closable": _vm.closable, "disabled": _vm.disabled }, on: { "close": _vm.handleClose } }, [_vm._t("default")], 2)], { "option": _vm.option, "size": _vm.size, "closable": _vm.closable, "disabled": _vm.disabled, "onClose": _vm.handleClose })], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdSelectChoice',
  components: {
    MtdTag: MtdTag
  },
  props: {
    theme: String,
    size: String,
    option: {
      type: Object,
      required: true
    },
    closable: Boolean,
    disabled: Boolean
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('select');
    }
  },
  methods: {
    handleClose: function handleClose() {
      this.$emit('close', this.option);
    }
  }
};
