// @ss/mtd-vue/es/components/modal/inside.js
import Icon from '@ss/mtd-vue/es/components/icon';
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: _vm.prefix }, [_vm.closable ? _c('span', { class: _vm.prefix + "-close", on: { "click": _vm.handleClose } }, [_c('Icon', { attrs: { "name": "close-thick" } })], 1) : _vm._e(), _vm.title || _vm.$slots.title ? _c('div', { class: _vm.prefix + "-header" }, [_vm._t("title", [_c('div', { class: _vm.prefix + "-title" }, [_vm._v(_vm._s(_vm.title))])])], 2) : _vm._e(), _c('div', { class: _vm.prefix + "-content-wrapper" }, [_c('div', { class: _vm.prefix + "-content" }, [_vm._t("default")], 2)]), _vm.$slots.footer ? _c('div', { class: _vm.prefix + "-footer" }, [_vm._t("footer")], 2) : _vm._e()]);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'Modal',
  components: {
    Icon: Icon
  },
  inheritAttrs: false,
  props: {
    title: String,
    closable: Boolean
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('modal');
    }
  },
  methods: {
    handleClose: function handleClose() {
      this.$emit('close');
    }
  }
};
