// @ss/mtd-vue/es/components/modal/modal.js
import { hasClass } from '@ss/mtd-vue/es/utils/dom';
import { lock, unlock } from '@ss/mtd-vue/es/utils/lock-scroll';
import { PopupManage } from '@ss/mtd-vue/es/utils/popper';
import { isNumber, isDef } from '@ss/mtd-vue/es/utils/type';
import Modal from './inside';
import { getConfig, CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';
import { hasProps } from '@ss/mtd-vue/es/utils/vnode';

var __vue_render__ = function __vue_render__() {
  var _obj, _obj$1;
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', [_c('transition', { attrs: { "name": "fade-in" } }, [_vm.mask ? _c('div', { directives: [{ name: "show", rawName: "v-show", value: _vm.visible, expression: "visible" }], ref: "mask", class: _vm.prefix + "-mask", style: { 'z-index': _vm.zIndex }, on: { "click": _vm.handleMaskClick } }) : _vm._e()]), _c('div', { directives: [{ name: "show", rawName: "v-show", value: _vm.wrapVisible, expression: "wrapVisible" }], class: (_obj = {}, _obj[_vm.prefix + "-wrapper"] = true, _obj[_vm.prefix + "-" + _vm.placement] = _vm.placement, _obj[_vm.prefix + "-through"] = _vm.enableClickThrough, _obj), style: { 'z-index': _vm.zIndex }, on: { "click": _vm.handleWrapClick } }, [_c('transition', { attrs: { "name": "modal-fade" }, on: { "after-enter": _vm.handleAfterEnter, "after-leave": _vm.handleAfterLeave } }, [_vm.modalVisible ? _c('modal', _vm._b({ directives: [{ name: "show", rawName: "v-show", value: _vm.visible, expression: "visible" }], class: (_obj$1 = {}, _obj$1[_vm.prefix + "-fullscreen"] = _vm.fullscreen, _obj$1), style: { 'width': _vm.modalWidth }, attrs: { "closable": _vm.closable, "title": _vm.title }, on: { "close": _vm.handleClose } }, 'modal', _vm.$attrs, false), [_vm.$slots.title ? _vm._t("title", null, { "slot": "title" }) : _vm._e(), _vm._t("default"), _vm.$slots.footer ? _vm._t("footer", null, { "slot": "footer" }) : _vm._e()], 2) : _vm._e()], 1)], 1)], 1);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdModal',
  components: {
    'modal': Modal
  },
  inheritAttrs: true,
  model: {
    prop: 'visible'
  },
  props: {
    title: String,
    appendToContainer: {
      type: Boolean,
      default: true
    },
    getPopupContainer: Function,
    mask: {
      type: Boolean,
      default: true
    },
    closable: {
      type: Boolean,
      default: true
    },
    maskClosable: {
      type: Boolean,
      default: false
    },
    fullscreen: Boolean,
    visible: Boolean,
    destroyOnClose: Boolean,
    lockScroll: {
      type: Boolean,
      default: true
    },
    placement: {
      type: String,
      validator: function validator(v) {
        return ['top', 'center'].indexOf(v) > -1;
      },

      default: 'center'
    },
    width: [String, Number],
    mountOnCreate: {
      type: Boolean,
      default: true
    },
    enableClickThrough: {
      // to support: https://tt.sankuai.com/ticket/detail?id=40402170
      type: Boolean,
      default: false
    },
    keyboard: Boolean
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  data: function data() {
    return {
      modalVisible: !this.mountOnCreate ? this.visible : true,
      wrapVisible: this.visible,
      zIndex: 2000
    };
  },

  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('modal');
    },
    modalWidth: function modalWidth() {
      return this.width ? isNumber(this.width) ? this.width + 'px' : this.width : undefined;
    },
    getContainer: function getContainer() {
      return this.getPopupContainer || this.config.getPContainer || getConfig().getPopupContainer;
    }
  },
  watch: {
    visible: function visible(n) {
      n ? this.openModal() : this.closeModal();
    }
  },
  created: function created() {
    this.$on('esc', this.handleEscClose);
  },
  mounted: function mounted() {
    this.init();
  },
  activated: function activated() {
    this.init();
  },
  deactivated: function deactivated() {
    this.destroy();
    this._modal = undefined;
  },
  beforeDestroy: function beforeDestroy() {
    this.destroy();
  },

  methods: {
    init: function init() {
      if (this.visible) {
        this.openModal();
      }
    },
    destroy: function destroy() {
      this.closeModal(this);
      var parentNode = this.$el.parentNode;
      parentNode && parentNode.removeChild(this.$el);
    },
    createModal: function createModal() {
      if (this.appendToContainer) {
        var getContainer = this.getContainer;
        var parent = getContainer();
        parent.appendChild(this.$el);
      }
      return true;
    },
    openModal: function openModal() {
      this.modalVisible = true;
      this.wrapVisible = true;
      this.zIndex = PopupManage.nextZIndex();
      PopupManage.open(this);
      if (!this._modal) {
        this._modal = this.createModal();
      }
      if (this.lockScroll) {
        lock(document.body);
      }
      this.$emit('open');
    },
    closeModal: function closeModal() {
      PopupManage.close(this);
      if (this.lockScroll) {
        unlock(document.body);
      }
    },
    close: function close() {
      this.$emit('close');
      this.$emit('input', false);
    },
    handleAfterLeave: function handleAfterLeave() {
      if (this.destroyOnClose) {
        this.modalVisible = false;
      }
      this.wrapVisible = false;
      this.$emit('closed');
    },
    handleAfterEnter: function handleAfterEnter() {
      this.$emit('opened');
    },
    handleMaskClick: function handleMaskClick() {
      this.maskClosable && this.visible && this.close();
    },
    handleClose: function handleClose() {
      this.visible && this.close();
    },
    handleWrapClick: function handleWrapClick(event) {
      if (hasClass(event.target, this.prefix + '-wrapper')) {
        this.handleMaskClick(event);
      }
    },
    handleEscClose: function handleEscClose(e) {
      // isDef: "keyboard" must be configured as a property in Confirm
      if ((hasProps(this, 'keyboard') && isDef(this.keyboard) ? this.keyboard : this.closable) && this.visible) {
        this.close();
      }
    }
  }
};
