// @ss/mtd-vue/es/components/date-picker/base/quarter-table.js
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import dayjs, { clearHours } from '@ss/mtd-vue/es/utils/date';
import { deepCopy } from '@ss/mtd-vue/es/utils/util';
import mixin from './mixin';
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: _vm.classes }, _vm._l(_vm.cells, function (cell, i) {
    return _c('span', { key: i, class: _vm.getCellCls(cell), on: { "click": function click($event) {
          return _vm.handleClick(cell);
        } } }, [_c('em', [_vm._v(_vm._s(cell.text))])]);
  }), 0);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  mixins: [mixin],

  props: {/* in mixin */},
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('date-picker');
    },
    prefixCls: function prefixCls() {
      return this.prefix + '-cells';
    },
    classes: function classes() {
      var prefixCls = this.prefixCls;

      return ['' + prefixCls, prefixCls + '-quarter'];
    },
    year: function year() {
      return this.tableDate.getFullYear();
    },
    cells: function cells() {
      var _this = this;

      var cells = [];
      var cellTmpl = {
        text: '',
        selected: false,
        disabled: false
      };

      var selectedQuarters = this.dates.filter(Boolean).filter(function (date) {
        return date.getFullYear() === _this.year;
      }).map(function (date) {
        return _this.getQuarter(date);
      });
      var focusedDate = clearHours(new Date(this.focusedDate.getFullYear(), 0, 1));
      var disabledTestFn = typeof this.disabledDate === 'function' && this.disabledDate;
      for (var i = 1; i < 5; i++) {
        var cell = deepCopy(cellTmpl);
        cell.date = new Date(this.year, (i - 1) * 3, 1);
        cell.disabled = cell.date && disabledTestFn && disabledTestFn(cell.date);
        var day = clearHours(cell.date);
        cell.selected = selectedQuarters.indexOf(i) > -1;
        cell.focused = day === focusedDate;
        cell.text = 'Q' + i;
        cells.push(cell);
      }

      return cells;
    }
  },
  methods: {
    getQuarter: function getQuarter(date) {
      return dayjs(date).quarter();
    },
    getCellCls: function getCellCls(cell) {
      var _ref;

      var prefixCls = this.prefixCls;

      return [prefixCls + '-cell', (_ref = {}, _defineProperty(_ref, prefixCls + '-cell-selected', cell.selected), _defineProperty(_ref, prefixCls + '-cell-disabled', cell.disabled), _defineProperty(_ref, prefixCls + '-cell-focused', cell.focused), _defineProperty(_ref, prefixCls + '-cell-range', cell.range && !cell.start && !cell.end), _ref)];
    }
  }
};
