// @ss/mtd-vue/es/components/cascader/menus.js
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _Object$assign from 'babel-runtime/core-js/object/assign';
import _defineProperty from 'babel-runtime/helpers/defineProperty';
import _extends from 'babel-runtime/helpers/extends';
import MtdLoading from '@ss/mtd-vue/es/components/loading';
import Menu from './menu';
// import { isDef } from '@ss/mtd-vue/es/utils/type';
import { DEFAULT_FIELD_NAMES, FILTERED_NODE_KEY } from './util.js';
import { escapeRegexpString } from '@ss/mtd-vue/es/utils/util';
import _scrollIntoView from '@ss/mtd-vue/es/utils/scroll-into-view';
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';

function linkedToArray(node) {
  var result = [];
  var n = node;
  var i = 0;
  while (n) {
    result[i] = n;
    i++;
    n = n.$parent;
  }
  return result.reverse();
}

function defaultFilter(filter, items, fieldNames) {
  return items.some(function (option) {
    return new RegExp(escapeRegexpString(filter), 'i').test(option[fieldNames.label]);
  });
}

export var SELECT_ALL_VALUE = '__SELECT_ALL__';

var __vue_render__ = function __vue_render__() {
  var _obj;
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: (_obj = {}, _obj[_vm.prefix] = true, _obj[_vm.prefix + "-filtered"] = !!_vm.filter, _obj) }, [_vm.loading ? _c('div', { class: _vm.mp + "-loading", style: { 'width': _vm.menuWidth } }, [_vm._t("loading", [_c('mtd-loading', { attrs: { "message": _vm.loadingText } })])], 2) : _vm._l(_vm.showNodes, function (nodes, index) {
    return _c('Menu', { key: index, ref: "menus", refInFor: true, attrs: { "nodes": nodes, "value": _vm.value[index], "no-data-text": _vm.noDataText, "no-match-text": _vm.noMatchText, "filtered": !!_vm.filter, "menu-width": _vm.menuWidth }, on: { "expand": _vm.handleExpand, "select": _vm.handleSelect, "click": _vm.handleClick, "checked": _vm.handleChecked, "mouseleave": _vm.handleMouseLeave }, scopedSlots: _vm._u([{ key: "default", fn: function fn(ref) {
          var node = ref.node;
          var data = ref.data;
          return [_vm._t("default", [_c('span', [_vm._v(_vm._s(node.label))])], { "node": node, "data": data })];
        } }], null, true) });
  })], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdCascaderMenus',
  components: {
    Menu: Menu,
    MtdLoading: MtdLoading
  },
  inheritAttrs: false,
  props: {
    data: Array,
    // 单选时当前选中项
    value: {
      type: Array, // 展开值
      default: function _default() {
        return [];
      }
    },
    expandedValue: {
      type: Array, // 展开值
      default: function _default() {
        return [];
      }
    },
    noDataText: {
      type: String,
      default: '暂无数据'
    },
    expandTrigger: {
      type: String,
      default: 'click',
      validator: function validator(v) {
        return ['click', 'hover'].indexOf(v) > -1;
      }
    },
    props: {
      type: Object,
      default: function _default() {
        return {};
      }
    },
    loadData: Function,
    loading: Boolean,
    loadingText: String,
    filterable: Boolean,
    filterParent: Boolean, // 是否要过滤掉 父节点
    filter: String,
    filterMethod: {
      type: Function,
      default: defaultFilter
    },
    noMatchText: {
      type: String,
      default: '暂无搜索结果'
    },
    menuWidth: String,
    multiple: Boolean,
    checkedValues: {
      // 二维数组，内部每一个数组表示一个链路
      type: Array, // 展开值
      default: function _default() {
        return [];
      }
    },
    checkStrictly: Boolean,
    checkedStrategy: String,
    changeOnSelect: Boolean,
    expandableOnDisabled: {
      // see: https://ones.sankuai.com/ones/product/4348/workItem/requirement/detail/9014200
      type: Boolean,
      default: false
    },
    disabledStrictly: {
      type: Boolean,
      default: true
    },
    defaultExpandFirstItem: {
      type: Boolean,
      default: false
    },
    showSelectAll: {
      type: Boolean,
      default: false
    },
    updateAfterLoaded: Boolean
  },
  provide: function provide() {
    return {
      menus: this
    };
  },

  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  data: function data() {
    return {
      forceNumber: 1,
      checkedNodes: [],
      unmatchedValues: [],

      // 当 data 由于异步更新的时候，已选节点和data的匹配情况需要重新计算
      // 首次是需要重新计算的
      isNeedRecalculateCheckedNode: true
    };
  },

  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('cascader-menus');
    },
    mp: function mp() {
      return this.config.getPrefixCls('cascader-menu');
    },
    fieldNames: function fieldNames() {
      return _extends({}, DEFAULT_FIELD_NAMES, this.props);
    },
    root: function root() {
      var mapping = {};
      var root = this.buildNodes(this.createNodes(this.data, mapping));
      if (this.forceNumber) {
        // 仅仅是让 root 重新计算
        root.forceNumber = this.forceNumber;
      }
      this.mapping = mapping; // eslint-disable-line
      return root;
    },
    menuNodes: function menuNodes() {
      return this.buildMenuNodes(this.root, []);
    },
    flattenNodes: function flattenNodes() {
      // 扁平化对象，仅当 filterable 时生效，格式为 [[...parentNode, childrenNode]]
      if (this.filterable) {
        return this.getFlattenNodes(this.root);
      }
      return [];
    },
    filteredNodes: function filteredNodes() {
      var _this = this;

      // 搜索结果
      var fieldNames = this.fieldNames;

      return this.flattenNodes.filter(function (nodes) {
        var paths = nodes.map(function (node) {
          return node.data;
        });
        return _this.filterMethod(_this.filter, paths, fieldNames);
      }).map(function (nodes) {
        var ns = nodes.filter(function (n) {
          return !n.isSelectAll;
        });
        var labels = ns.map(function (n) {
          return n.label;
        }).join(' / ');
        var values = ns.map(function (n) {
          return n.value;
        });
        var node = nodes[nodes.length - 1];
        var disabled = node.disabled;

        return _defineProperty({
          label: labels,
          value: values,
          isLeaf: true,
          disabled: disabled,
          checked: node.checked,
          data: node.data,
          nodes: ns,
          isSelectAll: node.isSelectAll
        }, FILTERED_NODE_KEY, true);
      });
    },
    showNodes: function showNodes() {
      if (this.filterable && this.filter) {
        return [this.filteredNodes];
      } else {
        return this.menuNodes;
      }
    },
    showCheckbox: function showCheckbox() {
      return this.multiple;
    }
  },
  watch: {
    data: function data(v) {
      this.isNeedRecalculateCheckedNode = true;
    }
  },
  methods: {
    createNodes: function createNodes(data, mapping, parent) {
      var _data,
          _this2 = this;

      var level = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
      var fieldNames = this.fieldNames,
          showSelectAll = this.showSelectAll,
          loadData = this.loadData;

      var nodes = !showSelectAll ? [] : [{
        $parent: parent,
        $index: level,
        level: level,
        data: (_data = {}, _defineProperty(_data, fieldNames.label, '全选'), _defineProperty(_data, fieldNames.value, SELECT_ALL_VALUE), _data),
        value: SELECT_ALL_VALUE,
        label: '全选',
        isLeaf: true,
        disabled: false,
        loading: false,
        selected: false,
        checked: false,
        isSelectAll: true,
        hasChildrenChecked: false,
        indeterminate: false,
        children: null,
        hasChildren: false
      }];

      return nodes.concat(data.map(function (item, i) {
        var disabledField = fieldNames.disabled;

        var node = {
          $parent: parent,
          $index: i,
          level: level,
          data: item,

          value: item[fieldNames.value],
          label: item[fieldNames.label],
          isLeaf: item[fieldNames.isLeaf],
          disabled: disabledField in item ? item[disabledField] : !_this2.disabledStrictly ? false : parent && parent.disabled,
          loading: item[fieldNames.loading],

          selected: false,
          checked: false,
          hasChildrenChecked: false,
          indeterminate: false
        };
        var children = item[fieldNames.children];
        children = children && children.length ? children : undefined;
        _Object$assign(node, {
          children: children ? _this2.createNodes(children, mapping, node, level + 1) : undefined,
          hasChildren: !!children || loadData && !node.isLeaf
        });
        mapping[node.value] = node;
        return node;
      }));
    },
    buildNodes: function buildNodes(nodes) {
      var _this3 = this;

      if (!this.multiple) {
        return nodes;
      }
      var checkedNodes = [];
      this.unmatchedValues = [];
      this.checkedValues.forEach(function (pathValue) {
        var matched = false;
        pathValue.reduce(function (items, v, i) {
          var node = items.find(function (item) {
            return item.value === v;
          });
          if (!node) return [];
          node.hasChildrenChecked = true;
          if (i === pathValue.length - 1) {
            matched = true;
            node.checked = true;
            // 计算时忽略 disabled 节点
            !node.disabled && checkedNodes.push(node);
          }
          return node.children || [];
        }, nodes);
        if (!matched) {
          _this3.unmatchedValues.push(pathValue);
        }
      });
      if (!this.checkStrictly) {
        checkedNodes.forEach(this.updateChildChecked);
        checkedNodes.forEach(this.updateParentChecked);
      }
      if (this.showSelectAll) {
        var first = nodes[0]; // 全选节点
        first.checked = nodes.every(function (node) {
          return node.isSelectAll || node.checked;
        });
      }
      return nodes;
    },
    buildMenuNodes: function buildMenuNodes(nodes, menuNodes) {
      var level = menuNodes.length;
      menuNodes[level] = nodes || [];

      var value = this.expandedValue[level];

      var node = nodes.find(function (item) {
        return item.value === value;
      });
      if (node && !node.loading && node.children) {
        this.buildMenuNodes(node.children, menuNodes);
      }
      return menuNodes;
    },
    getFlattenNodes: function getFlattenNodes(nodes) {
      var _this4 = this;

      var paths = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

      var flatNodes = [];
      nodes.forEach(function (node) {
        var children = node.children;
        var hasChildren = children && children.length;
        var stack = paths.concat([node]);
        if (!hasChildren) {
          flatNodes.push(stack);
        } else {
          if (!_this4.filterParent) {
            flatNodes.push(stack);
          }
          flatNodes = flatNodes.concat(_this4.getFlattenNodes(children, stack));
        }
      });
      return flatNodes;
    },
    creataLoadDataCallback: function creataLoadDataCallback(data, node) {
      var _this5 = this;

      var that = this;
      return function (children) {
        data[that.fieldNames.loading] = false;
        that.$set(data, that.fieldNames.children, children);
        var parent = that.$parent;
        if (parent && parent.updatePopper) {
          parent.updatePopper();
        }
        if (_this5.updateAfterLoaded) {
          _this5.handleChecked(node, { checked: node.checked });
        }
      };
    },
    updateParentChecked: function updateParentChecked(node) {
      var parent = node.$parent;

      if (parent && !parent.disabled) {
        var children = parent.children;

        var allChecked = children.filter(function (node) {
          return !node.isSelectAll;
        }).every(function (child) {
          return child.checked || child.disabled;
        });

        parent.checked = allChecked;
        parent.indeterminate = !allChecked;
        if (this.showSelectAll) {
          children[0].checked = parent.checked;
        }
        this.updateParentChecked(parent);
      }
    },
    updateChildChecked: function updateChildChecked(node) {
      var _this6 = this;

      var children = node.children;

      if (children) {
        children.forEach(function (item) {
          if (!item.checked && !item.disabled) {
            item.checked = true;
            _this6.updateChildChecked(item);
          }
        });
      }
    },
    updateParentUnChecked: function updateParentUnChecked(node) {
      var parent = node.$parent;

      if (parent) {
        var children = parent.children;

        var anyChecked = children.filter(function (node) {
          return !node.isSelectAll;
        }).some(function (child) {
          // 忽略 disabled 的选项
          return !child.disabled && (child.checked || child.indeterminate);
        });

        parent.checked = false;
        parent.indeterminate = anyChecked;
        this.updateParentUnChecked(parent);
      }
    },
    updateChildUnChecked: function updateChildUnChecked(node) {
      var _this7 = this;

      var children = node.children;

      if (children) {
        children.forEach(function (item) {
          if (!item.disabled) {
            item.checked = false;
            _this7.updateChildUnChecked(item);
          }
        });
      }
    },
    handleExpand: function handleExpand(node) {
      var _this8 = this;

      var handler = function handler() {
        if (!node) {
          _this8.$emit('expanded-change', [], []);
          return;
        }
        if (node.hasChildren) {
          if (!node.children && !node.loading) {
            // 动态加载
            var data = node.data;

            _this8.$set(data, _this8.fieldNames.loading, true);
            _this8.loadData(data, _this8.creataLoadDataCallback(data, node));
          }

          var _getValue = _this8.getValue(node),
              values = _getValue.values,
              nodes = _getValue.nodes;

          _this8.$emit('expanded-change', values, nodes);
        }
      };
      if (this.expandTrigger === 'hover') {
        this.timer = setTimeout(handler, 150);
      } else {
        handler();
      }
    },
    handleMouseLeave: function handleMouseLeave() {
      if (this.timer) {
        clearTimeout(this.timer);
        this.timer = null;
      }
    },
    handleSelect: function handleSelect(node) {
      var _getValue2 = this.getValue(node),
          values = _getValue2.values,
          nodes = _getValue2.nodes;

      this.$emit('select', values, nodes);
    },
    handleClick: function handleClick(node) {
      var _getValue3 = this.getValue(node),
          values = _getValue3.values,
          nodes = _getValue3.nodes;

      this.$emit('click', values, nodes);
    },
    handleAllChecked: function handleAllChecked(root, _ref2) {
      var _this9 = this;

      var checked = _ref2.checked;

      root.forEach(function (node) {
        node.checked = checked;
        if (checked) {
          _this9.updateChildChecked(node);
        } else {
          _this9.updateChildUnChecked(node);
        }
      });
      var checkedNodes = this.getCheckedNode(this.root);
      var checkedPaths = checkedNodes.map(function (node) {
        return linkedToArray(node);
      });
      var nextCheckedValues = this.getCheckedValue(checkedNodes);
      var _nextCheckedValues = [].concat(_toConsumableArray(this.unmatchedValues), _toConsumableArray(nextCheckedValues));
      this.$emit('update:checkedValues', _nextCheckedValues, checkedPaths, root);
      this.setNextRefresh();
    },
    setChecked: function setChecked(node, _ref3) {
      var _this10 = this;

      var checked = _ref3.checked;

      if (!node[FILTERED_NODE_KEY] && node.isSelectAll) {
        node = node.$parent;
        if (!node) {
          return this.handleAllChecked(this.root, { checked: checked });
        }
      }

      var nextCheckedValues = void 0;

      if (this.isNeedRecalculateCheckedNode) {
        // 能够保证 nextCheckedValues 不包含非匹配项的代码
        var checkedNodes = this.getCheckedNode(this.root);
        nextCheckedValues = this.getCheckedValue(checkedNodes);
        this.isNeedRecalculateCheckedNode = false;
      } else {
        nextCheckedValues = [].concat(_toConsumableArray(this.checkedValues));
      }

      var _getValue4 = this.getValue(node),
          nodes = _getValue4.nodes,
          values = _getValue4.values;

      var lastNode = nodes[nodes.length - 1];
      var checkedPaths = [];
      if (this.checkStrictly) {
        if (checked) {
          nextCheckedValues.push(values);
        } else {
          nextCheckedValues = nextCheckedValues.filter(function (checkedValues) {
            return checkedValues.length !== values.length || checkedValues.some(function (v, i) {
              return v !== values[i];
            });
          });
        }
        checkedPaths = nextCheckedValues.map(function (checkedValue) {
          return _this10.getPathByValue(_this10.root, checkedValue);
        });
      } else {
        lastNode.checked = checked;
        if (checked) {
          this.updateParentChecked(lastNode);
          this.updateChildChecked(lastNode);
        } else {
          this.updateParentUnChecked(lastNode);
          this.updateChildUnChecked(lastNode);
        }
        var _checkedNodes2 = this.getCheckedNode(this.root);
        checkedPaths = _checkedNodes2.map(function (node) {
          return linkedToArray(node);
        });
        nextCheckedValues = this.getCheckedValue(_checkedNodes2);
      }
      var _nextCheckedValues = [].concat(_toConsumableArray(this.unmatchedValues), _toConsumableArray(nextCheckedValues));
      return {
        nextCheckedValues: _nextCheckedValues,
        checkedPaths: checkedPaths,
        lastNode: lastNode
      };
    },
    handleChecked: function handleChecked(node, _ref4) {
      var checked = _ref4.checked;

      var r = this.setChecked(node, { checked: checked });
      /**
       * 全选在 setChecked 里面抛出了事件，外面不再处理
       */
      if (!r) {
        return;
      }
      var nextCheckedValues = r.nextCheckedValues,
          checkedPaths = r.checkedPaths,
          lastNode = r.lastNode;

      this.$emit('update:checkedValues', nextCheckedValues, checkedPaths, lastNode);
      this.setNextRefresh();
    },
    getNodes: function getNodes(values) {
      var nodes = [];
      values.reduce(function (items, v, i) {
        var node = items.find(function (item) {
          return item.value === v;
        });
        if (!node) return [];
        nodes.push(node);
        return node.children || [];
      }, this.root);
      return nodes;
    },
    getPathByValue: function getPathByValue(nodes, pathValue) {
      var paths = [];
      pathValue.reduce(function (items, v, i) {
        var node = (items || []).find(function (item) {
          return item.value === v;
        });
        node && paths.push(node);
        return node && node.children;
      }, nodes);
      return paths;
    },
    getValue: function getValue(node) {
      var nextValue = void 0;
      var nodes = void 0;
      if (node[FILTERED_NODE_KEY]) {
        var value = node.value,
            ns = node.nodes;

        nextValue = value;
        nodes = ns;
      } else {
        nodes = linkedToArray(node);
        nextValue = nodes.map(function (node) {
          return node.value;
        });
      }
      return {
        nodes: nodes,
        values: nextValue
      };
    },
    setValue: function setValue(node) {
      var nextValue = void 0;
      var nodes = void 0;
      if (node[FILTERED_NODE_KEY]) {
        var value = node.value,
            ns = node.nodes;

        nextValue = value;
        nodes = ns;
      } else {
        nodes = linkedToArray(node);
        nextValue = nodes.map(function (node) {
          return node.value;
        });
      }
      this.$emit('input', nextValue);
      this.$emit('expand', nextValue, nodes);
    },
    scrollIntoView: function scrollIntoView() {
      var _this11 = this;

      // scroll all
      (this.$refs.menus || []).forEach(function (menu) {
        var el = menu.$el;
        _scrollIntoView(el, el.getElementsByClassName(_this11.mp + '-item-active')[0]);
      });
    },
    getCheckedNode: function getCheckedNode(nodes) {
      var checkedStrategy = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.checkedStrategy;

      var checkedNodes = [];
      function flatChecked(node) {
        var children = node.children,
            checked = node.checked,
            isSelectAll = node.isSelectAll;

        if (checked && !isSelectAll) {
          checkedNodes.push(node);
        }
        children && children.forEach(flatChecked);
      }
      nodes.forEach(flatChecked);
      if (checkedStrategy === 'children') {
        return checkedNodes.filter(function (node) {
          return !node.children;
        });
      } else if (checkedStrategy === 'parent') {
        return checkedNodes.filter(function (node) {
          var $parent = node.$parent;

          return !$parent || node.disabled || !$parent.checked;
        });
      };
      return checkedNodes;
    },
    getCheckedValue: function getCheckedValue(checkedNodes) {
      var _checkedNodes = checkedNodes || this.getCheckedNode(this.root);
      return _checkedNodes.map(function (node) {
        var paths = linkedToArray(node);
        var values = paths.map(function (node) {
          return node.value;
        });
        return values;
      });
    },
    setNextRefresh: function setNextRefresh() {
      this.forceNumber++;
    },
    getNode: function getNode(value) {
      return this.mapping[value];
    }
  }
};
