// @ss/mtd-vue/es/components/picker/control.js
import createHoc from '@ss/mtd-vue/es/utils/hoc';
import Picker from './picker';

export default createHoc({
  name: 'MtdPicker',
  model: {
    prop: 'value'
  },
  props: {
    defaultVisible: {
      type: Boolean,
      default: false
    },
    visible: Boolean
  },
  data: function data() {
    return {
      vs: this.defaultVisible
    };
  },

  computed: {
    isControlled: function isControlled() {
      return 'visible' in this.$options.propsData;
    },
    _visible: function _visible() {
      return this.isControlled ? this.visible : this.vs;
    }
  },
  methods: {
    handleVisibleChange: function handleVisibleChange(v) {
      this.$emit('update:visible', v);
      if (!this.isControlled) {
        this.vs = v;
      }
    },
    updatePopper: function updatePopper() {
      var wrappedInstance = this.$refs.wrappedInstance;

      wrappedInstance && wrappedInstance.updatePopper();
    },
    getPopper: function getPopper() {
      var wrappedInstance = this.$refs.wrappedInstance;

      return wrappedInstance.$refs.popper;
    },
    focus: function focus() {
      var wrappedInstance = this.$refs.wrappedInstance;

      wrappedInstance.focus();
    },
    blur: function blur() {
      var wrappedInstance = this.$refs.wrappedInstance;

      wrappedInstance.blur();
    }
  }
}, {
  withRef: true,
  mapStateToProps: function mapStateToProps(context) {
    return {
      visible: context._visible
    };
  },
  mapMethodToListener: function mapMethodToListener(context) {
    return {
      'update:visible': context.handleVisibleChange
    };
  }
})(Picker);
