// @ss/mtd-vue/components/tabs/src/tab-one.js
import tabMore from './tab-more';
import Tab from './tab';
import Icon from '@components/icon';

export default {
  name: 'MtdTabOne',

  components: {
    tabMore,
    Icon,
  },

  props: {
    size: String,
    disabled: Boolean,
    active: Boolean,
    label: String,
    pane: {
      type: Object,
      required: true,
    },
    paneLabel: Array,
    icon: String,
    tabDropInfo: Object,
    currentValue: [String, Number, Object, Array],
    closable: Boolean,
    dropAutosize: {
      type: Boolean,
      default: false,
    },
    prefix: {
      type: String,
      required: true,
    },
    getPrefixCls: {
      type: Function,
      required: true,
    },
    tabComponent: [String, Object],
  },

  data () {
    return {
      labelContent: null,
      iconhover: false,
    };
  },
  updated () {
    this.$parent.repaint && this.$parent.repaint();
  },

  mounted () {
    this.$parent.addTabs && this.$parent.addTabs(this);
    this.pane.$tab = this;
  },

  methods: {
    emitTabClick (e) {
      const value = !this.tabDropInfo ? this.pane.value
        : this.tabDropInfo.$parent.currentValue;
      !this.disabled && this.$emit('tabClick', e, value);
    },
    handleClick (e) {
      const { moreTabs } = this.tabDropInfo || {};
      if (!moreTabs || !moreTabs.length) {
        this.emitTabClick(e);
      }
    },
    closeClick (e) {
      !this.disabled && this.$emit('tabRemove', e);
    },
    iconhoverenterEvent () {
      this.iconhover = true;
    },
    iconhoverleaveEvent () {
      this.iconhover = false;
    },
  },
  render (h) {
    const {
      size,
      disabled,
      active,
      icon,
      label,
      paneLabel,
      emitTabClick,
      handleClick,
      tabDropInfo,
      currentValue,
      closable,
      closeClick,
      iconhover,
      iconhoverenterEvent,
      iconhoverleaveEvent,
      dropAutosize,
      prefix,
      tabComponent,
    } = this;
    const { moreTabs, $parent } = tabDropInfo || {};
    let tabMoreData = null;
    $parent && ($parent.currentValue = currentValue);
    if (moreTabs && moreTabs.length) {
      tabMoreData = {
        props: {
          moreTabs,
          currentValue,
          label,
          icon,
          disabled,
          size,
          dropAutosize,
          prefix,
          tabComponent,
        },
        on: {
          tabSelectClick: (value) => {
            if (value) {
              $parent.currentValue = value;
              emitTabClick(undefined, value);
            }
          },
        },
        ref: 'more',
      };
    }

    // 如多每次都重新复制，会导致多次渲染问题
    // [Vue warn]: You may have an infinite update loop
    // in a component render function.

    // this.labelContent = this.labelContent || (
    //   <tab-more {...tabMoreData}></tab-more>
    // );
    const tabPrefix = this.getPrefixCls('tab');

    const children = <div class={`${prefix}-item-label`}>
      {tabMoreData && <tab-more {...tabMoreData}></tab-more>}
      {!tabMoreData && icon && <i
        class={{
          [`${tabPrefix}-icon`]: true,
          [`${icon}`]: icon,
        }}
      ></i>}
      {!tabMoreData && (paneLabel && paneLabel.length ? paneLabel : label)}
      {closable && <Icon name={iconhover ? 'error-circle' : 'close-thick'}
        class={`${prefix}-tab-close`}
        onClick={closeClick}
        onMouseenter={iconhoverenterEvent}
        onMouseleave={iconhoverleaveEvent}
      />}
    </div>;
    const attrs = {
      class: [{
        [`${prefix}-item`]: true,
        [`${prefix}-item-${size}`]: true,
        [`${prefix}-item-disabled`]: disabled,
        [`${tabPrefix}-active`]: active,
        [`${prefix}-closable`]: closable,
      }, this.pane.labelClass],
      on: {
        click: (ev) => {
          handleClick(ev);
        },
      },
      props: {
        tab: this.pane,
      },
    };
    const Component = tabComponent || Tab;
    return (
      <Component { ...attrs }>
        { children }
      </Component>
    );
  },
};
