// @ss/components-manage-tools/src/util/constant.ts
export const envHost = process.env.DEPLOY_ENV === 'test' ? 'http://material.bi.test.sankuai.com' : 'https://component.sankuai.com';

export const registerUrlMap = {
    'component': '/openapi/v1/develop/jointLib',
    'template': '/openapi/v1/develop/jointTemplate'
};

export const registerParamsKeyMap = {
    'component': {
        "teamId": '',
        "libName": '',
        "framework": '',
        "platform": '',
        'componentOrigin': '',
        "developLanguage": '',
        "templateId": '',
        "description": '',
        "npmName": '',
        "classPrefix": '',
        "administrator": '',
        "source": '',
        "gitUrl": ''
    },
    'template': {
        "teamId": '',
        "templateName": '',
        "framework": '',
        "platform": '',
        "developLanguage": '',
        "templateId": '',
        "description": '',
        "administrator": '',
        "source": ''
    },
};

export const getSignature = async () => {
    console.log('--- get signature start---');
    const { Auth } = require('@dp/patriot-sdk');
    const _auth = new Auth('com.sankuai.bfestatic.jointcomponent');
    const signature = await _auth.getTokenSignature('com.sankuai.bfestatic.jointcomponent');
    console.log('--- get signature end ---');
    return signature;
};

export const publishUrlMap = {
    'component': '/openapi/third/publish',
    'template': '/openapi/third/publish/template'
};

export const createPublishUrlMap = {
    'component': '/openapi/publish',
    'template': '/openapi/publish/template'
};
