// @ss/components-manage-tools/src/util/log.ts
const boxen = require('boxen');
const chalk = require('chalk');

export function boxenLog(message: string, options: any = {}) {
    // tslint:disable-next-line
    console.log(
        boxen(
            message,
            Object.assign(
                {
                    padding: { left: 4, right: 4, top: 1, bottom: 1 },
                    margin: 1,
                    borderStyle: 'round',
                    borderColor: 'green',
                },
                options,
            ),
        ),
    );
}

export function welcome() {
    boxenLog(`${chalk.green('Welcome to use mtd manage commander')}`);
}

export function fatalLog(content: string) {
    // tslint:disable-next-line
    console.info(chalk.red(`[fatal]: ${content}`));
    process.exit(-1);
}

export function successLog(content: string) {
    // tslint:disable-next-line
    console.info(chalk.green(`[success]: ${content}`));
}

export function warningLog(content: string) {
    // tslint:disable-next-line
    console.info(chalk.yellow(`[warning]: ${content}`));
}

export function infoLog(content: string) {
    // tslint:disable-next-line
    console.info(chalk.blueBright(`[info]: ${content}`));
}
