// @ss/mtd-vue/es/components/steps/steps.js
import { CONFIG_PROVIDER, getPrefixCls, getSize } from '@ss/mtd-vue/es/utils/config';
var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: [_vm.prefix, _vm._size && _vm.prefix + "-" + _vm._size, !_vm.isCenter && _vm.prefix + "-right", _vm.direction && _vm.prefix + "-" + _vm.direction] }, [_vm._t("default")], 2);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdSteps',
  props: {
    size: String,
    dot: {
      type: Boolean,
      default: false
    },
    space: [String, Number],
    direction: {
      type: String,
      default: 'horizontal'
    },
    active: {
      type: Number,
      default: 0
    },
    status: {
      type: String,
      default: 'process'
    },
    isCenter: {
      type: Boolean,
      default: true
    },
    simple: Boolean
  },
  provide: function provide() {
    return {
      steps: this
    };
  },

  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls,
        getSize: getSize
      }
    }
  },
  data: function data() {
    return {
      steps: []
    };
  },

  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('steps');
    },
    _size: function _size() {
      return this.config.getSize(this);
    }
  },
  watch: {
    steps: function steps(_steps) {
      _steps.forEach(function (child, index) {
        child.index = index;
      });
    }
  },
  created: function created() {
    this.$on('addStep', this.add);
    this.$on('removeStep', this.remove);
  },

  methods: {
    add: function add(step) {
      this.steps.push(step);
    },
    remove: function remove(step) {
      var index = this.steps.indexOf(step);
      if (index > -1) {
        this.steps.splice(index, 1);
      }
    }
  }
};
