// @ss/mtd-vue/es/components/steps/with.js
import createHoc from '@ss/mtd-vue/es/utils/hoc';

export default function withSteps(name, Components) {
  return createHoc({
    name: name,
    inject: ['steps'],
    props: {
      status: String
    },
    data: function data() {
      return {
        index: -1 // use in steps component
      };
    },
    created: function created() {
      this.steps.$emit('addStep', this);
    },
    beforeDestroy: function beforeDestroy() {
      this.steps.$emit('removeStep', this);
    },


    computed: {
      internalStatus: function internalStatus() {
        if (this.index > -1) {
          var val = this.steps.active;
          if (this.index < val) {
            return 'finish';
          } else if (this.index === val) {
            return this.steps.status || 'process';
          } else if (this.index > val) {
            return 'wait';
          }
        }
      },
      currentStatus: function currentStatus() {
        return this.status || this.internalStatus;
      }
    }
  }, {
    mapStateToProps: function mapStateToProps(context) {
      return {
        index: context.index,
        status: context.currentStatus,
        space: context.steps.space,
        direction: context.steps.direction,
        dot: context.steps.dot
      };
    }
  })(Components);
}
