// @ss/mtd-vue/es/components/card/card.js
import { CONFIG_PROVIDER, getPrefixCls } from '@ss/mtd-vue/es/utils/config';
var __vue_render__ = function __vue_render__() {
  var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { class: [_vm.prefix, _vm.shadow ? _vm.prefix + "-" + _vm.shadow + "-shadow" : _vm.prefix + "-always-shadow"] }, [_vm.$slots.title || _vm.title ? _c('div', { class: [_vm.prefix + "-title", _vm.titleClass] }, [_vm._t("title", [_vm._v(_vm._s(_vm.title))])], 2) : _vm._e(), _vm.$slots.extra ? _c('div', { class: _vm.prefix + "-extra" }, [_vm._t("extra")], 2) : _vm._e(), _c('div', { class: [_vm.prefix + "-body", _vm.bodyClass] }, [_vm._t("default")], 2)]);
};
var __vue_staticRenderFns__ = [];

export default {
  render: __vue_render__,

  staticRenderFns: __vue_staticRenderFns__,

  name: 'MtdCard',
  props: {
    title: {
      type: String
    },
    shadow: {
      type: String,
      default: 'always'
    },
    bodyClass: String,
    titleClass: String
  },
  inject: {
    config: {
      from: CONFIG_PROVIDER,
      default: {
        getPrefixCls: getPrefixCls
      }
    }
  },
  computed: {
    prefix: function prefix() {
      return this.config.getPrefixCls('card');
    }
  }
};
