// @mtfe/sc-web/webpack.config.dev.js
const path = require('path');
const webpack = require('webpack');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const packageJson = require('./package.json');

module.exports = {
    mode: process.env.NODE_ENV || 'development',
    entry: path.resolve(__dirname, './src/dev.ts'),
    // 仅用于错误调试
    devtool: 'source-map',
    devServer: {},
    output: {
        filename: 'sc-web.js',
        path: path.resolve(__dirname, './dist'),
        publicPath: '/',
        auxiliaryComment: `/* Version ${packageJson.version}*/`,
        // 生成前清空输出目录
        clean: true
    },
    module: {
        rules: [{
            test: /\.js$/,
            exclude: /(node_modules|bower_components)/,
            use: {
                loader: 'babel-loader',
                options: {
                    presets: ['@babel/preset-env'],
                    plugins: ['@babel/plugin-proposal-function-bind']
                }
            }
        },
        {
            test: /\.ts$/,
            loader: "ts-loader",
            exclude: '/node_modules/'
        }]
    },
    resolve: {
        extensions: [ '.js', '.ts' ]
    },
    plugins: [
        new webpack.DefinePlugin({
            'process.env': JSON.stringify({
                ...process.env
            })
        }),
        new HtmlWebpackPlugin({
            title: '服务组件化-SCWeb',
            template: path.resolve(__dirname, './public/index.html')
        })
    ],
};
