// @datafe/mdbi-common-tools/src/permission/api/safemeta.ts
import request from './config/axios.config'
export interface GetSafemetaInfoRes {
  // 资源是否包含安全meta
  safeMeta: boolean
  // 出仓状态 OUTBASED,OUTBASING,UNOUTBASE
  outbaseStatus: 'OUTBASED' | 'OUTBASING' | 'UNOUTBASE'
  // 当前IP是否是堡垒机
  isSafeEnvironment: boolean
}

export function outbase(options: {
  mis: string
  message: string
  resource: {
    id: string
    appType: string
    subType: string
  }
}): Promise<{
  code: number
  message: string
  data?: boolean
}> {
  return request.post('/safe/safemeta/outbase', options)
}

export function revokeOutbase(options: {
  mis: string
  resource: {
    id: string
    appType: string
    subType: string
  }
}): Promise<{
  code: number
  message: string
  data?: boolean
}> {
  return request.post('/safe/safemeta/revokeOutbase', options)
}

export function getSafemetaInfo(options: { id: string; appType: string }): Promise<{
  code: number
  message: string
  data?: GetSafemetaInfoRes
}> {
  return request.post('/safe/safemeta/safeInfo', options)
}
