// @mtfe/sso-web/src/config/const.ts
/**
 * Created by leon on 6/14/18.
 */
interface SSO_CONFIG_URL_TYPE {
  SSO_CONFIG_URL: string;
}

interface ConfigAPIUrl {
  readonly [key: string]: SSO_CONFIG_URL_TYPE;
}

export const CONFIG_API_URL: ConfigAPIUrl = {
  ppe: {
    SSO_CONFIG_URL: '//ssoconfigs.it.beta.sankuai.com',
  },
  test: {
    SSO_CONFIG_URL: '//ssoconfigs.it.test.sankuai.com',
  },
  dev: {
    SSO_CONFIG_URL: '//ssoconfigs.it.dev.sankuai.com',
  },
  product: {
    SSO_CONFIG_URL: 'https://ssosv.sankuai.com/configs',
  },
  staging: {
    SSO_CONFIG_URL: 'https://ssosv.sankuai.com/configs',
  },
};

interface SSOSpecialRoute {
  readonly callbackUrl: string;
  readonly logoutUrl: string;
  readonly robotUrl: string;
}

export const SSO_SPECIAL_ROUTE: SSOSpecialRoute = {
  callbackUrl: '/sso/callback',
  logoutUrl: '/sso/logout',
  robotUrl: '/sso/robot',
};

interface DefaultSSOAccessConfig {
  readonly [key: string]: any;
}

export const DEFAULT_SSO_ACCESS_CONFIG: DefaultSSOAccessConfig = {
  dev: {
    accessEnv: 'dev',
    http: {
      openApiHost: '//ssosv.it.test.sankuai.info/open',
      ssoSvHost: '//ssosv.it.test.sankuai.info/sson',
    },
    id: 'dev_master',
    loginUrl: '//ssosv.it.test.sankuai.info/sson/login',
    thrift: {
      openApiThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoopenapi',
        remoteServerPort: 9002,
      },
      ssoSvThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoservice',
        remoteServerPort: 9002,
      },
    },
  },
  test: {
    accessEnv: 'test',
    http: {
      openApiHost: '//ssosv.it.test.sankuai.com/open',
      ssoSvHost: '//ssosv.it.test.sankuai.com/sson',
    },
    id: 'test_master',
    loginUrl: '//ssosv.it.test.sankuai.com/sson/login',
    thrift: {
      openApiThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoopenapi',
        remoteServerPort: 9002,
      },
      ssoSvThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoservice',
        remoteServerPort: 9002,
      },
    },
  },
  staging: {
    accessEnv: 'staging',
    http: {
      openApiHost: '//ssosv.it.st.sankuai.com/open',
      ssoSvHost: '//ssosv.it.st.sankuai.com/sson',
    },
    id: 'staging_master',
    loginUrl: '//ssosv.it.st.sankuai.com/sson/login',
    thrift: {
      openApiThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoopenapi',
        remoteServerPort: 9002,
      },
      ssoSvThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoservice',
        remoteServerPort: 9002,
      },
    },
  },
  ppe: {
    accessEnv: 'ppe',
    http: {
      openApiHost: '//ssosv.it.beta.sankuai.com/open',
      ssoSvHost: '//ssosv.it.beta.sankuai.com/sson',
    },
    id: 'ppe_master',
    loginUrl: '//ssosv.it.beta.sankuai.com/sson/login',
    thrift: {
      openApiThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoopenapi',
        remoteServerPort: 9002,
      },
      ssoSvThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoservice',
        remoteServerPort: 9002,
      },
    },
  },
  product: {
    accessEnv: 'product',
    http: {
      openApiHost: 'https://ssosv.sankuai.com/open',
      ssoSvHost: 'https://ssosv.sankuai.com/sson',
    },
    id: 'product_master',
    loginUrl: 'https://ssosv.sankuai.com/sson/login',
    thrift: {
      openApiThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoopenapi',
        remoteServerPort: 9002,
      },
      ssoSvThrift: {
        remoteAppkey: 'com.sankuai.it.iam.ssoservice',
        remoteServerPort: 9002,
      },
    },
  },
};

interface ConfigMappingModel {
  readonly loginUrl: {
    layer: string;
  };
  readonly accessEnv: {
    layer: string;
  };
  readonly openApiHost: {
    layer: string;
  };
  readonly ssoSvHost: {
    layer: string;
  };
}

export const configMappingModel: ConfigMappingModel = {
  loginUrl: {
    layer: 'accessPoint',
  },
  accessEnv: {
    layer: 'accessPoint',
  },
  openApiHost: {
    layer: 'accessPoint.http',
  },
  ssoSvHost: {
    layer: 'accessPoint.http',
  },
};

export const HARD_CODE_METHOD = 'SHA256';

export const WEB_DEFAULT_IP = '0.0.0.0';
