// @ss/mtd-vue/script/publish/talos-message.js
const https = require('https');
const mtdPackage = require('../../package.json');
const execa = require('execa');

const hostname = 'mufasa.sankuai.com';

async function talosHooks () {
  const { stdout: branch } = await execa('git', ['branch', '--show-current']);
  const data = {
    "target":{
      "ssh_url": mtdPackage.repository.url,
    },
    "to": `refs/heads/${branch}`,
  }
  const postData = JSON.stringify(data);
  console.log("notify talos broadcast release");
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: hostname,
      path: '/mufasa-api/common/webhook',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json; charset=utf-8'
      },
      timeout: 2000
    }, function (res) {
      const responseBuffer = [];
      res.on('data', function (chunk) {
        responseBuffer.push(chunk);
      });
      res.on('end', function () {
        let responseData = Buffer.concat(responseBuffer);
        responseData = responseData.toString('utf-8');
        resolve(responseData);
      });
    });
    req.on('error', function (error) {
      reject(error);
    });
    req.write(postData);
    req.end();
    setTimeout(() => {
      req.abort();
      reject('abort request')
    }, 2000);
  })
}
module.exports = talosHooks;
