// @mtfe/sc-web/lib/config/const.js
export const DEFAULT_SC_ACCESS_CONFIG = {
    test: {
        accessEnv: 'test',
        http: {
            apiHost: 'https://proxy.fe.test.sankuai.com',
        },
        id: 'test_master',
        loginUrl: 'https://proxy.fe.test.sankuai.com/login',
        loginOrigin: 'https://proxy.fe.test.sankuai.com',
    },
    staging: {
        accessEnv: 'staging',
        http: {
            apiHost: 'https://proxy.fe.st.sankuai.com',
        },
        id: 'staging_master',
        loginUrl: 'https://proxy.fe.st.sankuai.com/login',
        loginOrigin: 'https://proxy.fe.st.sankuai.com',
    },
    production: {
        accessEnv: 'production',
        http: {
            apiHost: 'https://proxy.sankuai.com',
            apiHost_out_mt: 'https://service-component.meituan.com',
            apiHost_out_sk: 'https://proxy-pub.sankuai.com',
        },
        id: 'production_master',
        loginUrl: 'https://proxy.sankuai.com/login',
        loginOrigin: 'https://proxy.sankuai.com',
        loginUrl_out_mt: 'https://service-component.meituan.com/login',
        loginOrigin_out_mt: 'https://service-component.meituan.com',
        loginUrl_out_sk: 'https://proxy-pub.sankuai.com/login',
        loginOrigin_out_sk: 'https://proxy-pub.sankuai.com',
    }
};
export const PROXY_PATH = 'sc-proxy';
export const LOGIN_COUNTER = 5;
export const IFRAME_ID = '__sc_login_iframe__';
export const LOGIN_TIMEOUT = 10000;
export const LOGIN_THROTTLE_TIME = 1000;
